---
title: "Mouse Immune HiMyc"
output: html_document
date: "2023-01-10"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Immune Figures
Subset mouse seurat object integrated by strain for immune cell clusters 

```{r}
# Libraries
library(Seurat)
library(dplyr)
library(readr)
library(DataCombine)
library(viridis)
library(ggplot2)
library(glmGamPoi)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

```

# Subset and examine immune cell clusters
sctransform seurat object,dimensionality reduction by PCA and UMAP embedding
```
# load seurat object
strain.combined.sct <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")
DimPlot(strain.combined.sct)

# subset to immune cells
immune_seurat <- subset(strain.combined.sct, idents = c("Macrophages", "T cells", "Mast cells"))
DimPlot(immune_seurat, group.by = "cell_type")

# run sctransform
immune_seurat <- SCTransform(immune_seurat, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE)

# dimensionality reduction
immune_seurat <- RunPCA(immune_seurat, verbose = FALSE, npcs = 50)
immune_seurat <- RunUMAP(immune_seurat, reduction = "pca", dims = 1:50)
immune_seurat <- FindNeighbors(immune_seurat, dims = 1:50, verbose = FALSE)

# cluster resolution
immune_seurat <- FindClusters(immune_seurat, verbose = FALSE, resolution = 0.4)
DimPlot(immune_seurat, label = TRUE, group.by = "SCT_snn_res.0.4", label.size = 6) + NoLegend()

```

Identity of immune cell types
```
Immune <- "Ptprc"
Macrophages <- c("C1qb","C1qa","C1qc","Cd68")
Mast_cells <- c("Mcemp1", "Flt3")
Tcells <- c("Cd2","Cd3d", "Cd3g")


immune_genes <- c(Immune, Macrophages, Mast_cells, Tcells)

Basal <- c("Krt5", "Krt15", "Krt14")
Epithelial <- c("Krt8", "Krt18")
Luminal <- "Agr2"
Foxi1 <- c("Foxi1")

epithelial_gens <- c(Basal, Epithelial, Luminal, Foxi1)

DoHeatmap(subset(immune_seurat, downsample = 100), assay = "SCT", features = c(immune_genes, epithelial_gens), group.by = "SCT_snn_res.0.4") +
  scale_fill_viridis()
```

Filter epithelial cluster (cluster 6) and perform SCTransform
```
immune_seurat_v2 <- subset(immune_seurat, idents = 6, invert = TRUE)
DimPlot(immune_seurat, label = TRUE)
DimPlot(immune_seurat_v2, label = TRUE)

# run sctransform
immune_seurat_v2 <- SCTransform(immune_seurat_v2, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE)

# dimensionality reduction
immune_seurat_v2 <- RunPCA(immune_seurat_v2, verbose = FALSE, npcs = 50)
immune_seurat_v2 <- RunUMAP(immune_seurat_v2, reduction = "pca", dims = 1:50)
immune_seurat_v2 <- FindNeighbors(immune_seurat_v2, dims = 1:50, verbose = FALSE)

# cluster resolution
immune_seurat_v2 <- FindClusters(immune_seurat_v2, verbose = FALSE, resolution = 0.2)

Idents(immune_seurat_v2) <- "SCT_snn_res.0.2"
DoHeatmap(subset(immune_seurat_v2, downsample = 100), assay = "SCT", features = c(immune_genes), group.by = "SCT_snn_res.0.2") +
  scale_fill_viridis()

# save immune seurat
saveRDS(immune_seurat_v2, file = "mouse_immune_seurat.rds")
```

# strain specific differences in immune cells
0 - Macrophage BL6
1 - Macrophage FVB
2 - Mast cells
3 - Macrophage
4- T cells
```
# previous vs new clustering
table(immune_seurat_v2$genotype, immune_seurat_v2$SCT_snn_res.0.2, immune_seurat_v2$strain)
table(immune_seurat_v2$strain, immune_seurat_v2$SCT_snn_res.0.2)

# update clustering name
Idents(immune_seurat_v2) <- "SCT_snn_res.0.2"
new.cluster.ids <- c("Macrophages Cd72", 
                     "Macrophages Cxcl2", "Mast cells", "Macrophages Trem2","T cells")

names(new.cluster.ids) <- levels(immune_seurat_v2)
immune_seurat_v2[["immune"]] <- new.cluster.ids[as.numeric(as.matrix(immune_seurat_v2@meta.data$SCT_snn_res.0.2))+1]
immune_seurat_v2 <- RenameIdents(immune_seurat_v2, new.cluster.ids)
```

```{r}
immune_seurat_v2 <- readRDS(file = "mouse_immune_seurat.rds")
DimPlot(immune_seurat_v2, label = TRUE) + NoLegend() + ggtitle("Immune Mouse Clusters")
```

# Cell proportions test by RAISIN
https://github.com/zji90/raisin//
```
# add immune metadata to complete dataset
strain.combined.sct$immune <- strain.combined.sct$cell_type
strain.combined.sct$immune_subset <- immune_seurat_v2$immune
strain.combined.sct$immune <- paste(strain.combined.sct$immune, strain.combined.sct$immune_subset, sep = "")
Idents(strain.combined.sct) <- "immune"
table(strain.combined.sct$immune)

strain.combined.sct$immune <- gsub("NA", "", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("T cellsT cells", "T cells", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("T cellsMast cells", "Mast cells", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("Mast cellsMast cells", "Mast cells", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("MacrophagesMacrophages", "Macrophages", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("Mast cellsMacrophages", "Macrophages", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("MacrophagesT cells", "T cells", strain.combined.sct$immune)
strain.combined.sct$immune <- gsub("MacrophagesMast cells", "Mast cells", strain.combined.sct$immune)

Idents(strain.combined.sct) <- "immune"
levels(strain.combined.sct)
table(strain.combined.sct$immune)

strain.combined.filter <- subset(strain.combined.sct, idents = "Macrophages", invert = TRUE)
levels(strain.combined.filter)
table(strain.combined.filter$immune)

# save updated seurat object
saveRDS(strain.combined.filter, file = "strain.combined.filter.rds")
```
cell proportion by genotype
```{r}
library(raisin)

strain.combined.filter <- readRDS(file = "strain.combined.filter.rds")

# extract metadata from seurat object
Idents(strain.combined.filter) <- "immune"
levels(strain.combined.filter)
cluster <- Idents(strain.combined.filter)

Idents(strain.combined.filter) <- "library_id"
levels(strain.combined.filter)
individual <- Idents(strain.combined.filter)

# make the design 
design <- data.frame(intercept=1,contrast=c(0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1),
                     row.names = c("AB1", "AB2", "AF1", "AF2", "AHMB1",
                                   "AHMB2", "AHMB3", "AHMF1", "AHMF2",
                                   "DB1", "DB2", "DF1", "DF2", "DHMB1",
                                   "DHMB2", "DHMB3", "DHMF1", "DHMF2",
                                   "LB1", "LB2", "LF1", "LF2", "LHMB1",
                                   "LHMB2", "LHMB3", "LHMF1", "LHMF2",
                                   "VB1", "VB2", "VF1", "VF2", "VHMB1",
                                   "VHMB2", "VHMB3", "VHMF1", "VHMF2"))

# run the test
prop_test_results_genotype <- proportiontest(cluster,individual,design)
prop_test_results_genotype <- prop_test_results_genotype %>% tibble::rownames_to_column("Cluster")
colnames(prop_test_results_genotype) <- c("Cluster", "Genotype logFC", "Genotype P.Value", "Genotype adj.P.Val")
prop_test_results_genotype

```

cell proportion by strain
```{r}
library(raisin)

# extract metadata from seurat object
Idents(strain.combined.filter) <- "immune"
levels(strain.combined.filter)
cluster <- Idents(strain.combined.filter)

Idents(strain.combined.filter) <- "library_id"
levels(strain.combined.filter)
individual <- Idents(strain.combined.filter)

# make the design 
design <- data.frame(intercept=1,contrast=c(0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1),
                     row.names = c("AB1", "AB2", "AHMB1","AHMB2", "AHMB3",
                                   "AF1", "AF2", "AHMF1", "AHMF2",
                                   "DB1", "DB2", "DHMB1", "DHMB2", "DHMB3", 
                                   "DF1", "DF2", "DHMF1", "DHMF2",
                                   "LB1", "LB2", "LHMB1","LHMB2", "LHMB3",
                                   "LF1", "LF2", "LHMF1", "LHMF2",
                                   "VB1", "VB2", "VHMB1","VHMB2", "VHMB3",
                                   "VF1", "VF2", "VHMF1", "VHMF2"))

# run the test
prop_test_results_strain <- proportiontest(cluster,individual,design)
prop_test_results_strain <- prop_test_results_strain %>% tibble::rownames_to_column("Cluster")
colnames(prop_test_results_strain) <- c("Cluster", "Strain logFC", "Strain P.Value", "Strain adj.P.Val")

prop_test_results_strain
```

Plot cell proportion strain differences of immune cells
```{r}
library(ggpubr)
library(patchwork)
library_counts <- as.data.frame(table(strain.combined.filter$library_id))

# create dataframe containing all relevant metadata
cell_type_enrich <- as.data.frame(table(strain.combined.filter$library_id,strain.combined.filter$strain, strain.combined.filter$genotype, strain.combined.filter$immune))

cell_type_enrich$library <- library_counts$Freq
cell_type_enrich$fraction <- cell_type_enrich$Freq / cell_type_enrich$library

# rename columns to meaningful column names
colnames(cell_type_enrich) <- c("Library", "Strain", "Genotype", "Cluster", "Counts", "Library_Counts", "Fraction")

# filter to library
AB1 <- cell_type_enrich %>% filter(Library == "AB1", Strain == "C57BL/6J", Genotype == "WT")
AB2 <- cell_type_enrich %>% filter(Library == "AB2", Strain == "C57BL/6J", Genotype == "WT")
AF1 <- cell_type_enrich %>% filter(Library == "AF1", Strain == "FVB/NJ", Genotype == "WT")
AF2 <- cell_type_enrich %>% filter(Library == "AF2", Strain == "FVB/NJ", Genotype == "WT")
AHMB1 <- cell_type_enrich %>% filter(Library == "AHMB1", Strain == "C57BL/6J", Genotype == "WT")
AHMB2 <- cell_type_enrich %>% filter(Library == "AHMB2", Strain == "C57BL/6J", Genotype == "MYC")
AHMB3 <- cell_type_enrich %>% filter(Library == "AHMB3", Strain == "C57BL/6J", Genotype == "MYC")
AHMF1 <- cell_type_enrich %>% filter(Library == "AHMF1", Strain == "FVB/NJ", Genotype == "MYC")
AHMF2 <- cell_type_enrich %>% filter(Library == "AHMF2", Strain == "FVB/NJ", Genotype == "MYC")
DB1 <- cell_type_enrich %>% filter(Library == "DB1", Strain == "C57BL/6J", Genotype == "WT")
DB2 <- cell_type_enrich %>% filter(Library == "DB2", Strain == "C57BL/6J", Genotype == "WT")
DF1 <- cell_type_enrich %>% filter(Library == "DF1", Strain == "FVB/NJ", Genotype == "WT")
DF2 <- cell_type_enrich %>% filter(Library == "DF2", Strain == "FVB/NJ", Genotype == "WT")
DHMB1 <- cell_type_enrich %>% filter(Library == "DHMB1", Strain == "C57BL/6J", Genotype == "WT")
DHMB2 <- cell_type_enrich %>% filter(Library == "DHMB2", Strain == "C57BL/6J", Genotype == "MYC")
DHMB3 <- cell_type_enrich %>% filter(Library == "DHMB3", Strain == "C57BL/6J", Genotype == "MYC")
DHMF1 <- cell_type_enrich %>% filter(Library == "DHMF1", Strain == "FVB/NJ", Genotype == "MYC")
DHMF2 <- cell_type_enrich %>% filter(Library == "DHMF2", Strain == "FVB/NJ", Genotype == "MYC")
LB1 <- cell_type_enrich %>% filter(Library == "LB1", Strain == "C57BL/6J", Genotype == "WT")
LB2 <- cell_type_enrich %>% filter(Library == "LB2", Strain == "C57BL/6J", Genotype == "WT")
LF1 <- cell_type_enrich %>% filter(Library == "LF1", Strain == "FVB/NJ", Genotype == "WT")
LF2 <- cell_type_enrich %>% filter(Library == "LF2", Strain == "FVB/NJ", Genotype == "WT")
LHMB1 <- cell_type_enrich %>% filter(Library == "LHMB1", Strain == "C57BL/6J", Genotype == "WT")
LHMB2 <- cell_type_enrich %>% filter(Library == "LHMB2", Strain == "C57BL/6J", Genotype == "MYC")
LHMB3 <- cell_type_enrich %>% filter(Library == "LHMB3", Strain == "C57BL/6J", Genotype == "MYC")
LHMF1 <- cell_type_enrich %>% filter(Library == "LHMF1", Strain == "FVB/NJ", Genotype == "MYC")
LHMF2 <- cell_type_enrich %>% filter(Library == "LHMF2", Strain == "FVB/NJ", Genotype == "MYC")
VB1 <- cell_type_enrich %>% filter(Library == "VB1", Strain == "C57BL/6J", Genotype == "WT")
VB2 <- cell_type_enrich %>% filter(Library == "VB2", Strain == "C57BL/6J", Genotype == "WT")
VF1 <- cell_type_enrich %>% filter(Library == "VF1", Strain == "FVB/NJ", Genotype == "WT")
VF2 <- cell_type_enrich %>% filter(Library == "VF2", Strain == "FVB/NJ", Genotype == "WT")
VHMB1 <- cell_type_enrich %>% filter(Library == "VHMB1", Strain == "C57BL/6J", Genotype == "WT")
VHMB2 <- cell_type_enrich %>% filter(Library == "VHMB2", Strain == "C57BL/6J", Genotype == "MYC")
VHMB3 <- cell_type_enrich %>% filter(Library == "VHMB3", Strain == "C57BL/6J", Genotype == "MYC")
VHMF1 <- cell_type_enrich %>% filter(Library == "VHMF1", Strain == "FVB/NJ", Genotype == "MYC")
VHMF2 <- cell_type_enrich %>% filter(Library == "VHMF2", Strain == "FVB/NJ", Genotype == "MYC")

filtered_library_fraction <- rbind(AB1, AB2, AF1, AF2,
                                   AHMB1, AHMB2, AHMB3, AHMF1, AHMF2,
                                   DB1, DB2, DF1, DF2,
                                   DHMB1, DHMB2, DHMB3, DHMF1, DHMF2,
                                   LB1, LB2, LF1, LF2, 
                                   LHMB1, LHMB2, LHMB3, LHMF1, LHMF2,
                                   VB1, VB2, VF1, VF2,
                                   VHMB1, VHMB2, VHMB3, VHMF1, VHMF2)

# combine DF with cell proportion test results
combined_df <- merge(filtered_library_fraction, prop_test_results_genotype, by.x = "Cluster", by.y = "Cluster")
combined_df <- merge(combined_df, prop_test_results_strain, by.x = "Cluster", by.y = "Cluster")

# Add pvalue to cluster name
library(scales)
combined_df$strain_adjP <- paste(combined_df$Cluster,
                                  "\n FVB vs C57BL6",
                                  "\n logFC = ", 
                                  signif(combined_df$`Strain logFC`, 2),
                                  "\nadj p-value = ", scientific(combined_df$`Strain adj.P.Val`, 1))

combined_df$genotype_adjP <- paste(combined_df$Cluster,
                                  "\n Hi-Myc vs WT",
                                  "\n logFC = ", 
                                  signif(combined_df$`Genotype logFC`, 2),
                                  "\nadj p-value = ", scientific(combined_df$`Genotype P.Value`, 1))

# order x axis
combined_df$Genotype <- factor(combined_df$Genotype, levels = c("WT",  "MYC"))

# Plot immune clusters by strain
combined_df %>% 
  filter(Cluster %in% c("Macrophages Cd72", "Macrophages Cxcl2", "Macrophages Trem2",
                        "Mast cells", "T cells")) %>%
  ggplot(aes(x = `Strain`, y = Fraction))+
  geom_jitter(aes(fill = Genotype, color = Genotype), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Immune Clusters Difference by Strain") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~strain_adjP, scales = "free", ncol = 3) +
  scale_color_manual(values = c("#a8dadc", "#e63946")) 
  

# Plot immune clusters by genotype
combined_df %>% 
  filter(Cluster %in% c("Macrophages Cd72", "Macrophages Cxcl2", "Macrophages Trem2",
                        "Mast cells", "T cells")) %>%
  ggplot(aes(x = Genotype, y = Fraction))+
  geom_jitter(aes(fill = Strain, color = Strain), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Immune Clusters Difference by Genotype") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~genotype_adjP, scales = "free", ncol = 3) +
  scale_color_manual(values = c("#219ebc", "#ffb703"))  

```

Differential expression by seurat
```
# differential gene expression by cluster
Idents(immune_seurat_v2) <- "immune"
levels(immune_seurat_v2) <- c("Macrophages Trem2" , "Macrophages Cd72", "Macrophages Cxcl2",
                              "Mast cells", "T cells")


seurat_immune_genes <- FindAllMarkers(immune_seurat_v2)
top_seurat_immune <- seurat_immune_genes %>% group_by(cluster) %>% filter(avg_log2FC > 0) %>% top_n(10, -p_val_adj)

Idents(immune_seurat_v2) <- "strain_immune"
levels(immune_seurat_v2) <- c("FVB/NJ Macrophages Trem2", "C57BL/6J Macrophages Trem2",
                              "FVB/NJ Macrophages Cd72", "C57BL/6J Macrophages Cd72",
                              "FVB/NJ Macrophages Cxcl2", "C57BL/6J Macrophages Cxcl2",
                              "FVB/NJ Mast cells", "C57BL/6J Mast cells",
                              "FVB/NJ T cells", "C57BL/6J T cells")  
DoHeatmap(subset(immune_seurat_v2, downsample = 25), assay = "SCT", features = top_seurat_immune$gene, draw.lines = TRUE) +
  scale_fill_viridis()

table(immune_seurat_v2$strain_immune)

saveRDS(seurat_immune_genes, file = "seurat_immune_genes.rds")
```

Heatmap of cell type strain specific differences
```{r}
seurat_immune_genes <- readRDS(file = "seurat_immune_genes.rds")
top_seurat_immune <- seurat_immune_genes %>% group_by(cluster) %>% filter(avg_log2FC > 0) %>% top_n(10, -p_val_adj)

immune_seurat_v2$strain_immune <- paste(immune_seurat_v2$strain, immune_seurat_v2$immune, sep = " ")
Idents(immune_seurat_v2) <- "strain_immune"
levels(immune_seurat_v2) <- c("FVB/NJ Macrophages Trem2", "C57BL/6J Macrophages Trem2",
                              "FVB/NJ Macrophages Cd72", "C57BL/6J Macrophages Cd72",
                              "FVB/NJ Macrophages Cxcl2", "C57BL/6J Macrophages Cxcl2",
                              "FVB/NJ Mast cells", "C57BL/6J Mast cells",
                              "FVB/NJ T cells", "C57BL/6J T cells")  

# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# expression matrix
Idents(immune_seurat_v2) <- "strain_immune"
AvgExp <- AverageExpression(immune_seurat_v2)
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# filter to top differentially expressed genes of macrophage clusters
AvgExp_Strain <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(top_seurat_immune$gene)) %>%
   as.matrix()

# colors
RdYlBu <- brewer.pal(6, "RdYlBu")

# column orders
 sample_order <- c("FVB/NJ Macrophages Trem2", "C57BL/6J Macrophages Trem2",
                              "FVB/NJ Macrophages Cd72", "C57BL/6J Macrophages Cd72",
                              "FVB/NJ Macrophages Cxcl2", "C57BL/6J Macrophages Cxcl2",
                              "FVB/NJ Mast cells", "C57BL/6J Mast cells",
                              "FVB/NJ T cells", "C57BL/6J T cells")  
# complex heatmap
ht <- Heatmap(AvgExp_Strain,
        show_column_dend = FALSE,
        show_row_dend = FALSE,
        row_names_side = "left",
        column_dend_reorder = FALSE,
        column_order = sample_order,
        row_order = c(top_seurat_immune$gene),
        column_title = "Immune Cluster Genes",
        col = rev(RdYlBu),
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm")
))
draw(ht, padding = unit(c(2, 30, 2, 2), "mm"))


```

UMAPs
```{r}
Idents(immune_seurat_v2) <- "immune"
DimPlot(immune_seurat_v2, label = TRUE, raster = TRUE, pt.size = 2) + NoLegend() + 
  ggtitle("Mouse Immune Clusters") +
  theme(plot.title = element_text(hjust = 0.5))


Idents(immune_seurat_v2) <- "strain"
levels(immune_seurat_v2) <- c("C57BL/6J", "FVB/NJ")
DimPlot(immune_seurat_v2, label = FALSE, raster = TRUE, cols =  c("#219ebc", "#ffb703"), shuffle = TRUE, pt.size = 2) +
  ggtitle("Mouse Strain") +
  theme(plot.title = element_text(hjust = 0.5))

#Trem2
pal1 <- c("#3B9AB2", "#78B7C5", "#E1AF00", "#F21A00")

FeaturePlot(immune_seurat_v2, reduction = "umap", features = "Trem2", order = TRUE, raster = TRUE, pt.size = 2)+ 
  scale_color_gradientn(colours = pal1)

VlnPlot(immune_seurat_v2, features = "Trem2", group.by = "immune", split.by = "strain", cols =  c("#219ebc", "#ffb703"))

```
