---
title: "MYC mice prostate"
author: "Mindy Kim Graham"
date: "6/14/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Single cell analysis of mouse prostate (WT and Hi-MYC) at 6 months and 10 months
Here we merge libraries from the dorsal and lateral lobes of FVB mice aged ~10 months in WT and HiMYC mouse strains. At 6 months, HiMYC FVB mice predominately have PIN in DLV, while at 10 months DL lobes have significant invasive carcinoma.
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(sctransform)
library(dplyr)
library(DataCombine)
library(viridis)
library(tidyverse)
library(glmGamPoi)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

```

# Load gene x cell count matrices, merge and generate a seurat object
FASTQ and gene x cell count matrices available on NCBI GEO (GSE228945)
```
# Function to combine count matrices with matrices containing MYC annotations
merge_matrix <- function(matrix1, matrix2) {
  # create a vector with 0s and add it as a new row to matrix1. This new row will hold the data for the human MYC gene
  MYC <- rep(0, times = ncol(matrix1))
  matrix1 <- rbind(matrix1, MYC)
  # get the cells present in both matrices
  cells_myc <- intersect(colnames(matrix1), colnames(matrix2))
  # replace the 0 in the MYC row with their actual values from matrix2. Only the data for cells present in matrix1 and matrix2 will be replaced.
  matrix1[c("MYC"), cells_myc] <- matrix2[c("human_c-MYC"), cells_myc]
  # return the merged matrix
  return(matrix1)
}

# Load 6 month DLP WT and HiMYC libraries
DF1 <- Read10X(data.dir = "DF1/", strip.suffix = TRUE)
DF1_MYC <- Read10X(data.dir = "DF1_myc_featurecounts/")
DF1_merge <- merge_matrix(DF1, DF1_MYC)

DF2 <- Read10X(data.dir = "/DF2/", strip.suffix = TRUE)
DF2_MYC <- Read10X(data.dir = "DF2_myc_featurecounts/")
DF2_merge <- merge_matrix(DF2, DF2_MYC)

DHMF1 <- Read10X(data.dir = "DHMF1_filtered_feature_bc_matrix/", strip.suffix = TRUE)
DHMF1_MYC <- Read10X(data.dir = "DHMF1_myc_featurecounts/")
DHMF1_merge <- merge_matrix(DHMF1, DHMF1_MYC)

DHMF2 <- Read10X(data.dir = "DHMF2_filtered_feature_bc_matrix/", strip.suffix = TRUE)
DHMF2_MYC <- Read10X(data.dir = "DHMF2_myc_featurecounts/")
DHMF2_merge <- merge_matrix(DHMF2, DHMF2_MYC)

LF1 <- Read10X(data.dir = "LF1/", strip.suffix = TRUE)
LF1_MYC <- Read10X(data.dir = "LF1_myc_featurecounts/")
LF1_merge <- merge_matrix(LF1, LF1_MYC)

LF2 <- Read10X(data.dir = "LF2/", strip.suffix = TRUE)
LF2_MYC <- Read10X(data.dir = "LF2_myc_featurecounts/")
LF2_merge <- merge_matrix(LF2, LF2_MYC)

LHMF1 <- Read10X(data.dir = "LHMF1_filtered_feature_bc_matrix/", strip.suffix = TRUE)
LHMF1_MYC <- Read10X(data.dir = "LHMF1_myc_featurecounts/")
LHMF1_merge <- merge_matrix(LHMF1, LHMF1_MYC)

LHMF2 <- Read10X(data.dir = "LHMF2_filtered_feature_bc_matrix/", strip.suffix = TRUE)
LHMF2_MYC <- Read10X(data.dir = "LHMF2_myc_featurecounts/")
LHMF2_merge <- merge_matrix(LHMF2, LHMF2_MYC)

#  Load 10 month DLP WT and HiMYC libraries
# Function to combine count matrices with matrices containing MYC annotations
merge_matrix <- function(matrix1, matrix2) {
  # create a vector with 0s and add it as a new row to matrix1. This new row will hold the data for the human MYC gene
  MYC <- rep(0, times = ncol(matrix1))
  matrix1 <- rbind(matrix1, MYC)
  # get the cells present in both matrices
  cells_myc <- intersect(colnames(matrix1), colnames(matrix2))
  # replace the 0 in the MYC row with their actual values from matrix2. Only the data for cells present in matrix1 and matrix2 will be replaced.
  matrix1[c("MYC"), cells_myc] <- matrix2[c("human_MYC"), cells_myc]
  # return the merged matrix
  return(matrix1)
}

H13 <- Read10X(data.dir = "H13DGE_filtered_feature_bc_matrix/",
               strip.suffix = TRUE)
H13_MYC <- Read10X(data.dir = "H13DGE_MYC/")
H13_merge <- merge_matrix(H13, H13_MYC)

H23 <- Read10X(data.dir = "H23DGE_filtered_feature_bc_matrix/",
               strip.suffix = TRUE)
H23_MYC <- Read10X(data.dir = "H23DGE_MYC/")
H23_merge <- merge_matrix(H23, H23_MYC)

W13 <- Read10X(data.dir = "W13DGE_filtered_feature_bc_matrix/",
               strip.suffix = TRUE)
W13_MYC <- Read10X(data.dir = "W13DGE_MYC/")
W13_merge <- merge_matrix(W13, W13_MYC)

W23 <- Read10X(data.dir = "W23DGE_filtered_feature_bc_matrix/",
               strip.suffix = TRUE)
W23_MYC <- Read10X(data.dir = "W23DGE_MYC/")
W23_merge <- merge_matrix(W23, W23_MYC)

# Merge seurat objects

```
# create seurat object and integrate
https://satijalab.org/seurat/archive/v3.0/integration.html
```
# create a list with all the matrices
matrix.list <- c(WT_D6_1 = DF1_merge,
                 WT_D6_2 = DF2_merge,
                 WT_L6_1 = LF1_merge,
                 WT_L6_2 = LF2_merge,
                 HiMYC_D6_1 = DHMF1_merge,
                 HiMYC_D6_2 = DHMF2_merge,
                 HiMYC_L6_1 = LHMF1_merge,
                 HiMYC_L6_2 = LHMF2_merge,
                  WT_B9_1 = W13_merge, 
                  WT_B9_2 = W23_merge,
                  HiMYC_B9_1 = H13_merge,
                  HiMYC_B9_2 = H23_merge)

# create empty list for Seurat objects
seurat.list <- vector(mode = "list")

# This for loop iterates through each matrix in the list, creates a Seurat object and then adds it to a new list. This new list will contain the Seurat objects
for (i in 1:length(matrix.list)) {
  seurat.object <- CreateSeuratObject(counts = matrix.list[[i]], project = names(matrix.list)[i], min.cells = 3, min.features = 200)
  seurat.list <- append(seurat.list, seurat.object)
}

# remove matrices
rm(DF1, DF1_MYC, DF1_merge,
   DF2, DF2_MYC, DF2_merge,
   LF1, LF1_MYC, LF1_merge,
   LF2, LF2_MYC, LF2_merge,
   DHMF1, DHMF1_MYC, DHMF1_merge,
   DHMF2, DHMF2_MYC, DHMF2_merge,
   LHMF1, LHMF1_MYC, LHMF1_merge,
   LHMF2, LHMF2_MYC, LHMF2_merge,
   W13, W13_MYC, W13_merge,
   W23, W23_MYC, W23_merge,
   H13, H13_MYC, H13_merge,
   H23, H23_MYC, H23_merge)
# perform sctransform on each seurat object
seurat.list <- lapply(X = seurat.list, FUN = SCTransform, method = "glmGamPoi", vst.flavor = "v2")

# integration
features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features)
seurat.list <- lapply(X = seurat.list, FUN = RunPCA, features = features)

# integrate to reference HiMYC and WT 10 month old mice DLP
levels(seurat.list[[9]])
levels(seurat.list[[11]])
anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT",
    anchor.features = features, dims = 1:30, reduction = "rpca", k.anchor = 30, reference = c(9, 11))
seurat.int <- IntegrateData(anchorset = anchors, normalization.method = "SCT", dims = 1:30)

# remove anchors, unused lists
rm(anchors, matrix.list, seurat.list)
gc()

# save integrated object
saveRDS(seurat.int, file = "FVB_DLP_int.rds")

```

add metadata
```
# Add columns for genotype
seurat.int$genotype <- "WT"
seurat.int$genotype[substring(seurat.int@meta.data$orig.ident, 1, 1) == "H"] <- "MYC"
table(seurat.int$orig.ident, seurat.int$genotype)

# Add columns for age
seurat.int$age <- "6mo"
seurat.int$age[substring(seurat.int@meta.data$orig.ident, 5, 5) == "9"] <- "10mo"
seurat.int$age[substring(seurat.int@meta.data$orig.ident, 8, 8) == "9"] <- "10mo"
table(seurat.int$orig.ident, seurat.int$age)

# Add columns for lobe
seurat.int$lobe <- "Dorsolateral"
seurat.int$lobe[substring(seurat.int@meta.data$orig.ident, 7, 7) == "D"] <- "Dorsal"
seurat.int$lobe[substring(seurat.int@meta.data$orig.ident, 7, 7) == "L"] <- "Lateral"
seurat.int$lobe[substring(seurat.int@meta.data$orig.ident, 4, 4) == "D"] <- "Dorsal"
seurat.int$lobe[substring(seurat.int@meta.data$orig.ident, 4, 4) == "L"] <- "Lateral"
table(seurat.int$orig.ident, seurat.int$lobe)

```

Filter for high mitochondrial expressing cells
```
# Add column for mitochondrial ratio
seurat.int$mitoRatio <- PercentageFeatureSet(object = seurat.int, pattern = "^mt-", assay = "RNA")
seurat.int$mitoRatio <- seurat.int@meta.data$mitoRatio / 100

# mitochondrial ratio analysis
library(stats)
VlnPlot(seurat.int, features = "mitoRatio", group.by = "orig.ident")
quantile(seurat.int$mitoRatio, probs = c(0, 0.5, 0.8, 0.85, 1))

# Filter out low quality reads using selected thresholds, > 85% of cells retained 
seurat.int <- subset(x = seurat.int,
                           (mitoRatio < 0.15))
VlnPlot(seurat.int, features = "mitoRatio", group.by = "orig.ident")

```

Make dimensionality reduction plot and cluster
```
# dimensional reduction
seurat.int <- RunPCA(seurat.int, verbose = FALSE)
seurat.int <- RunUMAP(seurat.int, reduction = "pca", dims = 1:50)
DimPlot(seurat.int, group.by = "age", shuffle = TRUE)

# clustering analysis
seurat.int <- FindNeighbors(seurat.int, reduction = "pca", dims = 1:50)
seurat.int <- FindClusters(seurat.int, resolution = 0.2)
DimPlot(seurat.int, label = T, repel = F, shuffle = TRUE, group.by = "integrated_snn_res.0.2") + ggtitle("Unsupervised clustering")

```


# SCT transform without integration
Perform SCTransform for all genes, keeping integrating clustering
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
# run sctransform
seurat.int <- SCTransform(seurat.int, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

DimPlot(seurat.int, label = T, repel = F, shuffle = TRUE, group.by = "integrated_snn_res.0.2") + ggtitle("Unsupervised clustering")
```

# cell type ID
Marker genes for each cell type
```{r}

# load seurat object
seurat.int <- readRDS(file = "FVB_DLP_int.rds")

Basal <- c("Krt15","Krt5", "Krt14")
Epithelial <- c("Krt8", "Krt18")
Luminal <- "Agr2"
Foxi1 <- c("Foxi1")
HiMYC_luminal <- c("Timp1")
Luminal_Ly6d <- c("Ly6d", "Psca")
Neuroendocrine <- c("Chga", "Chgb", "Syp")
sm_pericytes <- c("Tagln", "Acta2")
Smooth_Muscle <- c("Acta1", "Actg2")
Pericytes <- c("Rgs4", "Notch3")
Fibroblast <- c("Pdgfra", "Fbln1")
Endothelial <- c("Eng","Flt1","Cdh5")
Macrophages <- c("C1qb","C1qa","C1qc","Cd68")
Immune <- "Ptprc"
Mast_cells <- c("Mcemp1", "Flt3")
Tcells <- c("Cd2","Cd3d", "Cd3g")
Glial <- c("Plp1", "Ngfr", "Mbp")

cell_markers <- c(Epithelial, Basal, Luminal_Ly6d, Luminal, Foxi1, HiMYC_luminal,  
                  Endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast,Glial,
                  Immune,Macrophages,  Mast_cells, Tcells
                  )
```

update cluster name by cell type
```
Idents(seurat.int) <- "integrated_snn_res.0.2"
DoHeatmap(subset(seurat.int, downsample = 100), assay = "SCT", features = c(cell_markers)) +
  scale_fill_viridis()

table(seurat.int$integrated_snn_res.0.2, seurat.int$lobe)

# cell type ID
Idents(seurat.int) <- "integrated_snn_res.0.2"
new.cluster.ids <- c("Basal",
                     "Luminal MYC 1", "Luminal Psca", "Luminal", "Luminal", "Fibroblasts",
                     "Macrophages", "Endothelial", "Luminal MYC 1", "Luminal MYC 1", "T cells",
                     "Pericytes", "Luminal", "Smooth Muscle", "Mast cells", "Ionocytes",
                     "Glial"
                     )

names(new.cluster.ids) <- levels(seurat.int)
seurat.int[["cell_type"]] <- new.cluster.ids[as.numeric(as.matrix(seurat.int@meta.data$integrated_snn_res.0.2))+1]
seurat.int <- RenameIdents(seurat.int, new.cluster.ids)
```

```{r}
Idents(seurat.int) <- "cell_type"
levels(seurat.int) <- c("Basal", "Luminal Psca", "Luminal", "Ionocytes", "Luminal MYC 1", 
                        "Endothelial", "Smooth Muscle", "Pericytes", "Fibroblasts", "Glial",
                        "Macrophages",  "Mast cells", "T cells")     
                     

DoHeatmap(subset(seurat.int, downsample = 50), assay = "SCT", features = c(cell_markers)) +
  scale_fill_viridis()
```
UMAP plots
```{r}
DimPlot(seurat.int, reduction = "umap", label = TRUE, pt.size = 1, shuffle = TRUE, raster = TRUE) + NoLegend() + ggtitle("Cell Type")

# age and genotype
seurat.int$genotype <- factor(seurat.int$genotype, levels = c('WT', 'MYC'))

pal1 <- c("#219ebc", "#fb8500")
pal2 <- c("#a8dadc", "#e63946")

DimPlot(seurat.int, reduction = "umap", label = FALSE, pt.size = 1, group.by = "age", shuffle = TRUE, cols = pal1, raster = TRUE) +  ggtitle("Age")

DimPlot(seurat.int, reduction = "umap", label = FALSE, pt.size = 1, group.by = "genotype", shuffle = TRUE, cols = pal2, raster = TRUE) +  ggtitle("Genotype")

# saveRDS(seurat.int, file = "FVB_DLP_int.rds")
```
# GSEA luminal HiMYC 6 months Luminal vs Luminal MYC 1
```{r}
library(DataCombine)
library(msigdbr)
library(presto)
library(fgsea)

# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "mouse", category = "H")
fgsea_sets<- hallmark_gene_sets %>% split(x = .$gene_symbol, f = .$gs_name)
```
6 month WT and HiMYC Luminal MYC 1 vs WT: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Idents(seurat.int) <- "age"
levels(seurat.int)

Luminal6mo_genes <- subset(seurat.int, idents = c("6mo")) %>% wilcoxauc(group_by = 'cell_type', seurat_assay = 'SCT', assay = 'data', groups_use = c("Luminal", "Luminal MYC 1"))
head(Luminal6mo_genes)

# saveRDS(Luminal6mo_genes, file = "GSEALuminal6mo_genes.rds")

# we have all the genes for each cluster
dplyr::count(Luminal6mo_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Luminal6mo_genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Luminal6mo_genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Luminal6mo_genes_group<- Luminal6mo_genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Luminal6mo<- deframe(Luminal6mo_genes_group)
head(ranks_Luminal6mo)
tail(ranks_Luminal6mo)

# perform analysis
fgseaRes_Luminal6mo <- fgsea(fgsea_sets, stats = ranks_Luminal6mo)

# save results
# saveRDS(fgseaRes_Luminal6mo, file = "fgseaRes_Luminal6mo.rds")

# tidy the data
fgseaResTidy_Luminal6mo <- fgseaRes_Luminal6mo %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_Luminal6mo$statistic <- -log10(fgseaResTidy_Luminal6mo$padj)

# add comparison
fgseaResTidy_Luminal6mo$comparison <- "Luminal MYC 1 vs Luminal in 6 month Mice"

```
# GSEA luminal HiMYC 6 months vs 10 months
Luminal Hi-Myc 6 months vs 10 months: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
seurat.int$age_genotype_celltype <- paste(seurat.int$age, seurat.int$genotype, seurat.int$cell_type, sep = " ")
Idents(seurat.int) <- "age_genotype_celltype"
levels(seurat.int)

Luminal_genes <- wilcoxauc(X = seurat.int, group_by = 'age_genotype_celltype', seurat_assay = 'SCT', assay = 'data', groups_use = c("6mo MYC Luminal MYC 1", "10mo MYC Luminal MYC 1"))
head(Luminal_genes)

# saveRDS(Luminal_genes, file = "GSEALuminal_genes.rds")

# we have all the genes for each cluster
dplyr::count(Luminal_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Luminal_genes %>%
  dplyr::filter(group == "10mo MYC Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Luminal_genes %>%
  dplyr::filter(group == "10mo MYC Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Luminal_genes_group<- Luminal_genes %>%
  dplyr::filter(group == "10mo MYC Luminal MYC 1") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Luminal<- deframe(Luminal_genes_group)
head(ranks_Luminal)
tail(ranks_Luminal)

# perform analysis
fgseaRes_Luminal <- fgsea(fgsea_sets, stats = ranks_Luminal)

# save results
# saveRDS(fgseaRes_Luminal, file = "fgseaRes_Luminal.rds")

# tidy the data
fgseaResTidy_Luminal <- fgseaRes_Luminal %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_Luminal$statistic <- -log10(fgseaResTidy_Luminal$padj)

# add comparison
fgseaResTidy_Luminal$comparison <- "Luminal MYC 1 10 months vs 6 months"

```
Plot fgsea results from two different comparison
```{r}
library(forcats)

fgseaResTidy_Luminal6mo$sort <- fgseaResTidy_Luminal6mo$NES
fgseaResTidy_Luminal$sort <- NA

# combine fgsea dataframes
fgsea_combineDF <- rbind(fgseaResTidy_Luminal6mo, fgseaResTidy_Luminal)
fgsea_combineDF <- fgsea_combineDF %>% group_by(pathway) %>% 
  mutate(NES_diff=NES-lag(NES))

# sort
fgsea_combineDF <- fgsea_combineDF %>% arrange(sort)

# top changes
fgsea_diff_top <- fgsea_combineDF %>% arrange(desc(abs(NES_diff))) %>% head(13) %>% arrange(sort)
fgsea_diff_top

## only plot the top pathway differences
pal3 <- c("#0081a7", "#f07167")

ggplot(fgsea_combineDF %>% filter(pathway %in% fgsea_diff_top$pathway),
       aes(x = fct_inorder(pathway), NES)) +
  geom_col(aes(fill= comparison),width=.5, position = "dodge") +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top Pathway Changes during MYC driven cancer progression") + 
  theme_classic() + labs(fill =  expression("Cluster Comparison")) +
  scale_fill_manual(values=pal3)

```

Top 20 leading edge genes
```{r}
# P53 leading edge
leadingEdge_P53_late <- fgseaRes_Luminal %>% filter(pathway == "HALLMARK_P53_PATHWAY") %>% select(leadingEdge) %>% unlist() %>% as.data.frame() %>% head(20)

# DNA repair leading edge
leadingEdge_DNA_late <- fgseaRes_Luminal %>% filter(pathway == "HALLMARK_DNA_REPAIR") %>% select(leadingEdge) %>% unlist() %>% as.data.frame() %>% head(20)

# Inflammatory leading edge
leadingEdge_inflame_late <- fgseaRes_Luminal %>% filter(pathway == "HALLMARK_INFLAMMATORY_RESPONSE") %>% select(leadingEdge) %>% unlist() %>% as.data.frame() %>% head(20)

# IL6 leading edge
leadingEdge_IL6_late <- fgseaRes_Luminal %>% filter(pathway == "HALLMARK_IL6_JAK_STAT3_SIGNALING") %>% select(leadingEdge) %>% unlist() %>% as.data.frame() %>% head(20)

```

heatmap of key pathways
```{r}
# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(wesanderson)

# expression matrix
Idents(seurat.int) <- "cell_type"
levels(seurat.int)
luminalMYC1_seurat <- subset(seurat.int, idents = c("Luminal MYC 1", "Luminal"))

luminalMYC1_seurat$age_celltype <- paste(luminalMYC1_seurat$age, luminalMYC1_seurat$cell_type, sep = " ")

Idents(luminalMYC1_seurat) <- "age_celltype"
levels(luminalMYC1_seurat)
AvgExp <- AverageExpression(luminalMYC1_seurat)
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# filter expression heatmap to leading edge genes
AvgExp_p53 <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_P53_late$.)) %>% arrange(`10mo Luminal MYC 1`) %>% as.matrix()
p53_gene_order <- rownames(AvgExp_p53)

AvgExp_DNA <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_DNA_late$.)) %>% arrange(`10mo Luminal MYC 1`) %>% as.matrix()
DNA_gene_order <- rownames(AvgExp_DNA)

AvgExp_inflame <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_inflame_late$.)) %>% arrange(`10mo Luminal MYC 1`) %>% as.matrix()
inflame_gene_order <- rownames(AvgExp_inflame)

AvgExp_IL6 <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_IL6_late$.)) %>% arrange(`10mo Luminal MYC 1`) %>% as.matrix()
IL6_gene_order <- rownames(AvgExp_IL6)

# sample order
sample_order <- c("6mo Luminal","10mo Luminal", "6mo Luminal MYC 1", "10mo Luminal MYC 1")

# sort by z score of Luminal MYC 1 9 month for column order, concatenate for top 20 of each leading edge pathway
pal <- wes_palette("Zissou1", 5, type = "discrete")

# complex heatmap
P53 <- Heatmap(AvgExp_p53,
               name = "P53",
        column_order = sample_order,
        row_order = p53_gene_order,
        row_dend_reorder = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = "right",
        row_title = "P53 PATHWAY",
        row_title_gp = gpar(fill = pal[1], col = "white", border = NA, fontface = "bold"),
        row_names_gp = gpar(fontsize = 12),
        column_names_rot = 45,
        heatmap_legend_param = list(
          title = "Z-score",
          legend_height = unit(4, "cm")
        ))

P53

DNA <- Heatmap(AvgExp_DNA,
               column_order = sample_order,
               row_order = DNA_gene_order,
               name = "DNA",
               column_dend_reorder = FALSE,
               show_row_dend = FALSE,
               show_column_dend = FALSE,
               row_names_side = "right",
               row_title = "DNA REPAIR",
               row_title_gp = gpar(fill = pal[2], col = "white", border = NA, fontface = "bold"),
               row_names_gp = gpar(fontsize = 12),
               column_names_rot = 45,
               heatmap_legend_param = list(
                 title = "Z-score",
                 legend_height = unit(4, "cm")
               ))

DNA

Inflame <- Heatmap(AvgExp_inflame,
               name = "Inflame",
               column_order = sample_order,
               row_order = inflame_gene_order,
               row_dend_reorder = FALSE,
               show_row_dend = FALSE,
               show_column_dend = FALSE,
               row_names_side = "right",
               row_title = "INFLAMMATORY RESPONSE",
               row_title_gp = gpar(fill = pal[3], col = "white", border = NA, fontface = "bold"),
               row_names_gp = gpar(fontsize = 12),
               column_names_rot = 45,
               heatmap_legend_param = list(
                 title = "Z-score",
                 legend_height = unit(4, "cm")
               ))

Inflame


IL6 <- Heatmap(AvgExp_IL6,
        column_order = sample_order,
        row_order = IL6_gene_order,
        name = "IL6",
        column_dend_reorder = FALSE,
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = "right",
        row_title = "IL6 JAK STAT3 SIGNALING",
        row_title_gp = gpar(fill = pal[4], col = "white", border = NA, fontface = "bold"),
        row_names_gp = gpar(fontsize = 12),
        column_names_rot = 45,
        heatmap_legend_param = list(
          title = "Z-score",
          legend_height = unit(4, "cm")
        ))

IL6
```
# save seurat object
```
saveRDS(seurat.int, file = "FVB_DLP_int.rds")
```