---
title: "MYC mice prostate"
author: "Mindy Kim Graham"
date: "6/14/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# single cell analysis of mouse prostate (WT and Hi-MYC) at 6 months and 10 months subsetted to the immune population
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(sctransform)
library(glmGamPoi)
library(dplyr)
library(DataCombine)
library(viridis)
library(tidyverse)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

```

Load integrated seurat object
```
seurat.int <- readRDS(file = "FVB_DLP_int.rds")

# subset to immune cells
Idents(seurat.int) <- "cell_type"
levels(seurat.int)

immune.int <- subset(seurat.int, idents = c("Macrophages", "Mast cells", "T cells"))
DimPlot(immune.int)
```

perform dimensionality reduction and clustering on subset
```
# run sctransform
immune.int <- SCTransform(immune.int, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE)

# dimensionality reduction
immune.int <- RunPCA(immune.int, verbose = FALSE, npcs = 50)
immune.int <- RunUMAP(immune.int, reduction = "pca", dims = 1:30)
immune.int <- FindNeighbors(immune.int, dims = 1:50, verbose = FALSE)

# cluster resolution
immune.int <- FindClusters(immune.int, verbose = FALSE, resolution = 0.1)
```

immune cell type markers
```{r}
# subtype markers
Macrophages <- c("C1qb","C1qa","C1qc","Cd68", "Adgre1")
Immune <- "Ptprc"
Mast_cells <- c("Mcemp1", "Flt3", "Kit")
Tcells <- c("Cd2","Cd3d", "Cd3g", "Trbc2")
TCD4 <- c("Cd4", "Cd8a")
TCD8 <- c("Cd8a")
Treg <- c("Foxp3", "Il2ra")
cytotoxic <- c("Prf1", "Gzmb")
killer_cells <- c("Klrb1b", "Klrd1", "Nkg7", "Ncr1")
Tmem_central <- c("Cd44", "Il7r", "Sell")
Tactive <- c("Cd69") # or Il2ra = CD25
Myeloid <- "Itgam"
Bcell <- c("Cd79a", "Cd79b")
Bcell_naive <- c("Ighd")
Bcell_memory <- c("Cd27")
M1 <- c("Cd86", "Cd80", "Nos2")
M2 <- c("Cd163", "Mrc1", "Arg1", "Arg2")
# plasmacytoid dendritic cells (pDCs)
pDCs <- c("Siglech", "Bst2")
neutrophils <- c("Ly6g")
Effector <- c("Cd44")
Monocyte <- "Cd14"
MDSCs <- c("S100a8", "S100a9")

# gene_features <-  c(Tcells,  Tactive, TCD4, Treg, TCD8, killer_cells, Myeloid, Bcell, Macrophages, "Irg1", "Trem2", pDCs, Mast_cells, MDSCs, Bcell)

# DoHeatmap(subset(immune.int, downsample = 50), features = c(gene_features), group.by = "SCT_snn_res.0.1") + scale_fill_viridis()
```

Differential gene expression of clusters
```
# differentially expressed genes for each cluster
levels(immune.int)
immune_clusters <- FindAllMarkers(immune.int)
top_immune <- immune_clusters %>% group_by(cluster) %>% filter(pct.2 < 0.2, pct.1 > 0.5) %>% top_n(5, -p_val_adj) %>% top_n(5, avg_log2FC)
#top_immune %>% filter(cluster=="5")

DoHeatmap(subset(immune.int, downsample = 50), features = c(top_immune$gene), group.by = "SCT_snn_res.0.1") +
  scale_fill_viridis()
```

name clusters
0 - Macrophages Trem2
1 - T cells CD8
2 - Macrophages Trem2
3 - Macrophages Trem2 Irg1/Acod
4 - NK cells
5 - T cells reg
6 - Macrophages Trem2
7 - Mast cells
8 - MDSCs
9 - B cells
10 - pDCs
```
# immune cells ID
Idents(immune.int) <- "SCT_snn_res.0.1"
new.cluster.ids <- c("Macrophages Trem2", 
                     "T cells CD8", "Macrophages Trem2", "Macrophages Trem2", "NK cells", "T cells reg", 
                     "Macrophages Trem2", "Mast cells", "MDSCs", "B cells", "Plasmacytoid Dendritic cells")

names(new.cluster.ids) <- levels(immune.int)
immune.int[["immune"]] <- new.cluster.ids[as.numeric(as.matrix(immune.int@meta.data$SCT_snn_res.0.1))+1]
immune.int <- RenameIdents(immune.int, new.cluster.ids)

saveRDS(immune.int, file = "seurat_immune.rds")
```

```{r}
# load seurat immune subset
immune.int <- readRDS(file = "seurat_immune.rds")

Idents(immune.int) <- "immune"
levels(immune.int) <- c("T cells CD8", "T cells reg", "NK cells", "B cells",
                        "Macrophages Trem2",
                        "MDSCs", "Mast cells", "Plasmacytoid Dendritic cells")

DimPlot(immune.int, label = TRUE, raster = TRUE, pt.size = 2) + NoLegend()

gene_features <-  c(Tcells,TCD8, TCD4, Treg,  killer_cells, Bcell, Macrophages, "Trem2", "Irg1",  Myeloid, MDSCs, Mast_cells, pDCs)
DoHeatmap(subset(immune.int, downsample = 50), features = gene_features) +
  scale_fill_viridis()
```

```
# immune subtype ID
Idents(immune.int) <- "SCT_snn_res.0.1"
new.cluster.ids <- c("Macrophages Trem2", 
                     "T cells CD8", "Macrophages Trem2", "Macrophages Trem2 Irg1", "NK cells", "T cells reg", 
                     "Macrophages Trem2", "Mast cells", "MDSCs", "B cells", "Plasmacytoid Dendritic cells")

names(new.cluster.ids) <- levels(immune.int)
immune.int[["immune_subtype"]] <- new.cluster.ids[as.numeric(as.matrix(immune.int@meta.data$SCT_snn_res.0.1))+1]
immune.int <- RenameIdents(immune.int, new.cluster.ids)
```

```{r}
Idents(immune.int) <- "immune_subtype"
levels(immune.int) <- c("T cells CD8", "T cells reg", "NK cells", "B cells",
                        "Macrophages Trem2", "Macrophages Trem2 Irg1",
                        "MDSCs", "Mast cells", "Plasmacytoid Dendritic cells")

levels(immune.int)

DimPlot(immune.int, label = TRUE, raster = TRUE, label.size = 5, pt.size = 2, repel = TRUE) +
  NoLegend() + ggtitle("Mouse Immune Clusters") +
  theme(plot.title = element_text(hjust = 0.5))

DimPlot(immune.int, label = FALSE, raster = TRUE, split.by = "genotype", pt.size = 2) + 
  ggtitle("Mouse Immune Clusters Split by Genotype") +
  theme(plot.title = element_text(hjust = 0.5)) +
  NoAxes()

gene_features <-  c(Tcells,TCD8, TCD4, Treg,  killer_cells, Bcell, Macrophages, "Trem2", "Irg1",  Myeloid, MDSCs, Mast_cells, pDCs)
DoHeatmap(subset(immune.int, downsample = 50), features = gene_features) +
  scale_fill_viridis()
```

# Cell proportions by RAISIN
Update immune cell types in seurat.int
```
# immune clusters
Idents(seurat.int) <- "cell_type"
DimPlot(seurat.int) 

seurat.int$immune <- seurat.int$cell_type
seurat.int$immune <- gsub("Mast cells", "", seurat.int$immune)
seurat.int$immune <- gsub("Macrophages", "", seurat.int$immune)
seurat.int$immune <- gsub("T cells", "", seurat.int$immune)

table(seurat.int$immune)

# add immune subset data to main seurat object
seurat.int$immune_subset <- immune.int$immune
seurat.int$immune <- paste0(seurat.int$immune_subset, seurat.int$immune)
table(seurat.int$immune)
seurat.int$immune <- gsub("NA", "", seurat.int$immune)

table(seurat.int$immune)
```

RAISIN
https://github.com/zji90/raisin//
cell proportion 10 months
```{r}
library(raisin)

# load seurat object
seurat.int <- readRDS(file = "FVB_DLP_int.rds")

# subset by MYC
Idents(seurat.int) <- "genotype"
levels(seurat.int)
seurat.int.MYC <- subset(seurat.int, idents = c("MYC"))

# extract metadata from seurat object
Idents(seurat.int.MYC) <- "immune"
levels(seurat.int.MYC)
cluster <- Idents(seurat.int.MYC)

Idents(seurat.int.MYC) <- "orig.ident"
levels(seurat.int.MYC)
individual <- Idents(seurat.int.MYC)

# make the design 
design <- data.frame(intercept=1,contrast=c(0,0,0,0,
                                            1,1),
                     row.names = c("HiMYC_D6_1", "HiMYC_D6_2", "HiMYC_L6_1", "HiMYC_L6_2", 
                                   "HiMYC_B9_1", "HiMYC_B9_2"))

# run the test
prop_test_results <- proportiontest(cluster,individual,design)
prop_test_results <- prop_test_results %>% tibble::rownames_to_column("Cluster")
prop_test_results
```

Plot cell proportion diferences of immune cells
```{r}
library(ggpubr)
table(seurat.int$immune)
library_counts <- as.data.frame(table(seurat.int$orig.ident))

# create dataframe containing all relevant metadata
cell_type_enrich <- as.data.frame(table(seurat.int$orig.ident,seurat.int$genotype, seurat.int$age, seurat.int$immune))

cell_type_enrich$library <- library_counts$Freq
cell_type_enrich$fraction <- cell_type_enrich$Freq / cell_type_enrich$library

# rename columns to meaningful column names
colnames(cell_type_enrich) <- c("Library", "Genotype", "Age", "Cluster", "Counts", "Library_Counts", "Fraction")

# filter to library
Idents(seurat.int) <- "orig.ident"
levels(seurat.int)
WT_D6_1 <- cell_type_enrich %>% filter(Library == "WT_D6_1", Genotype == "WT", Age == "6mo")
WT_D6_2 <- cell_type_enrich %>% filter(Library == "WT_D6_2", Genotype == "WT", Age == "6mo")
WT_L6_1 <- cell_type_enrich %>% filter(Library == "WT_L6_1", Genotype == "WT", Age == "6mo")
WT_L6_2 <- cell_type_enrich %>% filter(Library == "WT_L6_2", Genotype == "WT", Age == "6mo")

HiMYC_D6_1 <- cell_type_enrich %>% filter(Library == "HiMYC_D6_1", Genotype == "MYC", Age == "6mo")
HiMYC_D6_2 <- cell_type_enrich %>% filter(Library == "HiMYC_D6_2", Genotype == "MYC", Age == "6mo")
HiMYC_L6_1 <- cell_type_enrich %>% filter(Library == "HiMYC_L6_1", Genotype == "MYC", Age == "6mo")
HiMYC_L6_2 <- cell_type_enrich %>% filter(Library == "HiMYC_L6_2", Genotype == "MYC", Age == "6mo")

WT_B9_1 <- cell_type_enrich %>% filter(Library == "WT_B9_1", Genotype == "WT", Age == "10mo")
WT_B9_2 <- cell_type_enrich %>% filter(Library == "WT_B9_2", Genotype == "WT", Age == "10mo")
HiMYC_B9_1 <- cell_type_enrich %>% filter(Library == "HiMYC_B9_1", Genotype == "MYC", Age == "10mo")
HiMYC_B9_2 <- cell_type_enrich %>% filter(Library == "HiMYC_B9_2", Genotype == "MYC", Age == "10mo")

filtered_library_fraction <- rbind(WT_D6_1, WT_D6_2, WT_L6_1, WT_L6_2,
                                   HiMYC_D6_1, HiMYC_D6_2, HiMYC_L6_1, HiMYC_L6_2,
                                   WT_B9_1, WT_B9_2, HiMYC_B9_1, HiMYC_B9_2)

# combine DF with cell proportion test results
combined_df <- merge(filtered_library_fraction, prop_test_results, by.x = "Cluster", by.y = "Cluster")
combined_df$`Genotype Age` <- paste(combined_df$Genotype, combined_df$Age, sep = " ")

# Add pvalue to cluster name
library(scales)
combined_df$cluster_adjP <- paste(combined_df$Cluster,
                                  "\n Hi-Myc 10 months vs 6 months",
                                  "\n logFC = ", 
                                  signif(combined_df$logFC, 2),
                                  ", adj p-value = ", scientific(combined_df$adj.P.Val, 1))


# update level names
combined_df$`Genotype Age` <- gsub(" ", "\n", combined_df$`Genotype Age`)

# order x axis
combined_df$`Genotype Age` <- factor(combined_df$`Genotype Age`, levels = c("WT\n6mo", "WT\n10mo", "MYC\n6mo", "MYC\n10mo"))
levels(combined_df$`Genotype Age`)

# save dataframe
# saveRDS(combined_df, file = "immune_combined_df.rds")

# Plot all immune cells
combined_df %>% 
  filter(Cluster %in% levels(immune.int)) %>%
  ggplot(aes(x = `Genotype Age`, y = Fraction))+
  geom_jitter(aes(fill = Genotype, color = `Genotype Age`), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Immune Clusters: 6 Month and 10 month Mice") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~cluster_adjP, scales = "free") +
  scale_color_manual(values = c("#219ebc", "#023047", "#ffb703", "#fb8500")) +
  NoLegend()

# Plot immunosuppressive populations
combined_df %>% 
  filter(Cluster %in% c("Macrophages Trem2", "MDSCs", "T cells reg" )) %>%
  ggplot(aes(x = `Genotype Age`, y = Fraction))+
  geom_jitter(aes(fill = Genotype, color = `Genotype Age`), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Immune Cell Proportion") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~cluster_adjP, scales = "free") +
  scale_color_manual(values = c("#219ebc", "#023047", "#ffb703", "#fb8500")) +
  NoLegend()

```


Differential gene expression by seurat
```
markers_all_immunesubtype <- FindAllMarkers(seurat.int)
saveRDS(markers_all_immunesubtype, file = "markers_all_immunesubtype.rds")
write.csv(markers_all_immunesubtype, file = "markers_all_immunesubtype.csv")
```

Save immune seurat object
```
saveRDS(immune.int, file = "seurat_immune.rds")
```