---
title: "BMPC mice prostate"
author: "Mindy Kim Graham"
date: "2/05/2024"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# single cell analysis of mouse prostate (WT, Hi-Myc, BMPC)
Here we merge libraries from the FVB mice aged 6 months and 10 months in WT and HiMYC mouse strains and 12 week (AP, DP, LP, VP) and 6 month (lymph node met) BMPC.

At 6 months, HiMYC FVB mice predominately have PIN in DLV, while at 10 months DL lobes have significant invasive carcinoma.
At 3 months, BMPC mice predominately have PIN, predominately in AP, while at 6 months, metastasis is visible
https://aacrjournals.org/cancerres/article/76/2/283/613869/Combined-MYC-Activation-and-Pten-Loss-Are

Here we use a reference based integration 
https://satijalab.org/seurat/archive/v3.0/integration.html
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(sctransform)
library(glmGamPoi)
library(dplyr)
library(DataCombine)
library(viridis)
library(tidyverse)
library(patchwork)
library(wesanderson)

pal4 <- wes_palette("Zissou1", 5, "discrete")
pal5 <- wes_palette("Zissou1", 5, "continuous")
pal6 <- wes_palette("Zissou1", 6, "continuous")

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

```

# Load BMPC mouse data
```
BMPC_AP <- Read10X(data.dir = "BMPC outs file/lobe/AP/filtered_feature_bc_matrix/")
BMPC_AP <- CreateSeuratObject(counts = BMPC_AP, project = "BMPC AP", min.cells = 3, min.features = 200)
BMPC_AP$lobe <- "Anterior"
BMPC_AP$age <- "3mo"
BMPC_AP$strain <- "FVB/N"
BMPC_AP$genotype <- "BMPC"

BMPC_DP <- Read10X(data.dir = "BMPC outs file/lobe/DP/filtered_feature_bc_matrix/")
BMPC_DP <- CreateSeuratObject(counts = BMPC_DP, project = "BMPC DP", min.cells = 3, min.features = 200)
BMPC_DP$lobe <- "Dorsal"
BMPC_DP$age <- "3mo"
BMPC_DP$strain <- "FVB/N"
BMPC_DP$genotype <- "BMPC"

BMPC_LP <- Read10X(data.dir = "BMPC outs file/lobe/LP/filtered_feature_bc_matrix/")
BMPC_LP <- CreateSeuratObject(counts = BMPC_LP, project = "BMPC LP", min.cells = 3, min.features = 200)
BMPC_LP$lobe <- "Lateral"
BMPC_LP$age <- "3mo"
BMPC_LP$strain <- "FVB/N"
BMPC_LP$genotype <- "BMPC"

BMPC_VP <- Read10X(data.dir = "BMPC outs file/lobe/VP/filtered_feature_bc_matrix/")
BMPC_VP <- CreateSeuratObject(counts = BMPC_VP, project = "BMPC VP", min.cells = 3, min.features = 200)
BMPC_VP$lobe <- "Ventral"
BMPC_VP$age <- "3mo"
BMPC_VP$strain <- "FVB/N"
BMPC_VP$genotype <- "BMPC"

BMPC_met <- Read10X(data.dir = "BMPC outs file/MET2/filtered_feature_bc_matrix/")
BMPC_met <- CreateSeuratObject(counts = BMPC_met, project = "BMPC Met", min.cells = 3, min.features = 200)
BMPC_met$lobe <- "Lymph Node Met"
BMPC_met$age <- "6mo"
BMPC_met$strain <- "FVB/N"
BMPC_met$genotype <- "BMPC"
```

# Load seurat object for Hi-Myc FVB 6 month and 10 month
```
# 6 month seurat object
strain.combined.sct <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")
Idents(strain.combined.sct) <- "orig.ident"
levels(strain.combined.sct)

DimPlot(strain.combined.sct)

# subset to FVB dataset
Idents(strain.combined.sct) <- "strain"
levels(strain.combined.sct)
seurat_FVB_6mo <- subset(strain.combined.sct, idents = "FVB/NJ")
DimPlot(seurat_FVB_6mo)
rm(strain.combined.sct)

# load 10 month seurat object data
seurat_combined <- readRDS(file = "FVB_DLP_int.rds")
DimPlot(seurat_combined)
Idents(seurat_combined) <- "age"
levels(seurat_combined)

seurat_FVB_10mo <- subset(seurat_combined, idents = "10mo")
DimPlot(seurat_FVB_10mo)
rm(seurat_combined)
```

# merge seurat objects
```
seurat_combined <- merge(seurat_FVB_6mo, y = c(seurat_FVB_10mo, BMPC_AP, BMPC_DP, BMPC_LP, BMPC_VP, BMPC_met), project = "MYC models")

# remove uncombined seurat objects, and gc()
rm(seurat_FVB_6mo, seurat_FVB_10mo, BMPC_AP, BMPC_DP, BMPC_LP, BMPC_VP, BMPC_met)
gc()

# metadata
seurat_combined$orig.ident <- paste(seurat_combined$orig.ident, seurat_combined$library_id)
table(seurat_combined$orig.ident)
seurat_combined$orig.ident <- gsub(" NA", "", seurat_combined$orig.ident)
seurat_combined$orig.ident <- gsub("seurat_object ", "", seurat_combined$orig.ident)

# update age for 6mo Hi-Myc and WT animals
seurat_combined$age <- seurat_combined$age %>% replace_na("6mo")

```

# QC metrics
```
DefaultAssay(seurat_combined) <- "RNA"

# Add number of genes per UMI for each cell to metadata
seurat_combined$log10GenesPerUMI <- log10(seurat_combined$nFeature_RNA) / log10(seurat_combined$nCount_RNA)

# Compute percent mito ratio
seurat_combined$mitoRatio <- PercentageFeatureSet(object = seurat_combined, pattern = "^mt-")
seurat_combined$mitoRatio <- seurat_combined@meta.data$mitoRatio / 100

summary(seurat_combined$mitoRatio)

# mito ratio quantiles
library(stats)
quantile(seurat_combined$mitoRatio, probs = c(0, 0.5, 0.8, 0.94, 1))

# Filter out low quality reads using selected thresholds, > 94% of cells retained 
seurat_combined <- subset(x = seurat_combined,
                           (mitoRatio < 0.25))

summary(seurat_combined$mitoRatio)
```

# sctransformation, dimensionality reduction
```
# sctransformation
seurat_combined <- seurat_combined %>%
  SCTransform(vars.to.regress = c("mitoRatio"), 
              method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

# dimensionality reduction
seurat_combined <- RunPCA(seurat_combined, verbose = FALSE)
seurat_combined <- RunUMAP(seurat_combined, reduction = "pca", dims = 1:40)

UMAP_genotype <- DimPlot(seurat_combined, group.by = "genotype", shuffle = TRUE, cols = pal5[c(5, 3, 1)],
                         raster = TRUE) + NoAxes() + ggtitle("Mouse Model")
UMAP_genotype

UMAP_age <- DimPlot(seurat_combined, group.by = "age", shuffle = TRUE, raster = TRUE) + NoAxes() + ggtitle("Mouse Age")
UMAP_age

UMAP_library <- DimPlot(seurat_combined, group.by = "orig.ident", shuffle = TRUE, raster = TRUE) + NoAxes() + ggtitle("Library")
UMAP_library

UMAP_celltype <- DimPlot(seurat_combined, group.by = "immune", shuffle = TRUE, label = TRUE) + ggtitle("Cell Types")
UMAP_celltype

UMAP_genotype + UMAP_age + UMAP_celltype
# Myeloid, Lymphoid, and Stromal clusters contain cells from all libraries FVB WT (6mo, 10 mo), FVB Hi-Myc (6 mo, 10 mo), FVB BMPC (3 mo, 6 mo), suggesting batch effects are not significant. Integration unnecessary, but will complete after cell type annotation as luminal cells have significant subclustering that make epithelial subtype annotation complex
```
# cluster analysis
```
seurat_combined <- FindNeighbors(seurat_combined, dims = 1:50, verbose = FALSE)
seurat_combined <- FindClusters(seurat_combined, resolution = c(0.1, 0.2, 0.3, 0.4, 0.5))
seurat_combined <- FindClusters(seurat_combined, resolution = c(0.6))

UMAP_01 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.1", label = TRUE)  + NoLegend()
UMAP_02 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.2", label = TRUE)  + NoLegend()
UMAP_03 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.3", label = TRUE)  + NoLegend()
UMAP_04 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.4", label = TRUE)  + NoLegend()
UMAP_05 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.5", label = TRUE)  + NoLegend()
UMAP_06 <- DimPlot(seurat_combined, group.by = "SCT_snn_res.0.6", label = TRUE)  + NoLegend()

UMAP_01 + UMAP_02 + UMAP_03 + UMAP_04
UMAP_04 + UMAP_06

```

# cell type ID
Marker genes for each cell type
```{r}
Basal <- c("Krt5", "Krt14")
Epithelial <- c("Krt8", "Krt18")
Luminal <- "Agr2"
Foxi1 <- c("Foxi1")
HiMYC_luminal <- c("Timp1", "Wfdc12")
Luminal_Psca <- c("Psca")
Neuroendocrine <- c("Chga", "Syp")
sm_pericytes <- c("Tagln", "Acta2")
Smooth_Muscle <- c("Actg2") # Acta1
Pericytes <- c("Rgs4", "Notch3")
Fibroblast <- c("Pdgfra", "Fbln1", "Rorb", "Sult1e1")
Endothelial <- c("Eng","Flt1","Cdh5")
Macrophages <- c("C1qb","C1qa","C1qc","Cd68")
Immune <- "Ptprc"
Mast_cells <- c("Mcemp1", "Flt3")
Tcells <- c("Cd2","Cd3d", "Cd3g")
Bcells <- c("Cd79a", "Cd79b")
Plasma <- c("Tnfrsf17")
pDC <- c("Siglech")
MDSC <- c("Itgam", "S100a8","S100a9")
Glial <- c("Plp1", "Ngfr")

cell_markers <- c(Epithelial, Luminal, Luminal_Psca, Foxi1, Basal, 
                  Endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast,
                  Immune,Macrophages, MDSC,  Mast_cells, Tcells, Bcells, Plasma, pDC, Glial
                  )
```

```
Idents(seurat_combined) <- "SCT_snn_res.0.6"
heatmap_06 <- DoHeatmap(subset(seurat_combined, downsample = 100), assay = "SCT", features = c(cell_markers)) +
  scale_fill_viridis()

UMAP_06 + heatmap_06
```

update cluster names
```
# seurat clusters
seurat_clusters_06 <- FindAllMarkers(seurat_combined)
saveRDS(seurat_clusters_06, file = "seurat_clusters_06.rds")

# cell type ID
Idents(seurat_combined) <- "SCT_snn_res.0.6"
new.cluster.ids <- c("Luminal",
                     "Macrophages", "Basal", "Luminal", "Basal", "Fibroblast",
                     "Basal", "Luminal", "T cells", "Endothelial", "Luminal",
                     "Luminal", "Luminal", "Luminal", "Basal", "Basal",
                     "Luminal", "Luminal", "Luminal Psca", "Luminal", "Luminal",
                     "Luminal", "Luminal Psca", "Luminal", "Mix", "B cells",
                     "Luminal", "Fibroblast", "Luminal", "Pericytes", "Luminal",
                     "Mast cells", "Smooth Muscle", "MDSC", "Mix", "Luminal", 
                     "Luminal", "Glial", "Plasma cells", "Ionocytes", "Luminal",
                     "pDC")

names(new.cluster.ids) <- levels(seurat_combined)
seurat_combined[["cell_type"]] <- new.cluster.ids[as.numeric(as.matrix(seurat_combined@meta.data$SCT_snn_res.0.6))+1]
seurat_combined <- RenameIdents(seurat_combined, new.cluster.ids)

saveRDS(seurat_combined, file = "seurat_WT_HiMYC_BMPC.rds")
```

```
# load seurat object
seurat_combined <- readRDS(file = "seurat_WT_HiMYC_BMPC.rds")

Idents(seurat_combined) <- "cell_type"
levels(seurat_combined)

levels(seurat_combined) <- c("Luminal", "Luminal Psca", "Ionocytes", "Basal",
                        "Endothelial",  "Smooth Muscle", "Pericytes", "Fibroblast",
                        "Macrophages", "MDSC",  "Mast cells", "T cells", "B cells", "Plasma cells", "pDC", "Glial", "Mix")     
                     

DoHeatmap(subset(seurat_combined, downsample = 50), assay = "SCT", features = c(cell_markers), raster = TRUE) +
  scale_fill_viridis()

DimPlot(seurat_combined, label = TRUE, label.size = 6, raster = TRUE) + ggtitle("Cell Type")
```

# integrate libraries
sctransformation, dimensionality reduction
```
# split seurat object
table(seurat_combined$genotype_age)
seurat.list <- SplitObject(seurat_combined, split.by = "genotype_age")

rm(seurat_combined)
gc()

# SCTransform list
for (i in names(seurat.list)) {
    seurat.list[[i]] <- SCTransform(seurat.list[[i]], verbose = FALSE)
}

# integration features
seurat.features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = seurat.features)

# Select reference
reference_dataset <- which(names(seurat.list) == c("WT 10mo", "MYC 10mo"))

# integrate based on reference (you can list multiple references)
seurat.anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", 
    anchor.features = seurat.features, reference = reference_dataset)
seurat.integrated <- IntegrateData(anchorset = seurat.anchors, normalization.method = "SCT")

seurat.integrated <- RunPCA(object = seurat.integrated, verbose = FALSE)
seurat.integrated <- RunUMAP(object = seurat.integrated, dims = 1:30)
```

# cluster analysis
```
seurat.integrated <- FindNeighbors(seurat.integrated, dims = 1:50, verbose = FALSE)
seurat.integrated <- FindClusters(seurat.integrated, resolution = c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6))

UMAP_01 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.1", label = TRUE)  + NoLegend()
UMAP_02 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.2", label = TRUE)  + NoLegend()
UMAP_03 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.3", label = TRUE)  + NoLegend()
UMAP_04 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.4", label = TRUE)  + NoLegend()
UMAP_05 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.5", label = TRUE)  + NoLegend()
UMAP_06 <- DimPlot(seurat.integrated, group.by = "integrated_snn_res.0.6", label = TRUE)  + NoLegend()

table(seurat.integrated$integrated_snn_res.0.3, seurat.integrated$cell_type)
UMAP_01 + UMAP_02 + UMAP_03 + UMAP_04

```

# SCT transformation for aggregated dataset
```
seurat.integrated <- seurat.integrated %>%
  SCTransform(vars.to.regress = c("mitoRatio"), 
              method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")
```

# cell type annotation
0 - Basal
1 - Luminal
2 - Reactive Luminal
3 - Luminal
4 - Fibroblasts
5 - Luminal
6 - Macrophages
7 - Luminal
8 - Luminal
9 - T cells
10 - Endothelial
11 - Basal
12 - Luminal
13 - Luminal
14 - B cells
15 - Smooth Muscle
16 - Mast cells
17 - MDSCs
18 - Glial
19 - Ionocytes
20 - Plasma cells
21 - pDCs

```
Idents(seurat.integrated) <- "integrated_snn_res.0.4"
DoHeatmap(subset(seurat.integrated, downsample = 100), assay = "SCT", features = c(cell_markers)) +
  scale_fill_viridis()

# update clusters
new.cluster.ids <- c("Basal",
                     "Luminal", "Reactive Luminal", "Luminal", "Fibroblasts", "Luminal",
                     "Macrophages", "Luminal", "Luminal", "T cells", "Endothelial",
                     "Basal", "Luminal", "Luminal", "B cells", "Smooth Muscle",
                     "Mast cells", "MDSC", "Glial", "Ionocytes", "Plasma cells",
                     "pDCs")

names(new.cluster.ids) <- levels(seurat.integrated)
seurat.integrated[["cell_type"]] <- new.cluster.ids[as.numeric(as.matrix(seurat.integrated@meta.data$integrated_snn_res.0.4))+1]
seurat.integrated <- RenameIdents(seurat.integrated, new.cluster.ids)

Idents(seurat.integrated) <- "cell_type"
DimPlot(seurat.integrated, label = TRUE)

saveRDS(seurat.integrated, file = "seurat_BMPC_integ.rds")
```

```{r}
# load seurat object
seurat.integrated <- readRDS(file = "seurat_BMPC_integ.rds")

Idents(seurat.integrated) <- "cell_type"
levels(seurat.integrated)

levels(seurat.integrated) <- c("Luminal", "Reactive Luminal", "Ionocytes", "Basal",
                        "Endothelial",  "Smooth Muscle", "Pericytes", "Fibroblasts",
                        "Macrophages", "MDSC",  "Mast cells", "T cells", "B cells", "Plasma cells", "pDCs", "Glial")     
                     

DoHeatmap(subset(seurat.integrated, downsample = 50), assay = "SCT", features = c(cell_markers), raster = TRUE) +
  scale_fill_viridis()

DimPlot(seurat.integrated, label = TRUE, label.size = 6, raster = TRUE) + ggtitle("Cell Type") + NoLegend()

DimPlot(seurat.integrated, group.by = "genotype", shuffle = TRUE, cols = pal5[c(5, 3, 1)],
                         raster = TRUE) + NoAxes() + ggtitle("Mouse Model")

DimPlot(seurat.integrated, group.by = "age", shuffle = TRUE, raster = TRUE) + NoAxes() + ggtitle("Mouse Age")
```

# Ly6d expression in Basal cells
```{r}
seurat.integrated$genotype_age <- paste(seurat.integrated$genotype, seurat.integrated$age, sep = " ")
table(seurat.integrated$cell_type, seurat.integrated$genotype_age)

Idents(seurat.integrated) <- "genotype_age"
primary_seurat <- subset(seurat.integrated, idents = "BMPC 6mo", invert = TRUE)

Idents(primary_seurat) <- "genotype_age"
levels(primary_seurat) <- c("WT 6mo",   "WT 10mo", "MYC 6mo", "MYC 10mo", "BMPC 3mo")
primary_seurat$genotype_age <- primary_seurat@active.ident

seurat.integrated$genotype_age <- gsub("BMPC 6mo", "BMPC Met 6mo", seurat.integrated$genotype_age)
Idents(seurat.integrated) <- "genotype_age"
levels(seurat.integrated) <- c("WT 6mo",   "WT 10mo", "MYC 6mo", "BMPC 3mo", "MYC 10mo", "BMPC Met 6mo")
seurat.integrated$genotype_age <- seurat.integrated@active.ident 

Idents(primary_seurat) <- "cell_type"
seurat_basal <- subset(primary_seurat, idents = "Basal")
Idents(seurat_basal) <- "genotype_age"
levels(seurat_basal) <- c("WT 6mo",   "WT 10mo", "MYC 6mo", "BMPC 3mo", "MYC 10mo", "BMPC Met 6mo")

# expression of Ly6d in basal cells
seurat_basal %>% VlnPlot(features = c("Ly6d"), cols = pal6[c(1, 1, 4, 4, 5)], pt.size = 0) +
  ggtitle("Ly6d Expression \nin Basal Cells") +
  stat_summary(fun = median, geom='point', size = 10, colour = "black", shape = 95)
```

# TREM2 macrophages
```{r}
Idents(seurat.integrated) <- "cell_type"
seurat_macrophages <- subset(seurat.integrated, idents = "Macrophages") 
Idents(seurat_macrophages) <- "genotype_age"
levels(seurat_macrophages) <- c("WT 6mo",   "WT 10mo", "MYC 6mo", "BMPC 3mo", "MYC 10mo", "BMPC Met 6mo")
seurat_macrophages %>% VlnPlot(features = c("Trem2"), cols = pal6[c(1, 1, 4, 4, 5, 6)], pt.size = 0 ) +
  ggtitle("Trem2 Expression \nin Macrophages") +
  stat_summary(fun = median, geom='point', size = 10, colour = "black", shape = 95)
```

# Timp1 fibroblasts
```{r}
Idents(seurat.integrated) <- "cell_type"
seurat_fib <- subset(seurat.integrated, idents = "Fibroblasts") 
Idents(seurat_fib) <- "genotype_age"
levels(seurat_fib) <- c("WT 6mo",   "WT 10mo", "MYC 6mo", "BMPC 3mo", "MYC 10mo", "BMPC Met 6mo")
seurat_fib %>% VlnPlot(features = c("Timp1"), group.by = "genotype_age", cols = pal6[c(1, 1, 4, 4, 5, 6)], pt.size = 0 ) +
  ggtitle("Timp1 Expression \nin Fibroblasts") +
  stat_summary(fun = median, geom='point', size = 10, colour = "black", shape = 95)

```