---
title: "GSEA MYC"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prostatectomy Epithelial Clusters GSEA
 
For the analysis, I adapted this workflow to our dataset:
https://crazyhottommy.github.io/scRNA-seq-workshop-Fall-2019/scRNAseq_workshop_3.html
 
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(RColorBrewer)
library(dplyr)
library(tidyverse)
library(DataCombine)
library(msigdbr)
library(presto)
library(fgsea)
library(SeuratWrappers)
library(UCell)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)
```

load dataframes
```{r}
# load seurat object
seurat_epi <- readRDS(file = "seurat_epi.rds")

Idents(seurat_epi) <- "epithelial_clusters"
DimPlot(seurat_epi)
```

# Make gene lists from MSigDB
```{r}
# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "human", category = "H")

# select MYC Dang targets up from C2 collection
C2_CGP_gene_sets <- msigdbr(species = "human", category = "C2", subcategory = "CGP")
Dang_gene_sets <- C2_CGP_gene_sets %>% filter(gs_name %in% c("DANG_MYC_TARGETS_UP"))

# combined datsets
GSEA_gs <- rbind(Dang_gene_sets, hallmark_gene_sets)
fgsea_sets<- GSEA_gs %>% split(x = .$gene_symbol, f = .$gs_name)
```
# GSESA Cancer vs Luminal by cluster
Cancer1 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer1_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-1", "Luminal"))
head(Cancer1_genes)

# we have all the genes for each cluster
dplyr::count(Cancer1_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer1_genes %>%
  dplyr::filter(group == "Cancer-1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer1_genes %>%
  dplyr::filter(group == "Cancer-1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer1_genes<- Cancer1_genes %>%
  dplyr::filter(group == "Cancer-1") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer1<- deframe(Cancer1_genes)
head(ranks_Cancer1)
tail(ranks_Cancer1)

# perform analysis
fgseaRes_Cancer1 <- fgsea(fgsea_sets, stats = ranks_Cancer1)

# save results
# saveRDS(fgseaRes_Cancer1, file = "fgseaRes_Cancer1.rds")
```

Cancer2 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 

```{r}
Cancer2_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-2", "Luminal"))
head(Cancer2_genes)

# we have all the genes for each cluster
dplyr::count(Cancer2_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer2_genes %>%
  dplyr::filter(group == "Cancer-2") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer2_genes %>%
  dplyr::filter(group == "Cancer-2") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer2_genes<- Cancer2_genes %>%
  dplyr::filter(group == "Cancer-2") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer2<- deframe(Cancer2_genes)
head(ranks_Cancer2)
tail(ranks_Cancer2)

# perform analysis
fgseaRes_Cancer2 <- fgsea(fgsea_sets, stats = ranks_Cancer2)

# save results
# saveRDS(fgseaRes_Cancer2, file = "fgseaRes_Cancer2.rds")

```
Cancer3 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer3_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-3", "Luminal"))
head(Cancer3_genes)

# we have all the genes for each cluster
dplyr::count(Cancer3_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer3_genes %>%
  dplyr::filter(group == "Cancer-3") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer3_genes %>%
  dplyr::filter(group == "Cancer-3") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer3_genes<- Cancer3_genes %>%
  dplyr::filter(group == "Cancer-3") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer3<- deframe(Cancer3_genes)
head(ranks_Cancer3)
tail(ranks_Cancer3)

# perform analysis
fgseaRes_Cancer3 <- fgsea(fgsea_sets, stats = ranks_Cancer3)

# save results
# saveRDS(fgseaRes_Cancer3, file = "fgseaRes_Cancer3.rds")

```
Cancer4 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer4_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-4", "Luminal"))
head(Cancer4_genes)

# we have all the genes for each cluster
dplyr::count(Cancer4_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer4_genes %>%
  dplyr::filter(group == "Cancer-4") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer4_genes %>%
  dplyr::filter(group == "Cancer-4") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer4_genes<- Cancer4_genes %>%
  dplyr::filter(group == "Cancer-4") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer4<- deframe(Cancer4_genes)
head(ranks_Cancer4)
tail(ranks_Cancer4)

# perform analysis
fgseaRes_Cancer4 <- fgsea(fgsea_sets, stats = ranks_Cancer4)

# save results
# saveRDS(fgseaRes_Cancer4, file = "fgseaRes_Cancer4.rds")

```
Cancer5 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer5_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-5", "Luminal"))
head(Cancer5_genes)

# we have all the genes for each cluster
dplyr::count(Cancer5_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer5_genes %>%
  dplyr::filter(group == "Cancer-5") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer5_genes %>%
  dplyr::filter(group == "Cancer-5") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer5_genes<- Cancer5_genes %>%
  dplyr::filter(group == "Cancer-5") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer5<- deframe(Cancer5_genes)
head(ranks_Cancer5)
tail(ranks_Cancer5)

# perform analysis
fgseaRes_Cancer5 <- fgsea(fgsea_sets, stats = ranks_Cancer5)

# save results
# saveRDS(fgseaRes_Cancer5, file = "fgseaRes_Cancer5.rds")

```
Cancer6 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer6_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-6", "Luminal"))
head(Cancer6_genes)

# we have all the genes for each cluster
dplyr::count(Cancer6_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer6_genes %>%
  dplyr::filter(group == "Cancer-6") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer6_genes %>%
  dplyr::filter(group == "Cancer-6") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer6_genes<- Cancer6_genes %>%
  dplyr::filter(group == "Cancer-6") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer6<- deframe(Cancer6_genes)
head(ranks_Cancer6)
tail(ranks_Cancer6)

# perform analysis
fgseaRes_Cancer6 <- fgsea(fgsea_sets, stats = ranks_Cancer6)

# save results
# saveRDS(fgseaRes_Cancer6, file = "fgseaRes_Cancer6.rds")

```
Cancer7 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer7_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-7", "Luminal"))
head(Cancer7_genes)

# we have all the genes for each cluster
dplyr::count(Cancer7_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer7_genes %>%
  dplyr::filter(group == "Cancer-7") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer7_genes %>%
  dplyr::filter(group == "Cancer-7") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer7_genes<- Cancer7_genes %>%
  dplyr::filter(group == "Cancer-7") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer7<- deframe(Cancer7_genes)
head(ranks_Cancer7)
tail(ranks_Cancer7)

# perform analysis
fgseaRes_Cancer7 <- fgsea(fgsea_sets, stats = ranks_Cancer7)

# save results
# saveRDS(fgseaRes_Cancer7, file = "fgseaRes_Cancer7.rds")

```
Cancer8 vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Cancer8_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_clusters', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer-8", "Luminal"))
head(Cancer8_genes)

# we have all the genes for each cluster
dplyr::count(Cancer8_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Cancer8_genes %>%
  dplyr::filter(group == "Cancer-8") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Cancer8_genes %>%
  dplyr::filter(group == "Cancer-8") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Cancer8_genes<- Cancer8_genes %>%
  dplyr::filter(group == "Cancer-8") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Cancer8<- deframe(Cancer8_genes)
head(ranks_Cancer8)
tail(ranks_Cancer8)

# perform analysis
fgseaRes_Cancer8 <- fgsea(fgsea_sets, stats = ranks_Cancer8)

# save results
# saveRDS(fgseaRes_Cancer8, file = "fgseaRes_Cancer8.rds")

```
Make NES heatmap summary
```{r}
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# make dataframe NES
fgsea_NES <- fgseaRes_Cancer1[, c("pathway", "NES")]
colnames(fgsea_NES)[2] <- "Cancer-1"
fgsea_NES$`Cancer-2` <- fgseaRes_Cancer2$NES
fgsea_NES$`Cancer-3` <- fgseaRes_Cancer3$NES
fgsea_NES$`Cancer-4` <- fgseaRes_Cancer4$NES
fgsea_NES$`Cancer-5` <- fgseaRes_Cancer5$NES
fgsea_NES$`Cancer-6` <- fgseaRes_Cancer6$NES
fgsea_NES$`Cancer-7` <- fgseaRes_Cancer7$NES
fgsea_NES$`Cancer-8` <- fgseaRes_Cancer8$NES

fgsea_NES <- column_to_rownames(fgsea_NES, var = "pathway")

fgsea_NES <- as.matrix(t(fgsea_NES))
colnames(fgsea_NES) <- colnames(fgsea_NES)
fgsea_NES <- t(fgsea_NES)

# heatmap of NES
Heatmap(fgsea_NES, column_dend_reorder = TRUE, 
        show_row_dend = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 6),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "NES",
    legend_height = unit(4, "cm")
))

```

# Cancer vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
cancer_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer", "Luminal"))
head(cancer_genes)

# saveRDS(cancer_genes, file = "cancer_genes.rds")

# we have all the genes for each cluster
dplyr::count(cancer_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
cancer_genes<- cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_cancer<- deframe(cancer_genes)
head(ranks_cancer)
tail(ranks_cancer)

# perform analysis
fgseaRes_cancer <- fgsea(fgsea_sets, stats = ranks_cancer)

# save results
# saveRDS(fgseaRes_cancer, file = "fgseaRes_cancer.rds")

# tidy the data
fgseaResTidy_cancer <- fgseaRes_cancer %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_cancer$statistic <- -log10(fgseaResTidy_cancer$padj)

# only plot the top 20 pathways for cancer
plot_CvsL <- ggplot(fgseaResTidy_cancer %>% filter(padj < 0.01, ) %>% top_n(20, -padj), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 20 Pathways \nCancer vs Luminal Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 
plot_CvsL

```

# Intermediate vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}

intermediate_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Atrophic/Intermediate", "Luminal"))
head(intermediate_genes)

# saveRDS(intermediate_genes, file = "intermediate_genes.rds")

# we have all the genes for each cluster
dplyr::count(intermediate_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
intermediate_genes %>%
  dplyr::filter(group == "Atrophic/Intermediate") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

intermediate_genes %>%
  dplyr::filter(group == "Atrophic/Intermediate") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
intermediate_genes<- intermediate_genes %>%
  dplyr::filter(group == "Atrophic/Intermediate") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_intermediate<- deframe(intermediate_genes)
head(ranks_intermediate)
tail(ranks_intermediate)

# perform analysis
fgseaRes_intermediate <- fgsea(fgsea_sets, stats = ranks_intermediate)

# save results
# saveRDS(fgseaRes_intermediate, file = "fgseaRes_intermediate.rds")

# tidy the data
fgseaResTidy_intermediate <- fgseaRes_intermediate %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_intermediate$statistic <- -log10(fgseaResTidy_intermediate$padj)

## only plot the top 20 pathways for intermediate
plot_inter <- ggplot(fgseaResTidy_intermediate %>% filter(padj < 0.01, ) %>% top_n(20, -padj), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 20 Pathways \nAtrophic/Intermediate vs Luminal Clusters") + 
  theme_minimal() + labs(fill =  expression("-Log"[10]*"(adjusted p-value)")) 
plot_inter
```

# Intermediate vs cancer: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}

Int_cancer_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Atrophic/Intermediate", "Cancer"))
head(Int_cancer_genes)

# saveRDS(Int_cancer_genes, file = "Int_cancer_genes.rds")

# we have all the genes for each cluster
dplyr::count(Int_cancer_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Int_cancer_genes<- Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Int_cancer<- deframe(Int_cancer_genes)
head(ranks_Int_cancer)
tail(ranks_Int_cancer)

# perform analysis
fgseaRes_Int_cancer <- fgsea(fgsea_sets, stats = ranks_Int_cancer)

# save results
# saveRDS(fgseaRes_Int_cancer, file = "fgseaRes_Int_cancer.rds")

# tidy the data
fgseaResTidy_Int_cancer <- fgseaRes_Int_cancer %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_Int_cancer$statistic <- -log10(fgseaResTidy_Int_cancer$padj)

## only plot the top 20 pathways for Int_cancer
plot_interCan <- ggplot(fgseaResTidy_Int_cancer %>% filter(padj < 0.01, ) %>% top_n(20, -padj), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 20 Pathways \nCancer vs Atrophic/Intermediate Clusters") + 
  theme_minimal() + labs(fill =  expression("-Log"[10]*"(adjusted p-value)")) 
plot_interCan 
```
# Basal vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
Basal_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Basal", "Luminal"))
head(Basal_genes)

# saveRDS(Basal_genes, file = "Basal_genes.rds")

# we have all the genes for each cluster
dplyr::count(Basal_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Basal_genes %>%
  dplyr::filter(group == "Basal") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Basal_genes %>%
  dplyr::filter(group == "Basal") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Basal_genes<- Basal_genes %>%
  dplyr::filter(group == "Basal") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Basal<- deframe(Basal_genes)
head(ranks_Basal)
tail(ranks_Basal)

# perform analysis
fgseaRes_Basal <- fgsea(fgsea_sets, stats = ranks_Basal)

# save results
# saveRDS(fgseaRes_Basal, file = "fgseaRes_Basal.rds")

# tidy the data
fgseaResTidy_Basal <- fgseaRes_Basal %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_Basal$statistic <- -log10(fgseaResTidy_Basal$padj)

# only plot the top 20 pathways for Basal
plot_Basal <- ggplot(fgseaResTidy_Basal %>% filter(padj < 0.01, ) %>% top_n(20, -padj), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 20 Pathways \nBasal vs Luminal Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 
plot_Basal
```

# MYC gene signature
Identify genes that were upregulated in cancer
```{r}
library(UCell)
library(stringr)
library(patchwork)

DimPlot(seurat_epi, group.by = "epithelial_groups", label = TRUE, raster = TRUE, label.size = 5) + NoLegend() + ggtitle("Epithelial Groups")

# MYC cancer leading edge
leadingEdge_MYC <- fgseaRes_cancer %>% filter(pathway == "HALLMARK_MYC_TARGETS_V1") %>% select(leadingEdge) %>% unlist() %>% as.data.frame()

#as.character(fgseaRes_cancer$leadingEdge) 

# saveRDS(leadingEdge_MYC, file = "leadingEdgeMYCV1.rds")

GeneSig <- list()
GeneSig$MYC_GeneSig <- leadingEdge_MYC$.
GeneSig

# perform gene signature analysis using leading edge genes from MYC V1 pathway analysis
seurat_epi <- AddModuleScore_UCell(seurat_epi, features = GeneSig)

# UMAP of MYC expression
FeaturePlot(seurat_epi, reduction = "umap", features = "MYC", order = TRUE, raster = TRUE)

# UMAP of MYC target genes
pal1 <- c("#3B9AB2", "#78B7C5", "#E1AF00", "#F21A00")
FeaturePlot(seurat_epi, reduction = "umap", features = "MYC_GeneSig_UCell", order = TRUE, raster = FALSE) + 
  scale_color_gradientn(colours = pal1) + ggtitle("MYC V1 Leading Edge")


```

