---
title: "GSEA MYC"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prostatectomy Epithelial Clusters GSEA, MYC targets
This analysis focuses on the epithelial clusters of the prostatectomy dataset, with GSEA focusing on MYC gene signatures
 
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(RColorBrewer)
library(dplyr)
library(tidyverse)
library(DataCombine)
library(msigdbr)
library(presto)
library(fgsea)
library(SeuratWrappers)
library(UCell)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)
```

load dataframes
```{r}
# load seurat object
seurat_epi <- readRDS(file = "seurat_epi.rds")

Idents(seurat_epi) <- "epithelial_clusters"
DimPlot(seurat_epi)
```

# Make gene lists from msigdb
```{r}
# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "human", category = "H")
hallmark_gs <- hallmark_gene_sets %>% split(x = .$gene_symbol, f = .$gs_name)

MYC_Gene_Sets <- hallmark_gene_sets %>% filter(gs_name %in% c("HALLMARK_MYC_TARGETS_V1", "HALLMARK_MYC_TARGETS_V2"))

# select C2 gene sets related to myc
library(readxl)
list_of_MYC_genesets <- read_excel("list of MYC genesets.xlsx", 
    col_names = FALSE)
C2_CGP_gene_sets <- msigdbr(species = "human", category = "C2")
C2_MYC_gene_sets <- C2_CGP_gene_sets %>% filter(gs_name %in% list_of_MYC_genesets$...1)

GSEA_gs <- rbind(C2_MYC_gene_sets, MYC_Gene_Sets)

fgsea_sets<- GSEA_gs %>% split(x = .$gene_symbol, f = .$gs_name)
```

# Cancer vs Luminal: Differential expression analysis for ranked list and GSEA for Hallmark
Presto performs a fast Wilcoxon rank sum test and auROC analysis. 
```{r}
cancer_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Cancer", "Luminal"))
head(cancer_genes)

# saveRDS(cancer_genes, file = "cancer_genes.rds")


# we have all the genes for each cluster
dplyr::count(cancer_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
cancer_genes<- cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_cancer<- deframe(cancer_genes)
head(ranks_cancer)
tail(ranks_cancer)

# perform analysis
fgseaRes_cancer <- fgsea(fgsea_sets, stats = ranks_cancer)

# save results
# saveRDS(fgseaRes_cancer, file = "fgseaRes_cancer.rds")

# tidy the data
fgseaResTidy_cancer <- fgseaRes_cancer %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_cancer$statistic <- -log10(fgseaResTidy_cancer$padj)

# only plot the top 10 up and down pathways for cancer
top_up <- fgseaResTidy_cancer %>% filter(padj < 0.05 ) %>% top_n(10, NES)
top_down <- fgseaResTidy_cancer %>% filter(padj < 0.05 ) %>% top_n(10, -NES)

plot_CvsL <- ggplot(fgseaResTidy_cancer %>% filter(pathway %in% c(top_up$pathway, top_down$pathway)), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top Up and Down MYC GSEA Molecular Signatures\nCancer vs Luminal Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 
plot_CvsL

```

# MYC gene signature
To do: Identify genes that were upregulated in cancer and downregulated

```{r}
library(UCell)
library(stringr)
library(patchwork)

DimPlot(seurat_epi, group.by = "epithelial_groups", label = TRUE, raster = TRUE, label.size = 5) + NoLegend() + ggtitle("Epithelial Groups")

# MYC cancer leading edge
leadingEdge_MYC_Dang_Up <- fgseaRes_cancer %>% filter(pathway == "DANG_BOUND_BY_MYC") %>% select(leadingEdge) %>% unlist() %>% unname() 
leadingEdge_MYC_Dang_DN <- fgseaRes_cancer %>% filter(pathway == "DANG_REGULATED_BY_MYC_DN") %>% select(leadingEdge) %>% unlist() %>% unname()

# saveRDS(leadingEdge_MYC_Dang_Up, file = "leadingEdgeMYCDang.rds")
# saveRDS(leadingEdge_MYC_Dang_DN, file = "leadingEdgeMYCDangDN.rds")

GeneSig <- list()
GeneSig$MYC_Dang <- leadingEdge_MYC_Dang_Up
GeneSig$MYC_Dang_DN <- leadingEdge_MYC_Dang_DN

seurat_epi <- AddModuleScore_UCell(seurat_epi, features = GeneSig, name = "")

pal1 <- c("#3B9AB2", "#78B7C5", "#E1AF00", "#F21A00")
FeaturePlot(seurat_epi, reduction = "umap", features = "MYC_Dang", order = TRUE, raster = TRUE, pt.size = 2)+ 
  scale_color_gradientn(colours = pal1) + ggtitle("DANG BOUND BY MYC \nLeading Edge") + NoAxes()
FeaturePlot(seurat_epi, reduction = "umap", features = "MYC_Dang_DN", order = TRUE, raster = TRUE, pt.size = 2)+ 
  scale_color_gradientn(colours = pal1) + ggtitle("DANG REGULATED BY MYC DN \nLeading Edge") + NoAxes()

# MYC targets down regulated in cancer
FeaturePlot(seurat_epi, features = leadingEdge_MYC_Dang_DN[1:6])

```

# MYC down genes in cancer
```{r}
# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)
library(wesanderson)

# expression matrix
Idents(seurat_epi) <- "epithelial_groups"
levels(seurat_epi)
AvgExp <- seurat_epi %>% AverageExpression()
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# saveRDS(AvgExp_scale, file = "AvgExp_scale.rds")

# filter expression heatmap to leading edge genes
AvgExp_MYC_DN <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_MYC_Dang_DN)) %>% arrange(`Cancer`) %>% as.matrix()
AvgExp_MYC_DN_filter <- AvgExp_MYC_DN %>% as.data.frame() %>% filter(Cancer < 0 & `Atrophic/Intermediate` > 0) %>% as.matrix()

# list of genes down specifical in cancer compared to atrophy
MYC_genes_DN_cancer <- rownames(AvgExp_MYC_DN_filter)
# write.csv(MYC_genes_DN_cancer, file = "MYC_genes_DN_cancer.csv")

sample_order <- c("Basal","Luminal", "Atrophic/Intermediate", "Cancer")
MYC_DN_gene_order <- rownames(AvgExp_MYC_DN_filter)

```

pathway marked genes as annotations
```{r}

pal5 <- wes_palette("Zissou1", 5, "discrete")

#mark which genes are P53
intersect_P53 <- AvgExp_MYC_DN_filter %>% as.data.frame() %>% rownames_to_column("Gene") %>% filter(Gene %in% hallmark_gs$HALLMARK_P53_PATHWAY)
intersect_P53$`P53 Pathway` <- "P53 Pathway"

AvgExp_MYC_DN_df <- AvgExp_MYC_DN_filter %>% as.data.frame() %>% rownames_to_column("Gene")
AvgExp_MYC_DN_merge <- merge(AvgExp_MYC_DN_df, intersect_P53, all = TRUE)

#mark which genes are hypoxia
intersect_hypoxia <- AvgExp_MYC_DN_filter %>% as.data.frame() %>% rownames_to_column("Gene") %>% filter(Gene %in% hallmark_gs$HALLMARK_HYPOXIA)
intersect_hypoxia$Hypoxia <- "Hypoxia"

AvgExp_MYC_DN_merge <- merge(AvgExp_MYC_DN_merge, intersect_hypoxia, all = TRUE)

#mark which genes are apoptosis
intersect_apoptosis <- AvgExp_MYC_DN_filter %>% as.data.frame() %>% rownames_to_column("Gene") %>% filter(Gene %in% hallmark_gs$HALLMARK_APOPTOSIS)
intersect_apoptosis$Apoptosis <- "Apoptosis"

AvgExp_MYC_DN_merge <- merge(AvgExp_MYC_DN_merge, intersect_apoptosis, all = TRUE)

#mark which genes are TNFA
intersect_TNFA <- AvgExp_MYC_DN_filter %>% as.data.frame() %>% rownames_to_column("Gene") %>% filter(Gene %in% hallmark_gs$HALLMARK_TNFA_SIGNALING_VIA_NFKB)
intersect_TNFA$TNFA <- "TNFA signaling via NFKB"

AvgExp_MYC_DN_merge <- merge(AvgExp_MYC_DN_merge, intersect_TNFA, all = TRUE)

# column to row names
AvgExp_MYC_DN_merge <- AvgExp_MYC_DN_merge %>% column_to_rownames("Gene")

# row annotations
ha <- HeatmapAnnotation(
  `P53 Pathway` = AvgExp_MYC_DN_merge$`P53 Pathway`,
  Apoptosis = AvgExp_MYC_DN_merge$Apoptosis,
  Hypoxia = AvgExp_MYC_DN_merge$Hypoxia,
  TNFA = AvgExp_MYC_DN_merge$TNFA,
  #row_names_rot = 45,
  which = "row",
  na_col = "light grey",
  gp = gpar(col = "white"),
  col = list(`P53 Pathway` = c("P53 Pathway" = pal5[1]),
             Apoptosis = c("Apoptosis" = pal5[2]),
             Hypoxia = c("Hypoxia" = pal5[3]),
            TNFA = c("TNFA signaling via NFKB" = pal5[5]))
)


# heatmap with annotations
MYC_DN_annon <- Heatmap(as.matrix(AvgExp_MYC_DN_merge[1:4]),
                column_order = sample_order,
                row_dend_reorder = FALSE,
                show_row_dend = FALSE,
                show_column_dend = FALSE,
                row_names_side = "right",
                column_title = ("MYC target genes down regulated in Cancer"), 
                row_names_gp = gpar(fontsize = 12),
                column_names_rot = 45,
                left_annotation = ha,
                heatmap_legend_param = list(
                  title = "Z-score"
                ))

MYC_DN_annon

```

# GSEA overlap analysis
https://www.gsea-msigdb.org/gsea/msigdb/annotate.jsp
Perform overlap analysis using leading edge genes, save results and load
```{r}
# gene list
leadingEdge_MYC_Dang_DN

# overlap analysis results
library(readxl)
library(viridis)
MYC_dn_leadingEdge_overlap <- read_excel("MYC_dn_leadingEdge_overlap.xlsx", 
    skip = 8)
MYC_dn_leadingEdge_overlap$statistic <- -log10(MYC_dn_leadingEdge_overlap$`FDR q-value`)

# make a dotplot
GSEA_overlap <- ggplot(MYC_dn_leadingEdge_overlap, aes(reorder(`Gene Set Name`, `# Genes in Overlap (k)`), `# Genes in Overlap (k)`)) +
  geom_point(aes(color= statistic, size = `k/K`)) +
  coord_flip() +
  labs(x="Pathway", y="Number of Overlapped Genes (k)",
       title="Computed Overlaps with Hallmark Collection for \nDANG REGULATED BY MYC DN Lagging Genes") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) +
  scale_color_viridis()
GSEA_overlap
```


# PTEN/AKT/mTOR signaling axis
PTEN is a negative regulator of AKT/mTOR signaling
Loss of PTEN accelerates tumor progression
```{r}
# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "human", category = "H")
hallmark_gs <- hallmark_gene_sets %>% split(x = .$gene_symbol, f = .$gs_name)

PTEN_Gene_Sets <- hallmark_gene_sets %>% 
  filter(gs_name %in% c("HALLMARK_P53_PATHWAY", "HALLMARK_PI3K_AKT_MTOR_SIGNALING", "HALLMARK_MTORC1_SIGNALING",
                        "HALLMARK_APOPTOSIS", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_HYPOXIA"))

# select C2 gene sets related to PTEN/AKT
library(readxl)
list_of_PTEN_genesets <- read_excel("list of PTEN genesets.xlsx", 
    col_names = FALSE)
list_of_AKT_genesets <- read_excel("list of AKT genesets.xlsx", col_names = FALSE)
C2_CGP_gene_sets <- msigdbr(species = "human", category = "C2")
C2_PTEN_gene_sets <- C2_CGP_gene_sets %>% filter(gs_name %in% c(list_of_AKT_genesets$...1, list_of_PTEN_genesets$...1))

GSEA_gs <- rbind(C2_PTEN_gene_sets, PTEN_Gene_Sets)

fgsea_sets_PTEN<- GSEA_gs %>% split(x = .$gene_symbol, f = .$gs_name)
```

comparison of cancer and atrophy
```{r}
Int_cancer_genes <- wilcoxauc(X = seurat_epi, group_by = 'epithelial_groups', seurat_assay = 'SCT', assay = 'data', groups_use = c("Atrophic/Intermediate", "Cancer"))
head(Int_cancer_genes)

# saveRDS(Int_cancer_genes, file = "Int_cancer_genes.rds")
# Int_cancer_genes <- readRDS(file = "Int_cancer_genes.rds")

# we have all the genes for each cluster
dplyr::count(Int_cancer_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
Int_cancer_genes<- Int_cancer_genes %>%
  dplyr::filter(group == "Cancer") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_Int_cancer<- deframe(Int_cancer_genes)
head(ranks_Int_cancer)
tail(ranks_Int_cancer)

# perform analysis
fgseaRes_Int_cancer <- fgsea(fgsea_sets_PTEN, stats = ranks_Int_cancer)

# save results
# saveRDS(fgseaRes_Int_cancer, file = "fgseaRes_Int_cancer.rds")

# tidy the data
fgseaResTidy_int_cancer_PTEN <- fgseaRes_Int_cancer %>%
  as_tibble() %>%
 arrange(padj)

# add statistic
fgseaResTidy_int_cancer_PTEN$statistic <- -log10(fgseaResTidy_int_cancer_PTEN$padj)
```

```
# only plot the top 10 up and down pathways for cancer vs atrophys
top_up <- fgseaResTidy_int_cancer_PTEN %>% filter(padj < 0.05 ) %>% top_n(10, NES)
top_up$pathway
top_down <- fgseaResTidy_int_cancer_PTEN %>% filter(padj < 0.05 ) %>% top_n(10, -NES)
top_down$pathway
```

```{r}
pathway_list <- c("HALLMARK_MTORC1_SIGNALING", "HALLMARK_PI3K_AKT_MTOR_SIGNALING", "PID_PI3KCI_AKT_PATHWAY",
                  "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_APOPTOSIS", "HALLMARK_P53_PATHWAY", "HALLMARK_HYPOXIA",
                  "REACTOME_NEGATIVE_REGULATION_OF_THE_PI3K_AKT_NETWORK")


plot_int_cancer_PTEN <- ggplot(fgseaResTidy_int_cancer_PTEN %>% filter(pathway %in% pathway_list), aes(reorder(pathway, NES), NES)) +
geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Selected Genesets for GSEA Cancer vs Intermediate Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 
 plot_int_cancer_PTEN

# any overlap with cancer downregulated genes?
intersect(MYC_genes_DN_cancer, fgsea_sets$REACTOME_NEGATIVE_REGULATION_OF_THE_PI3K_AKT_NETWORK)
```

```{r}
# IER3 is a negative regulator of PI3K_AKT, and is suppressed in cancer
FeaturePlot(seurat_epi, features = "IER3", order = TRUE) + scale_color_gradientn(colors = pal5) + NoAxes()

```