---
title: "Hi-Myc Epithelial"
output: html_document
date: "2022-11-29"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Epithelial Subset
Load integrated seurat object of mouse data and subset to epithelial clusters. Perform normalization, dimensionality reduction and clustering analysis.

```{r}
# Libraries
library(Seurat)
library(dplyr)
library(readr)
library(DataCombine)
library(viridis)
library(ggplot2)
library(glmGamPoi)
library(wesanderson)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

# color palette
pal5 <- wes_palette("Zissou1", 5, "continuous")
```

load seurat object and subset
```
strain.mouse_seurat <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")
DimPlot(strain.mouse_seurat)

# subset to epithelial cells
epithelial_seurat <- subset(mouse_seurat_epi, idents = c("Basal", "Luminal", "Ionocytes"))
DimPlot(epithelial_seurat, group.by = "strain_lobe")
```

# integrate data by strain
https://satijalab.org/seurat/articles/integration_large_datasets.html
https://satijalab.org/seurat/articles/integration_rpca.html

```
# split the dataset into a list by strain (BL6 and FVB)
seurat_object.list <- SplitObject(epithelial_seurat, split.by = "strain")
seurat_object.list <- lapply(X = seurat_object.list, FUN = SCTransform, method = "glmGamPoi", vst.flavor = "v2")

# run SCTransform on each dataset split by strain
features <- SelectIntegrationFeatures(object.list = seurat_object.list, nfeatures = 3000)
seurat_object.list <- PrepSCTIntegration(object.list = seurat_object.list, anchor.features = features)
seurat_object.list <- lapply(X = seurat_object.list, FUN = RunPCA, features = features)

# integrate datasets
strain.anchors <- FindIntegrationAnchors(object.list = seurat_object.list, normalization.method = "SCT",
    anchor.features = features, dims = 1:30, reduction = "rpca", k.anchor = 20)
mouse_seurat_epi <- IntegrateData(anchorset = strain.anchors, normalization.method = "SCT", dims = 1:30)

# dimensional reduction
mouse_seurat_epi <- RunPCA(mouse_seurat_epi, verbose = FALSE)
mouse_seurat_epi <- RunUMAP(mouse_seurat_epi, reduction = "pca", dims = 1:40)
DimPlot(mouse_seurat_epi, group.by = "strain", shuffle = TRUE)
DimPlot(mouse_seurat_epi, group.by = "genotype", shuffle = TRUE)
DimPlot(mouse_seurat_epi, group.by = "lobe", shuffle = TRUE)
DimPlot(mouse_seurat_epi, group.by = "cell_type", shuffle = TRUE)

# clustering analysis
mouse_seurat_epi <- FindNeighbors(mouse_seurat_epi, reduction = "pca", dims = 1:30)
mouse_seurat_epi <- FindClusters(mouse_seurat_epi, resolution = 0.12)
DimPlot(mouse_seurat_epi, label = T, repel = F, shuffle = TRUE, group.by = "integrated_snn_res.0.12") + ggtitle("Unsupervised clustering")

```

# SCT transform without integration
Perform SCTransform for all genes, keeping integrating clustering
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
# run sctransform
mouse_seurat_epi <- SCTransform(mouse_seurat_epi, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

saveRDS(mouse_seurat_epi, file = "mouse_seurat_epi.rds.rds")
mouse_seurat_epi <- readRDS(file = "mouse_seurat_epi.rds.rds")
```

# Identifying Clusters
name clusters based on marker genes/lobe enrichment
0 Luminal MYC 1
1 Basal
2 Luminal Ventral
3 Luminal Anterior
4 Luminal Dorsal
5 Luminal MYC 2
6 Reactive Basal
7 Luminal Lateral
8 Luminal Psca Ly6d
```
# update clusters with unique cluster name
Idents(mouse_seurat_epi) <- "integrated_snn_res.0.12"
DimPlot(mouse_seurat_epi, label = TRUE)
new.cluster.ids <- c("Luminal MYC 1",
                     "Basal", "Luminal Ventral", "Luminal Anterior", "Luminal Dorsal", "Luminal MYC 2",
                     "Reactive Basal", "Luminal Lateral", "Luminal Psca Ly6d")

names(new.cluster.ids) <- levels(mouse_seurat_epi)
mouse_seurat_epi <- RenameIdents(mouse_seurat_epi, new.cluster.ids)
mouse_seurat_epi$Epi_clusters <- mouse_seurat_epi@active.ident
```
UMAP plot
```{r}
mouse_seurat_epi <- readRDS(file = "mouse_seurat_epi.rds")

Idents(mouse_seurat_epi) <- "Epi_clusters"
DimPlot(mouse_seurat_epi, reduction = "umap", label = TRUE, pt.size = 1, group.by = "Epi_clusters", shuffle = TRUE, raster = TRUE, label.size = 5, repel = TRUE) + NoLegend() + ggtitle("Mouse Epithelial Clusters")

# UMAP for Ly6d and Psca
FeaturePlot(mouse_seurat_epi, features = "Ly6d", order = TRUE, raster = TRUE, pt.size = 2) + scale_color_gradientn(colours = pal5) + NoAxes()

FeaturePlot(mouse_seurat_epi, features = "Psca", order = TRUE, raster = TRUE, pt.size = 2) + scale_color_gradientn(colours = pal5) + NoAxes()
```

# Performed subsetted analysis of the Luminal Psca Ly6d cluster
Prompted by reviewers comments, we sought to define the urothelial cells, as they express Ly6d and Psca 
See "mouseMYC_epithelial_integStrain_reviewer2.Rmd" for original analysis

subset cells to Luminal Psca cluster for subcluster analysis
```
levels(mouse_seurat_epi)
Psca_seurat <- subset(mouse_seurat_epi, idents = "Luminal Psca Ly6d")
DimPlot(Psca_seurat, group.by = "genotype", shuffle = TRUE)
FeaturePlot(Psca_seurat, features = c("Krt4", "Psca"), order = TRUE)

# sctransform
Psca_seurat <- SCTransform(Psca_seurat, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

# dimensionality reduction
Psca_seurat <- RunPCA(Psca_seurat, verbose = FALSE)
Psca_seurat <- RunUMAP(Psca_seurat, reduction = "pca", dims = 1:30)
Psca_seurat <- FindNeighbors(Psca_seurat, dims = 1:50, verbose = FALSE)
```

```
DimPlot(Psca_seurat, group.by = "genotype")
DimPlot(Psca_seurat, group.by = "strain")
```

integration by strain using harmony
There is clear separation by strain of the subclusters
```
library(harmony)
library(wesanderson)
library(patchwork)

# color palettes
pal5 <- wes_palette("Zissou1", 5, "continuous")

# make integrated object
harmonized_seurat <- RunHarmony(Psca_seurat, 
				group.by.vars = c("strain"), 
				reduction = "pca", assay.use = "SCT", reduction.save = "harmony")

harmonized_seurat <- RunUMAP(harmonized_seurat, reduction = "harmony", assay = "SCT", dims = 1:40)
harmonized_seurat <- FindNeighbors(object = harmonized_seurat, reduction = "harmony")
harmonized_seurat <- FindClusters(harmonized_seurat, resolution = c(0.05, 0.1, 0.2, 0.3))

DimPlot(harmonized_seurat, group.by = "genotype")
DimPlot(harmonized_seurat, group.by = "strain")

DimPlot(harmonized_seurat, group.by = "SCT_snn_res.0.05", label = TRUE)
DimPlot(harmonized_seurat, group.by = "SCT_snn_res.0.1", label = TRUE)
DimPlot(harmonized_seurat, group.by = "SCT_snn_res.0.2", label = TRUE)

table(harmonized_seurat$SCT_snn_res.0.1, harmonized_seurat$genotype)

Idents(harmonized_seurat) <- "SCT_snn_res.0.1"
clusters_psca <- FindAllMarkers(harmonized_seurat)
saveRDS(clusters_psca, file = "clusters_psca.rds")

clusters_psca_top <- clusters_psca %>% group_by(cluster) %>% filter(avg_log2FC > 0, pct.2 < 0.2) %>% top_n(10, -p_val_adj)

subset(harmonized_seurat, downsample = 100) %>% DoHeatmap(features = c("Psca", "Krt4", "Ly6d", clusters_psca_top$gene))

FeaturePlot(harmonized_seurat, features = c("Krt17"), min.cutoff = "q1", max.cutoff = "q99")
FeaturePlot(harmonized_seurat, features = c("Ly6d"), min.cutoff = "q1", max.cutoff = "q99")
FeaturePlot(harmonized_seurat, features = c("Krt4"), min.cutoff = "q1", max.cutoff = "q99")
FeaturePlot(harmonized_seurat, features = c("Slc25a48"), min.cutoff = "q1", max.cutoff = "q99")

# epithelial clusters
Idents(harmonized_seurat) <- "SCT_snn_res.0.1"
new.cluster.ids <- c("Reactive Luminal",
                     "Luminal", "Urothelial")
names(new.cluster.ids) <- levels(harmonized_seurat)
harmonized_seurat <- RenameIdents(harmonized_seurat, new.cluster.ids)
harmonized_seurat$luminal_ly6d <- harmonized_seurat@active.ident

# which lobe are the luminal cells from
table(harmonized_seurat$luminal_ly6d, harmonized_seurat$lobe)
harmonized_luminal <- subset(harmonized_seurat, idents = "Luminal")
harmonized_seurat$luminal_lobe <- harmonized_luminal$lobe
harmonized_seurat$luminal_lobe <- paste(harmonized_seurat$luminal_ly6d, harmonized_seurat$luminal_lobe, sep = " ")
harmonized_seurat$luminal_lobe <- gsub(" NA", "", harmonized_seurat$luminal_lobe)

table(harmonized_seurat$luminal_lobe)

# save seurat object
# saveRDS(harmonized_seurat, file = "seurat_Luminal_psca.rds")
```

```{r}
harmonized_seurat <- readRDS(file = "seurat_Luminal_psca.rds") 
clusters_psca <- readRDS(file = "clusters_psca.rds")

clusters_psca_top <- clusters_psca %>% group_by(cluster) %>% filter(avg_log2FC > 0, pct.2 < 0.2) %>% top_n(10, -p_val_adj)


Idents(harmonized_seurat) <- "luminal_ly6d"
levels(harmonized_seurat) <- c("Reactive Luminal",
                     "Luminal", "Urothelial")

DimPlot(harmonized_seurat, label = TRUE, raster = TRUE, pt.size = 3) + ggtitle("Mouse Luminal Ly6d Subclusters")
DimPlot(harmonized_seurat, label = FALSE, raster = TRUE, pt.size = 3, split.by = "genotype") & NoLegend()

harmonized_seurat %>% DotPlot(features = unique(c("Krt8", "Psca", "Krt4", "Ly6d", "Krt6a", clusters_psca_top$gene))) + scale_color_gradientn(colors = pal5) +
  theme(axis.text.x=element_text(angle=45,hjust=1))
```

Add subclustering to main epithelial seurat object
```
levels(mouse_seurat_epi)

mouse_seurat_epi$cell_type_Psca <- mouse_seurat_epi$Epi_clusters
mouse_seurat_epi$cell_type_Psca <- gsub("Luminal Psca Ly6d", "", mouse_seurat_epi$cell_type_Psca)

# add luminal psca ly6d subclusters
mouse_seurat_epi$Luminal_Psca_Ly6d <- harmonized_seurat$luminal_lobe
mouse_seurat_epi$cell_type_Psca <- paste(mouse_seurat_epi$cell_type_Psca, mouse_seurat_epi$Luminal_Psca_Ly6d, sep = "")
mouse_seurat_epi$cell_type_Psca <- gsub("NA", "", mouse_seurat_epi$cell_type_Psca)
mouse_seurat_epi$cell_type_Psca <- gsub("Luminal Psca Ly6d", "", mouse_seurat_epi$cell_type_Psca)

table(mouse_seurat_epi$cell_type_Psca)

# remove lobe distinction for luminal clusters
mouse_seurat_epi$epi_clusters_noLobe <- mouse_seurat_epi$cell_type_Psca
mouse_seurat_epi$epi_clusters_noLobe <- gsub(" Anterior", "", mouse_seurat_epi$epi_clusters_noLobe)
mouse_seurat_epi$epi_clusters_noLobe <- gsub(" Dorsal", "", mouse_seurat_epi$epi_clusters_noLobe)
mouse_seurat_epi$epi_clusters_noLobe <- gsub(" Lateral", "", mouse_seurat_epi$epi_clusters_noLobe)
mouse_seurat_epi$epi_clusters_noLobe <- gsub(" Ventral", "", mouse_seurat_epi$epi_clusters_noLobe)
table(mouse_seurat_epi$epi_clusters_noLobe)
```

UMAPs of updated epithelial seurat
```{r}
Idents(mouse_seurat_epi) <- "cell_type_Psca"
DimPlot(mouse_seurat_epi, raster = TRUE, label = TRUE) + NoLegend() + ggtitle("Mouse Prostate Epithelial Clusters")

# UMAPs for reactive stroma
FeaturePlot(mouse_seurat_epi, features = c("Tpm2", "Ly6c1", "Calcb"), 
                             order = TRUE, raster = TRUE, pt.size = 2, ncol = 1) & 
  scale_color_gradientn(colours = pal5) & NoAxes()
```

# Cell proportions test by RAISIN
https://github.com/zji90/raisin//
```{r}

library(raisin)

# extract metadata from seurat object
Idents(mouse_seurat_epi) <- "epi_clusters_noLobe"
cluster <- Idents(mouse_seurat_epi)

Idents(mouse_seurat_epi) <- "library_id"
individual <- Idents(mouse_seurat_epi)

# make the design 
design <- data.frame(intercept=1,contrast=c(
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1),
                     row.names = c("AB1", "AB2", "AF1", "AF2", "AHMB1",
                                   "AHMB2", "AHMB3", "AHMF1", "AHMF2",
                                   "DB1", "DB2", "DF1", "DF2", "DHMB1",
                                   "DHMB2", "DHMB3", "DHMF1", "DHMF2",
                                   "LB1", "LB2", "LF1", "LF2", "LHMB1",
                                   "LHMB2", "LHMB3", "LHMF1", "LHMF2",
                                   "VB1", "VB2", "VF1", "VF2", "VHMB1",
                                   "VHMB2", "VHMB3", "VHMF1", "VHMF2"))

# run the test
prop_test_results <- proportiontest(cluster,individual,design)
prop_test_results <- prop_test_results %>% tibble::rownames_to_column("Cluster")
prop_test_results
```

Plot cell proportion diferences
```{r}
library(ggpubr)
library_counts <- as.data.frame(table(mouse_seurat_epi$library_id))

# create dataframe containing all relevant metadata
cell_type_enrich <- as.data.frame(table(mouse_seurat_epi$library_id,mouse_seurat_epi$genotype, mouse_seurat_epi$lobe, mouse_seurat_epi$strain, mouse_seurat_epi$epi_clusters_noLobe))

cell_type_enrich$library <- library_counts$Freq
cell_type_enrich$fraction <- cell_type_enrich$Freq / cell_type_enrich$library

# rename columns to meaningful column names
colnames(cell_type_enrich) <- c("Library", "Genotype", "Lobe", "Strain", "Cluster", "Counts", "Library_Counts", "Fraction")

# filter to library
AB1 <- cell_type_enrich %>% filter(Library == "AB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Anterior")
AB2 <- cell_type_enrich %>% filter(Library == "AB2", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Anterior")
AF1 <- cell_type_enrich %>% filter(Library == "AF1", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Anterior")
AF2 <- cell_type_enrich %>% filter(Library == "AF2", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Anterior")
AHMB1 <- cell_type_enrich %>% filter(Library == "AHMB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Anterior")
AHMB2 <- cell_type_enrich %>% filter(Library == "AHMB2", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Anterior")
AHMB3 <- cell_type_enrich %>% filter(Library == "AHMB3", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Anterior")
AHMF1 <- cell_type_enrich %>% filter(Library == "AHMF1", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Anterior")
AHMF2 <- cell_type_enrich %>% filter(Library == "AHMF2", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Anterior")
DB1 <- cell_type_enrich %>% filter(Library == "DB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Dorsal")
DB2 <- cell_type_enrich %>% filter(Library == "DB2", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Dorsal")
DF1 <- cell_type_enrich %>% filter(Library == "DF1", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Dorsal")
DF2 <- cell_type_enrich %>% filter(Library == "DF2", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Dorsal")
DHMB1 <- cell_type_enrich %>% filter(Library == "DHMB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Dorsal")
DHMB2 <- cell_type_enrich %>% filter(Library == "DHMB2", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Dorsal")
DHMB3 <- cell_type_enrich %>% filter(Library == "DHMB3", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Dorsal")
DHMF1 <- cell_type_enrich %>% filter(Library == "DHMF1", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Dorsal")
DHMF2 <- cell_type_enrich %>% filter(Library == "DHMF2", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Dorsal")
LB1 <- cell_type_enrich %>% filter(Library == "LB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Lateral")
LB2 <- cell_type_enrich %>% filter(Library == "LB2", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Lateral")
LF1 <- cell_type_enrich %>% filter(Library == "LF1", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Lateral")
LF2 <- cell_type_enrich %>% filter(Library == "LF2", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Lateral")
LHMB1 <- cell_type_enrich %>% filter(Library == "LHMB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Lateral")
LHMB2 <- cell_type_enrich %>% filter(Library == "LHMB2", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Lateral")
LHMB3 <- cell_type_enrich %>% filter(Library == "LHMB3", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Lateral")
LHMF1 <- cell_type_enrich %>% filter(Library == "LHMF1", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Lateral")
LHMF2 <- cell_type_enrich %>% filter(Library == "LHMF2", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Lateral")
VB1 <- cell_type_enrich %>% filter(Library == "VB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Ventral")
VB2 <- cell_type_enrich %>% filter(Library == "VB2", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Ventral")
VF1 <- cell_type_enrich %>% filter(Library == "VF1", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Ventral")
VF2 <- cell_type_enrich %>% filter(Library == "VF2", Strain == "FVB/NJ", Genotype == "WT", Lobe == "Ventral")
VHMB1 <- cell_type_enrich %>% filter(Library == "VHMB1", Strain == "C57BL/6J", Genotype == "WT", Lobe == "Ventral")
VHMB2 <- cell_type_enrich %>% filter(Library == "VHMB2", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Ventral")
VHMB3 <- cell_type_enrich %>% filter(Library == "VHMB3", Strain == "C57BL/6J", Genotype == "MYC", Lobe == "Ventral")
VHMF1 <- cell_type_enrich %>% filter(Library == "VHMF1", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Ventral")
VHMF2 <- cell_type_enrich %>% filter(Library == "VHMF2", Strain == "FVB/NJ", Genotype == "MYC", Lobe == "Ventral")

filtered_library_fraction <- rbind(AB1, AB2, AF1, AF2,
                                   AHMB1, AHMB2, AHMB3, AHMF1, AHMF2,
                                   DB1, DB2, DF1, DF2,
                                   DHMB1, DHMB2, DHMB3, DHMF1, DHMF2,
                                   LB1, LB2, LF1, LF2, 
                                   LHMB1, LHMB2, LHMB3, LHMF1, LHMF2,
                                   VB1, VB2, VF1, VF2,
                                   VHMB1, VHMB2, VHMB3, VHMF1, VHMF2)

# combine DF with cell proportion test results
combined_df <- merge(filtered_library_fraction, prop_test_results, by.x = "Cluster", by.y = "Cluster")

combined_df$Genotype <- factor(combined_df$Genotype, levels = c("WT", "MYC"))

# Add pvalue to cluster name
library(scales)
combined_df$cluster_adjP <- paste(combined_df$Cluster, 
                                  "\n logFC = ", 
                                  signif(combined_df$logFC, 2),
                                  "\n adj p-value = ", scientific(combined_df$adj.P.Val, 1))

# Plot clusters
combined_df %>% 
  ggplot(aes(x = Genotype, y = Fraction))+
  geom_jitter(aes(fill = Genotype, color = Lobe), size = 2, 
             width = 0.2) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
                 geom = "crossbar", width = 0.5) +
  ggtitle("Epithelial Clusters") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~cluster_adjP, scales = "free", ncol = 4) +
  scale_color_manual(values = c("#219ebc", "#023047", "#ffb703", "#fb8500"))
```

# GSEA luminal cells by genotype
```{r}
library(msigdbr)
library(presto)
library(fgsea)
library(DataCombine)
library(tidyverse)

# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "mouse", category = "H")
fgsea_sets<- hallmark_gene_sets %>% split(x = .$gene_symbol, f = .$gs_name)
```

WT vs Hi-Myc
```{r}
# subset to luminal cells
Idents(mouse_seurat_epi) <- "cell_type"
levels(mouse_seurat_epi)
luminal_seurat <- subset(mouse_seurat_epi, idents = "Luminal")
DimPlot(luminal_seurat)

luminal_genes <- wilcoxauc(X = luminal_seurat, group_by = 'genotype', seurat_assay = 'SCT', assay = 'data')
head(luminal_genes)

# we have all the genes for each cluster
dplyr::count(luminal_genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
luminal_genes %>%
  dplyr::filter(group == "MYC") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

luminal_genes %>%
  dplyr::filter(group == "MYC") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
luminal_genes<- luminal_genes %>%
  dplyr::filter(group == "MYC") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_luminal<- deframe(luminal_genes)
head(ranks_luminal)
tail(ranks_luminal)

# perform analysis
fgseaRes_luminal <- fgsea(fgsea_sets, stats = ranks_luminal)

# save results
# saveRDS(fgseaRes_luminal, file = "fgseaRes_MYCluminal.rds")

# tidy the data
fgseaResTidy_luminal <- fgseaRes_luminal %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_luminal$statistic <- -log10(fgseaResTidy_luminal$padj)

# only plot the top 20 pathways for luminal
ggplot(fgseaResTidy_luminal %>% filter(padj < 0.01, ) %>% top_n(20, -padj), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 20 Pathways \nHi-Myc vs WT Luminal Cells") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 

# MYC leading edge
leadingEdge_MYC <- fgseaRes_luminal %>% filter(pathway == "HALLMARK_MYC_TARGETS_V1") %>% select(leadingEdge) %>% unlist() %>% as.data.frame() %>% head(50)

```

Heatmap of MYC V1 leading edge genes
```{r}
# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# expression matrix
Idents(mouse_seurat_epi) <- "cell_type_Psca"
levels(mouse_seurat_epi)
AvgExp <- AverageExpression(mouse_seurat_epi)
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# filter to leading edge genes
AvgExp_MYCV1 <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(leadingEdge_MYC$.)) %>% as.matrix()
AvgExp_MYC <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c("MYC", "Myc")) %>% t()
AvgExp_MYC_order <- AvgExp_MYC %>% as.data.frame() %>% arrange(desc(MYC))
colnames(AvgExp_MYC)[2] <- "MYC Transgene"
AvgExp_MYC <- AvgExp_MYC[, c("MYC Transgene", "Myc")]

# top annotation
library(circlize)
MYC.color <- colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
top.annotations = HeatmapAnnotation(`MYC Z-score` = AvgExp_MYC, col = list(`MYC Z-score` = MYC.color), annotation_name_side = "left")

# complex heatmap
heatmap_mycv1 <- Heatmap(t(AvgExp_MYCV1), row_order = row.names(AvgExp_MYC_order),
        #top_annotation = top.annotations,
        column_dend_reorder = FALSE, 
        row_dend_reorder = FALSE, 
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm")
))

heatmap_myc <- Heatmap(AvgExp_MYC, row_order = row.names(AvgExp_MYC_order),
        #top_annotation = top.annotations,
        column_dend_reorder = FALSE, 
        row_dend_reorder = FALSE, 
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm"),
    at = seq(-2, 2, 1)
))

heatmap_myc + heatmap_mycv1
```

# Dotplot of Ly6d expression cluster genes
```{r}

Idents(mouse_seurat_epi) <- "cell_type_Psca"
levels(mouse_seurat_epi)

# Find differentially expressed genes for clusters
# seurat_clusters_mouse_epi <- FindAllMarkers(mouse_seurat_epi)
# saveRDS(seurat_clusters_mouse_epi, file = "seurat_clusters_mouse_epi.rds")
seurat_clusters_mouse_epi <- readRDS(file = "seurat_clusters_mouse_epi.rds")

# Filter for top genes
Seurat_basalLy6d  <- seurat_clusters_mouse_epi %>% filter(cluster == "Reactive Basal", avg_log2FC > 0.5, pct.1 > 0.5, pct.2 < 0.2) %>% top_n(10, -p_val_adj) %>% top_n(10, avg_log2FC) %>% arrange(pct.1)

Seurat_luminalLy6d  <- seurat_clusters_mouse_epi %>% filter(cluster == "Reactive Luminal", avg_log2FC > 0.5, pct.1 > 0.5, pct.2 < 0.2) %>% top_n(10, -p_val_adj) %>% top_n(10, avg_log2FC) %>% arrange(pct.1)

Seurat_urothelial  <- seurat_clusters_mouse_epi %>% filter(cluster == "Urothelial", avg_log2FC > 0.5, pct.1 > 0.5, pct.2 < 0.2) %>% top_n(10, -p_val_adj) %>% top_n(10, avg_log2FC) %>% arrange(pct.1)

ly6d_intersect <- intersect(Seurat_luminalLy6d$gene, Seurat_basalLy6d$gene)
ly6d_intersect


# dotplot of top genes for Basal Ly6d cluster including urethral markers
# Sca1/Ly6a, Tacstd2/Trop2, and Psca
library(wesanderson)
pal <- wes_palette("Zissou1", 6, type = "continuous")

Idents(mouse_seurat_epi) <- "epi_clusters_noLobe" 
levels(mouse_seurat_epi) <- c("Reactive Basal", "Reactive Luminal", "Urothelial", "Basal", "Luminal", "Luminal MYC 1", "Luminal MYC 2")

DotPlot(mouse_seurat_epi, features = unique(c("Krt4", "Psca", "Ly6a", "Tacstd2", "Ly6d", Seurat_basalLy6d$gene, Seurat_luminalLy6d$gene, Seurat_urothelial$gene))) +
  scale_color_gradientn(colours = pal) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))#, vjust = 0.5, hjust=1))

```

UMAPs of mouse data
```{r}
Idents(mouse_seurat_epi) <- "genotype"
levels(mouse_seurat_epi) <- as.factor(c("WT", "MYC"))
mouse_seurat_epi$genotype <- mouse_seurat_epi@active.ident

Idents(mouse_seurat_epi) <- "cell_type_Psca"
DimPlot(mouse_seurat_epi, raster = TRUE, label = TRUE, pt.size = 2) + ggtitle("Mouse Prostate \nEpithelial Clusters")

DimPlot(mouse_seurat_epi, raster = TRUE, label = FALSE, split.by = "genotype", pt.size = 2) + NoLegend()
```

# Ly6d correlation
co-expressed genes
```
# find co-expressed genes
matrix<-mouse_seurat_epi@assays$SCT@scale.data
matrix_mod<-as.matrix(matrix)
gene<-as.numeric(matrix_mod["Ly6d",])
correlations<-apply(matrix_mod,1,function(x){cor(gene,x)})
correlations_ly6d <- as.data.frame(correlations) %>% tibble::rownames_to_column("gene") %>% arrange(desc(correlations))
correlations_ly6d_top20 <- top_n(correlations_ly6d, 21, correlations)
```

UMAP of co-expression with Krt6a and Ly6d
```{r}
FeaturePlot(mouse_seurat_epi, features = c("Krt6a", "Ly6d"), order = T, min.cutoff = "q1",max.cutoff = "q99", blend = T, cols = c("light grey", "#219ebc", "#ffb703"), blend.threshold = 0, raster = TRUE) + NoAxes()

```

