---
title: "Hi-Myc Epithelial"
output: html_document
date: "2022-11-29"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Epithelial Subset
Take 10x genomics outs of mouse prostate 6 month libraries and generate an aggregated seurat object. Previously performed by Roshan, but repeated here to ensure reproducibility. 

```{r}
# Libraries
library(Seurat)
library(dplyr)
library(readr)
library(DataCombine)
library(viridis)
library(ggplot2)


# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

# load seurat object
seurat_epi <- readRDS(file = "mouse_seurat_epi.rds")
```

# GSEA luminal vs luminal MYC 1
The MYC1 cluster expresses the most MYC, and appears to be a Hi-Myc associated cluster enriched in dorsal and lateral lobe. Likely to be the source of cells that are PIN and invasive carcinoma
```{r}
library(msigdbr)
library(presto)
library(fgsea)
library(DataCombine)
library(tidyverse)

# select hallmark geneset
hallmark_gene_sets <- msigdbr(species = "mouse", category = "H")
fgsea_sets<- hallmark_gene_sets %>% split(x = .$gene_symbol, f = .$gs_name)
```

Luminal MYC 1 vs Luminal
```{r}
# subset to luminal cells
Idents(seurat_epi) <- "epi_clusters_noLobe"
levels(seurat_epi)

luminalMYC1genes <- wilcoxauc(X = seurat_epi, group_by = 'epi_clusters_noLobe', seurat_assay = 'SCT', assay = 'data',
                               groups_use = c("Luminal", "Luminal MYC 1"))
head(luminalMYC1genes)

# we have all the genes for each cluster
dplyr::count(luminalMYC1genes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
luminalMYC1genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

luminalMYC1genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
luminalMYC1genes<- luminalMYC1genes %>%
  dplyr::filter(group == "Luminal MYC 1") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_luminalMYC1<- deframe(luminalMYC1genes)
head(ranks_luminalMYC1)
tail(ranks_luminalMYC1)

# perform analysis
fgseaRes_luminalMYC1 <- fgsea(fgsea_sets, stats = ranks_luminalMYC1)

# save results
# saveRDS(fgseaRes_luminalMYC1, file = "fgseaRes_LuminalMYC1.rds")

# tidy the data
fgseaResTidy_luminalMYC1 <- fgseaRes_luminalMYC1 %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_luminalMYC1$statistic <- -log10(fgseaResTidy_luminalMYC1$padj)

# add comparison
fgseaResTidy_luminalMYC1$comparison <- "Luminal MYC 1 vs Luminal"

# only plot the top pathways
LuminalMYC1_GSEA <- fgseaResTidy_luminalMYC1 %>% top_n(15, -padj) %>%
ggplot(aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Top 15 GSEA Hallmark Pathways \nLuminal MYC 1 vs Luminal Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 

LuminalMYC1_GSEA
```

# GSEA Basal vs Reactive Basal
Basal vs Reactive Basal
```{r}
# subset to luminal cells
Idents(seurat_epi) <- "epi_clusters_noLobe"
levels(seurat_epi)

BasalLy6dgenes <- wilcoxauc(X = seurat_epi, group_by = 'epi_clusters_noLobe', seurat_assay = 'SCT', assay = 'data',
                               groups_use = c("Basal", "Reactive Basal"))
head(BasalLy6dgenes)

# we have all the genes for each cluster
dplyr::count(BasalLy6dgenes, group)

# check of the gene-level statistic following wilcox rank sum test and auROC analysis 
BasalLy6dgenes %>%
  dplyr::filter(group == "Reactive Basal") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  head(n = 10)

BasalLy6dgenes %>%
  dplyr::filter(group == "Reactive Basal") %>%
  arrange(desc(logFC), desc(statistic)) %>%
  tail(n = 10)

# select only the feature and statistic columns for fgsea
BasalLy6dgenes<- BasalLy6dgenes %>%
  dplyr::filter(group == "Reactive Basal") %>%
  arrange(desc(statistic)) %>% 
  dplyr::select(feature, statistic)

ranks_BasalLy6d<- deframe(BasalLy6dgenes)
head(ranks_BasalLy6d)
tail(ranks_BasalLy6d)

# perform analysis
fgseaRes_BasalLy6d <- fgsea(fgsea_sets, stats = ranks_BasalLy6d)

# save results
# saveRDS(fgseaRes_BasalLy6d, file = "fgseaRes_BasalLy6d.rds")

# tidy the data
fgseaResTidy_BasalLy6d <- fgseaRes_BasalLy6d %>%
  as_tibble() %>%
  arrange(padj)

# add statistic
fgseaResTidy_BasalLy6d$statistic <- -log10(fgseaResTidy_BasalLy6d$padj)

# add comparison
fgseaResTidy_BasalLy6d$comparison <- "Reactive Basal vs Luminal"

# only plot the top pathways
BasalLy6d_GSEA <- fgseaResTidy_BasalLy6d %>% top_n(15, -padj) %>%
ggplot(aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= statistic)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Reactive Basal vs Basal Clusters") + 
  theme_minimal() + labs(fill = expression("-Log"[10]*"(adjusted p-value)")) 

BasalLy6d_GSEA
```

