---
title: "Human Prostate Stroma"
author: "Mindy Kim Graham"
date: "4/21/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Human Subjects Aggregated Prostatectomy Study: Stroma
```{r}
# Libraries
library(Seurat)
library(ggplot2)
library(sctransform)
library(glmGamPoi)
library(RColorBrewer)
library(ggsci)
library(viridis)
library(dplyr)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)
```

load data
```
# load seurat object
seurat_object <- readRDS(file = "seurat_prostatectomy.rds")

Idents(seurat_object) <- "cell_type"
DimPlot(seurat_object)
```

# subset to stroma cells, sctransform, dimensionality reduction
refer to https://satijalab.org/seurat/articles/sctransform_vignette.html
```

# subset for stroma cells
levels(seurat_object)
seurat_stroma <- subset(seurat_object, idents = c("Endothelial", "Fibroblasts", "Pericytes", "Smooth Muscle"))
DimPlot(seurat_stroma)

# sctransform
seurat_stroma <- SCTransform(seurat_stroma, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

# dimensionality reduction
seurat_stroma <- RunPCA(seurat_stroma, verbose = FALSE)
seurat_stroma <- RunUMAP(seurat_stroma, reduction = "pca", dims = 1:40)
seurat_stroma <- FindNeighbors(seurat_stroma, dims = 1:40, verbose = FALSE)

DimPlot(seurat_stroma, group.by = "cell_type")
DimPlot(seurat_stroma, group.by = "Subject", shuffle = TRUE)

# cluster resolution
seurat_stroma <- FindClusters(seurat_stroma, verbose = FALSE, resolution = 0.2)
DimPlot(seurat_stroma, label = TRUE, group.by = "SCT_snn_res.0.2", shuffle = TRUE)

table(seurat_stroma$cell_type, seurat_stroma$SCT_snn_res.0.2)
```

cell type markers
0 Pericytes
1 Endothelial
2 Endothelial
3 Fibroblast
4 Mixed
5 Endothelial
6 Smooth Muscle
```

# stroma subtypes
Luminal <- c("KRT8", "KRT18")
Basal <- c("KRT5", "KRT14")
Fibroblast <- c("FBLN1", "DCN", "PTGDS")
Smooth_Muscle <- c("DES", "ACTG2", "ACTA2", "TAGLN", "MYL9")
Pericytes <- c("RGS5", "NOTCH3")
Endothelial <- c("CD93", "PECAM1")

celltype_genes <- c(Fibroblast, Smooth_Muscle, Pericytes, Endothelial, "TIMP1", "COL1A1", "COL1A2", "COL3A1", Luminal, Basal, "AMACR", "PCA3", "ERG", "FOLH1")

# heatmap
DoHeatmap(subset(seurat_stroma, downsample = 50), 
          features = celltype_genes) + 
  scale_fill_viridis()

# exclude cluster 4. Several cell marker genes expressed, including luminal and cancer markers
seurat_stroma_filter <- subset(seurat_stroma, idents = 4, invert = TRUE)
DimPlot(seurat_stroma_filter)
```

Dimensionality reduction and clustering analysis with mixed cluster filtered out
```
# sctransform
seurat_stroma_filter <- SCTransform(seurat_stroma_filter, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")

# dimensionality reduction
seurat_stroma_filter <- RunPCA(seurat_stroma_filter, verbose = FALSE)
seurat_stroma_filter <- RunUMAP(seurat_stroma_filter, reduction = "pca", dims = 1:50)
seurat_stroma_filter <- FindNeighbors(seurat_stroma_filter, dims = 1:50, verbose = FALSE)

DimPlot(seurat_stroma_filter, group.by = "cell_type")
DimPlot(seurat_stroma_filter, group.by = "Subject", shuffle = TRUE)

# cluster resolution
seurat_stroma_filter <- FindClusters(seurat_stroma_filter, verbose = FALSE, resolution = 0.3)
DimPlot(seurat_stroma_filter, label = TRUE, shuffle = TRUE)
```
cell type markers
```{r}
# stroma subtypes
Luminal <- c("KRT8", "KRT18")
Basal <- c("KRT5", "KRT14")
Fibroblast <- c("FBLN1", "DCN", "PTGDS")
Smooth_Muscle <- c("DES", "ACTG2", "ACTA2", "TAGLN", "MYL9")
Pericytes <- c("RGS5", "NOTCH3")
Endothelial <- c("CD93", "PECAM1")
Glial <- c("PLP1", "MPZ", "NRXN1")

celltype_genes <- c(Fibroblast, Smooth_Muscle, Pericytes, Endothelial, Glial, "TIMP1", "COL1A1", "COL1A2", "COL3A1")
```

```
# heatmap
DoHeatmap(subset(seurat_stroma_filter, downsample = 50), 
          features = celltype_genes) + 
  scale_fill_viridis()
```

Differential Gene Expression
```
seurat_DGE <- FindAllMarkers(seurat_stroma_filter)
top_stroma <- seurat_DGE %>% group_by(cluster) %>% filter(avg_log2FC > 0, pct.2 < 0.1) %>% top_n(5, -p_val_adj) %>% top_n(5, avg_log2FC)

DoHeatmap(subset(seurat_stroma_filter, downsample = 100), features = c(top_stroma$gene)) +
  scale_fill_viridis()
```

stromal cell type ID
0 Endothelial HMOX1
1 Pericytes
2 Endothelial
3 Pericytes INHBA
4 Endothelial
5 Fibroblast
6 Endothelial IGFBP3
7 Smooth Muscle
8 Endothelial HMOX1 CXCL8
9 Glial
```

# stroma
Idents(seurat_stroma_filter) <- "SCT_snn_res.0.3"

new.cluster.ids <- c("Endothelial",
                     "Pericytes", "Endothelial", "Pericytes", "Endothelial", "Fibroblast",
                     "Endothelial", "Smooth Muscle", "Endothelial", "Glial")

names(new.cluster.ids) <- levels(seurat_stroma_filter)
seurat_stroma_filter[["stroma"]] <- new.cluster.ids[as.numeric(as.matrix(seurat_stroma_filter@meta.data$SCT_snn_res.0.3))+1]
seurat_stroma_filter <- RenameIdents(seurat_stroma_filter, new.cluster.ids)
DimPlot(seurat_stroma_filter, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "stroma", shuffle = TRUE)

#stroma subset
Idents(seurat_stroma_filter) <- "SCT_snn_res.0.3"

new.cluster.ids <- c("Endothelial HMOX1",
                     "Pericytes", "Endothelial", "Pericytes INHBA", "Endothelial", "Fibroblast",
                     "Endothelial IGFBP3", "Smooth Muscle", "Endothelial HMOX1 CXCL8", "Glial")

names(new.cluster.ids) <- levels(seurat_stroma_filter)
seurat_stroma_filter[["stroma_subcluster"]] <- new.cluster.ids[as.numeric(as.matrix(seurat_stroma_filter@meta.data$SCT_snn_res.0.3))+1]
seurat_stroma_filter <- RenameIdents(seurat_stroma_filter, new.cluster.ids)
DimPlot(seurat_stroma_filter, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "stroma_subcluster", shuffle = TRUE)

saveRDS(seurat_stroma_filter, file = "Humanseurat_stroma.rds")
```

```{r}
# load stroma seurat object
seurat_stroma_filter <- readRDS(file = "Humanseurat_stroma.rds")

# Figure 1: Heatmap of cell type specific genes
Idents(seurat_stroma_filter) <- "stroma"
levels(seurat_stroma_filter) <- c("Fibroblast", "Smooth Muscle", 
                                  "Pericytes",
                                  "Endothelial",
                                  "Glial")

celltype_genes <- c(Fibroblast, Smooth_Muscle, Pericytes, Endothelial, Glial)

DoHeatmap(subset(seurat_stroma_filter, downsample = 50), features = celltype_genes) +
  scale_fill_viridis()

# UMAPs by cell type, patient, tissue source
DimPlot(seurat_stroma_filter, label = TRUE, raster = TRUE, label.size = 5, pt.size = 2) + ggtitle("Human Stromal Cells") + NoLegend() +
  theme(plot.title = element_text(hjust = 0.5))

DimPlot(seurat_stroma_filter, label = FALSE, group.by = "Tissue", shuffle = TRUE, raster = TRUE, pt.size = 2) + ggtitle("Tissue") +
  theme(plot.title = element_text(hjust = 0.5))
```

# Mouse Fibroblast Timp1 gene signature
https://www.r-bloggers.com/2016/10/converting-mouse-to-human-gene-names-with-biomart-package/
```
# load DGE of mouse stroma
ms_stroma_clusters <- readRDS(file = "seurat_stroma_DGE.rds")
top_stroma <- ms_stroma_clusters %>% group_by(cluster) %>% filter(avg_log2FC > 0, pct.2 < 0.3) %>% top_n(5, -p_val_adj) %>% top_n(10, avg_log2FC)
FibTimp1 <- top_stroma %>% filter(cluster %in% "Fibroblasts Timp1")

# Basic function to convert mouse to human gene names
convertMouseGeneList <- function(x){

require("biomaRt")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")

genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = x , mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
humanx <- genesV2

# Print the first 6 genes found to the screen
print(head(humanx))
return(humanx)
}

FibTimp1_human <- convertMouseGeneList(FibTimp1$gene)
FeaturePlot(seurat_stroma_filter, features = FibTimp1_human$HGNC.symbol, order = TRUE, ncol = 3)
saveRDS(FibTimp1_human, file = "FibTimp1_human.rds")
```

GeneSignature Score
```{r}
library(SeuratWrappers)
library(UCell)

# load converted genes
FibTimp1_human <- readRDS(file = "FibTimp1_human.rds")

GeneSig <- list()
GeneSig$FibTimp1 <- FibTimp1_human$HGNC.symbol
GeneSig

seurat_stroma_filter <- AddModuleScore_UCell(seurat_stroma_filter, features = GeneSig)
pal1 <- c("#3B9AB2", "#78B7C5", "#E1AF00", "#F21A00")
FeaturePlot(seurat_stroma_filter, reduction = "umap", features = "FibTimp1_UCell", order = TRUE, pt.size = 1)+ 
  scale_color_gradientn(colours = pal1) + ggtitle("Gene Signature: \nHi-Myc Fibroblast Timp1")+
  theme(plot.title = element_text(hjust = 0.5))

```

# save seurat object
```
saveRDS(seurat_stroma_filter, file = "Humanseurat_stroma.rds")
```

