---
title: "Mouse Stroma HiMyc by Strain"
output: html_document
date: "2023-01-10"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Hi-Myc mouse prostate stroma
The stromal clusters were of 6 month WT and Hi-Myc mice were subset and further evaluated to investigate strain specific differences of stromal cells in response to MYC driven prostate cancer

```{r}
# Libraries
library(Seurat)
library(dplyr)
library(readr)
library(DataCombine)
library(viridis)
library(ggplot2)
library(glmGamPoi)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)
```

# Subset and examine stromal cell clusters
sctransform seurat object,dimensionality reduction by PCA and UMAP embedding
```
# load seurat object
strain.combined.sct <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")
DimPlot(strain.combined.sct)

# subset to stromal cells
levels(strain.combined.sct)
stroma_seurat <- subset(strain.combined.sct, idents = c("Fibroblasts", "Endothelial", "Pericytes", "Smooth Muscle"))
DimPlot(stroma_seurat, group.by = "cell_type")
```
integrate data by strain
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
# split the dataset into a list by strain (FVB and C57BL/6)
stroma_seurat.list <- SplitObject(stroma_seurat, split.by = "strain")

FVB <- stroma_seurat.list[["FVB/NJ"]]
BL6 <- stroma_seurat.list[["C57BL/6J"]]

```

Perform integration using pearson residuals
```
# normalize and run dimensionality reduction on FVB libraries
FVB <- SCTransform(FVB, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
BL6 <- SCTransform(BL6, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)

stroma_seurat.list <- list(`FVB/NJ` = FVB, `C57BL/6J` = BL6)
features <- SelectIntegrationFeatures(object.list = stroma_seurat.list, nfeatures = 3000)
stroma_seurat.list <- PrepSCTIntegration(object.list = stroma_seurat.list, anchor.features = features)

# integrate datasets
strain.anchors <- FindIntegrationAnchors(object.list = stroma_seurat.list, normalization.method = "SCT",
    anchor.features = features)
stroma_seurat <- IntegrateData(anchorset = strain.anchors, normalization.method = "SCT")
```

Perform an integrated analysis
```
DefaultAssay(stroma_seurat) <- "integrated" 
stroma_seurat <- RunPCA(stroma_seurat, verbose = FALSE)
stroma_seurat <- RunUMAP(stroma_seurat, reduction = "pca", dims = 1:50, verbose = FALSE)
stroma_seurat <- FindNeighbors(stroma_seurat, reduction = "pca", dims = 1:50)
stroma_seurat <- FindClusters(stroma_seurat, resolution = 0.3)

P1 <- DimPlot(stroma_seurat, label = T, repel = T, shuffle = TRUE) + ggtitle("Unsupervised clustering")
P2 <- DimPlot(stroma_seurat, label = T, repel = T, shuffle = TRUE, group.by = "strain")
P3 <- DimPlot(stroma_seurat, label = T, repel = T, shuffle = TRUE, group.by = "genotype")

P1+P2+P3

table(stroma_seurat$cell_type, stroma_seurat$integrated_snn_res.0.3)
```

SCT transform without integration
Perform SCTransform for all genes, keeping integrating clustering
dimensionality reduction by PCA and UMAP embedding
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
library(glmGamPoi)

# run sctransform
stroma_seurat <- SCTransform(stroma_seurat, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")
```

seurat cluster DGE
```
seurat_cluster_markers <- FindAllMarkers(stroma_seurat)
top_seurat_markers <- seurat_cluster_markers %>% filter(avg_log2FC > 0) %>% group_by(cluster)%>% top_n(5, -p_val_adj) %>% top_n(5, avg_log2FC) 

DoHeatmap(subset(stroma_seurat, downsample = 50), assay = "SCT", features = c(top_seurat_markers$gene, "Tnfsf9"), group.by = "integrated_snn_res.0.3") +
  scale_fill_viridis()
```

Marker genes for each cell type
https://www.sciencedirect.com/science/article/pii/S1550413121002175 LECs
```{r}
# markers
sm_pericytes <- c("Tagln", "Acta2")
Smooth_Muscle <- c("Acta1", "Actg2")
Pericytes <- c("Rgs4", "Notch3")
Fibroblast <- c("Pdgfra", "Fbln1")
Fibroblast_subtypes <- c("Rorb", "Sult1e1", "Timp1")
Endothelial <- c("Flt1","Eng","Cdh5")
Glial <- c("Plp1", "Ngfr", "Mbp")
Epithelial <- c("Krt8", "Krt18")
lymphatic_endothelial <- c("Prox1", "Mmrn1", "Flt4", "Nts")
```

```
cell_markers <- c(Endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast, "Rorb", "Timp1", "Sult1e1", Glial, lymphatic_endothelial)

DoHeatmap(subset(stroma_seurat, downsample = 50), assay = "SCT", features = c(cell_markers), 
          group.by = "integrated_snn_res.0.3") +
  scale_fill_viridis()
```

Filter cluster 8
```
# exclude cluster 8, this is a mixed cluster
levels(stroma_seurat)
stroma_seurat_filter <- subset(stroma_seurat, idents = 8, invert = TRUE)
```

# Dimensionality reduction and clustering with filtered seurat object
integrate data by strain
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
# split the dataset into a list by strain (FVB and C57BL/6)
stroma_seurat_filter.list <- SplitObject(stroma_seurat_filter, split.by = "strain")

FVB <- stroma_seurat_filter.list[["FVB/NJ"]]
BL6 <- stroma_seurat_filter.list[["C57BL/6J"]]

```

Perform integration using pearson residuals
```
# normalize and run dimensionality reduction on FVB libraries
FVB <- SCTransform(FVB, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE)
BL6 <- SCTransform(BL6, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE)

stroma_seurat_filter.list <- list(`FVB/NJ` = FVB, `C57BL/6J` = BL6)
features <- SelectIntegrationFeatures(object.list = stroma_seurat_filter.list, nfeatures = 3000)
stroma_seurat_filter.list <- PrepSCTIntegration(object.list = stroma_seurat_filter.list, anchor.features = features)

# integrate datasets
strain.anchors <- FindIntegrationAnchors(object.list = stroma_seurat_filter.list, normalization.method = "SCT",
                                         anchor.features = features)
stroma_seurat_filter <- IntegrateData(anchorset = strain.anchors, normalization.method = "SCT")
```

Perform an integrated analysis
```
DefaultAssay(stroma_seurat_filter) <- "integrated" 
stroma_seurat_filter <- RunPCA(stroma_seurat_filter, verbose = FALSE)
stroma_seurat_filter <- RunUMAP(stroma_seurat_filter, reduction = "pca", dims = 1:50, verbose = FALSE)
stroma_seurat_filter <- FindNeighbors(stroma_seurat_filter, reduction = "pca", dims = 1:50)
stroma_seurat_filter <- FindClusters(stroma_seurat_filter, resolution = 0.1)

P1 <- DimPlot(stroma_seurat_filter, label = T, repel = T, shuffle = TRUE) + ggtitle("Unsupervised clustering")
P2 <- DimPlot(stroma_seurat_filter, label = T, repel = T, shuffle = TRUE, group.by = "strain")
P3 <- DimPlot(stroma_seurat_filter, label = T, repel = T, shuffle = TRUE, group.by = "genotype")

P1+P2+P3

table(stroma_seurat_filter$cell_type, stroma_seurat_filter$integrated_snn_res.0.1)
```

SCT transform without integration
Perform SCTransform for all genes, keeping integrating clustering
dimensionality reduction by PCA and UMAP embedding
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
library(glmGamPoi)

# run sctransform
stroma_seurat_filter <- SCTransform(stroma_seurat_filter, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")
```

seurat cluster DGE
```
seurat_cluster_markers <- FindAllMarkers(stroma_seurat_filter)
top_seurat_markers <- seurat_cluster_markers %>% filter(avg_log2FC > 0) %>% group_by(cluster)%>% top_n(5, -p_val_adj) %>% top_n(5, avg_log2FC) 

DoHeatmap(subset(stroma_seurat_filter, downsample = 50), assay = "SCT", features = c(top_seurat_markers$gene), group.by = "integrated_snn_res.0.1") +
  scale_fill_viridis()
```

Marker genes for each cell type
https://www.sciencedirect.com/science/article/pii/S1550413121002175 LECs
```
DefaultAssay(stroma_seurat_filter) <- "SCT" 

cell_markers <- c(Endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast, "Rorb", "Timp1", "Sult1e1", Glial, lymphatic_endothelial, "Tnfsf9", "Selp")

DoHeatmap(subset(stroma_seurat_filter, downsample = 50), assay = "SCT", features = c(cell_markers), 
          group.by = "integrated_snn_res.0.1") +
  scale_fill_viridis()
```

Update cluster names
0 - Fibroblasts Subglandular
1 - Endothelial
2 - Fibroblast Timp1
3 - Smooth Muscle
4 - Endothelial Selp
5 - Pericytes
6 - Fibroblasts Interstitial
7 - Fibroblasts Tnfrsf9
8 - Lymphatic endothelial
9 - Glial

```
# previous vs new clustering
table(stroma_seurat_filter$integrated_snn_res.0.1, stroma_seurat_filter$genotype)
table(stroma_seurat_filter$integrated_snn_res.0.1, stroma_seurat_filter$strain)

# update clustering name
Idents(stroma_seurat_filter) <- "integrated_snn_res.0.1"
new.cluster.ids <- c("Fibroblasts Subglandular", 
                     "Endothelial", "Fibroblasts Timp1", "Smooth Muscle", "Endothelial Selp", "Pericytes",
                     "Fibroblasts Interstitial", "Fibroblasts Tgif2", "Lymphatic endothelial", "Glial")

names(new.cluster.ids) <- levels(stroma_seurat_filter)
stroma_seurat_filter[["stroma"]] <- new.cluster.ids[as.numeric(as.matrix(stroma_seurat_filter@meta.data$integrated_snn_res.0.1))+1]
stroma_seurat_filter <- RenameIdents(stroma_seurat_filter, new.cluster.ids)

DimPlot(stroma_seurat_filter, group.by = "stroma")
saveRDS(stroma_seurat_filter, file = "stroma_seurat_strain.rds")
```

UMAPs and Heatmaps of integrated stromal clusters
```{r}
# load seurat object
stroma_seurat_filter <- readRDS(file = "stroma_seurat_strain.rds")

Idents(stroma_seurat_filter) <- "stroma"
levels(stroma_seurat_filter) <- c("Endothelial", "Endothelial Selp", "Lymphatic endothelial",
                           "Smooth Muscle", "Pericytes",
                           "Fibroblasts Subglandular", "Fibroblasts Timp1", "Fibroblasts Tgif2","Fibroblasts Interstitial",
                           "Glial")

cell_markers <- c(Endothelial, "Selp", lymphatic_endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast, "Rorb", "Timp1", "Tgif2", "Sult1e1", Glial)

DoHeatmap(subset(stroma_seurat_filter, downsample = 50), assay = "SCT", features = cell_markers) +
  scale_fill_viridis()

DimPlot(stroma_seurat_filter, label = TRUE, raster = TRUE, label.size = 4, pt.size = 2) + NoLegend() + ggtitle("Stroma Mouse Clusters")+
  theme(plot.title = element_text(hjust = 0.5))

DimPlot(stroma_seurat_filter, label = FALSE, raster = TRUE, label.size = 5, pt.size = 2,
                  group.by = "strain", cols =  c("#219ebc", "#ffb703"), shuffle = TRUE) + ggtitle("Strain")+
  theme(plot.title = element_text(hjust = 0.5))

DimPlot(stroma_seurat_filter, label = FALSE, raster = TRUE, label.size = 5, pt.size = 2,
        group.by = "genotype", shuffle = TRUE) + ggtitle("Genotype")+
  theme(plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values = c("#a8dadc", "#e63946")) 

```


#Cell proportions 
test by RAISIN
https://github.com/zji90/raisin//
  cell proportion by genotype
```
# add stroma metadata to complete dataset
strain.combined.sct$stroma <- strain.combined.sct$cell_type
strain.combined.sct$stroma <- gsub("Fibroblasts", "", strain.combined.sct$stroma)
strain.combined.sct$stroma <- gsub("Endothelial", "", strain.combined.sct$stroma)
strain.combined.sct$stroma <- gsub("Pericytes", "", strain.combined.sct$stroma)
strain.combined.sct$stroma <- gsub("Smooth Muscle", "", strain.combined.sct$stroma)
strain.combined.sct$stroma_subset <- stroma_seurat_filter$stroma
strain.combined.sct$stroma <- paste(strain.combined.sct$stroma, strain.combined.sct$stroma_subset, sep = "")
strain.combined.sct$stroma <- gsub("NA", "", strain.combined.sct$stroma)

Idents(strain.combined.sct) <- "stroma"
levels(strain.combined.sct)
table(strain.combined.sct$stroma)
```

```{r}
library(raisin)

# load seurat object
strain.combined.sct <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")
Idents(strain.combined.sct) <- "stroma"
levels(strain.combined.sct)

# extract metadata from seurat object
Idents(strain.combined.sct) <- "stroma"
levels(strain.combined.sct)
cluster <- Idents(strain.combined.sct)

Idents(strain.combined.sct) <- "library_id"
levels(strain.combined.sct)
individual <- Idents(strain.combined.sct)

# make the design 
design <- data.frame(intercept=1,contrast=c(0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1),
                     row.names = c("AB1", "AB2", "AF1", "AF2", "AHMB1",
                                   "AHMB2", "AHMB3", "AHMF1", "AHMF2",
                                   "DB1", "DB2", "DF1", "DF2", "DHMB1",
                                   "DHMB2", "DHMB3", "DHMF1", "DHMF2",
                                   "LB1", "LB2", "LF1", "LF2", "LHMB1",
                                   "LHMB2", "LHMB3", "LHMF1", "LHMF2",
                                   "VB1", "VB2", "VF1", "VF2", "VHMB1",
                                   "VHMB2", "VHMB3", "VHMF1", "VHMF2"))

# run the test
prop_test_results_genotype <- proportiontest(cluster,individual,design)
prop_test_results_genotype <- prop_test_results_genotype %>% tibble::rownames_to_column("Cluster")
colnames(prop_test_results_genotype) <- c("Cluster", "Genotype logFC", "Genotype P.Value", "Genotype adj.P.Val")
prop_test_results_genotype

```

Cell proportions test by RAISIN
https://github.com/zji90/raisin//
  cell proportion by strain
```{r}

# extract metadata from seurat object
Idents(strain.combined.sct) <- "stroma"
levels(strain.combined.sct)
cluster <- Idents(strain.combined.sct)

Idents(strain.combined.sct) <- "library_id"
levels(strain.combined.sct)
individual <- Idents(strain.combined.sct)

# make the design 
design <- data.frame(intercept=1,contrast=c(0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1,
                                            0,0,0,0,0,
                                            1,1,1,1),
                     row.names = c("AB1", "AB2", "AHMB1","AHMB2", "AHMB3",
                                   "AF1", "AF2", "AHMF1", "AHMF2",
                                   "DB1", "DB2", "DHMB1", "DHMB2", "DHMB3", 
                                   "DF1", "DF2", "DHMF1", "DHMF2",
                                   "LB1", "LB2", "LHMB1","LHMB2", "LHMB3",
                                   "LF1", "LF2", "LHMF1", "LHMF2",
                                   "VB1", "VB2", "VHMB1","VHMB2", "VHMB3",
                                   "VF1", "VF2", "VHMF1", "VHMF2"))

# run the test
prop_test_results_strain <- proportiontest(cluster,individual,design)
prop_test_results_strain <- prop_test_results_strain %>% tibble::rownames_to_column("Cluster")
colnames(prop_test_results_strain) <- c("Cluster", "Strain logFC", "Strain P.Value", "Strain adj.P.Val")

prop_test_results_strain
```

Plot cell proportion strain differences of stroma cells
```{r}
library(ggpubr)
library_counts <- as.data.frame(table(strain.combined.sct$orig.ident))

# create dataframe containing all relevant metadata
cell_type_enrich <- as.data.frame(table(strain.combined.sct$library_id,strain.combined.sct$strain, strain.combined.sct$genotype, strain.combined.sct$stroma))

cell_type_enrich$library <- library_counts$Freq
cell_type_enrich$fraction <- cell_type_enrich$Freq / cell_type_enrich$library

# rename columns to meaningful column names
colnames(cell_type_enrich) <- c("Library", "Strain", "Genotype", "Cluster", "Counts", "Library_Counts", "Fraction")

# filter to library
AB1 <- cell_type_enrich %>% filter(Library == "AB1", Strain == "C57BL/6J", Genotype == "WT")
AB2 <- cell_type_enrich %>% filter(Library == "AB2", Strain == "C57BL/6J", Genotype == "WT")
AF1 <- cell_type_enrich %>% filter(Library == "AF1", Strain == "FVB/NJ", Genotype == "WT")
AF2 <- cell_type_enrich %>% filter(Library == "AF2", Strain == "FVB/NJ", Genotype == "WT")
AHMB1 <- cell_type_enrich %>% filter(Library == "AHMB1", Strain == "C57BL/6J", Genotype == "WT")
AHMB2 <- cell_type_enrich %>% filter(Library == "AHMB2", Strain == "C57BL/6J", Genotype == "MYC")
AHMB3 <- cell_type_enrich %>% filter(Library == "AHMB3", Strain == "C57BL/6J", Genotype == "MYC")
AHMF1 <- cell_type_enrich %>% filter(Library == "AHMF1", Strain == "FVB/NJ", Genotype == "MYC")
AHMF2 <- cell_type_enrich %>% filter(Library == "AHMF2", Strain == "FVB/NJ", Genotype == "MYC")
DB1 <- cell_type_enrich %>% filter(Library == "DB1", Strain == "C57BL/6J", Genotype == "WT")
DB2 <- cell_type_enrich %>% filter(Library == "DB2", Strain == "C57BL/6J", Genotype == "WT")
DF1 <- cell_type_enrich %>% filter(Library == "DF1", Strain == "FVB/NJ", Genotype == "WT")
DF2 <- cell_type_enrich %>% filter(Library == "DF2", Strain == "FVB/NJ", Genotype == "WT")
DHMB1 <- cell_type_enrich %>% filter(Library == "DHMB1", Strain == "C57BL/6J", Genotype == "WT")
DHMB2 <- cell_type_enrich %>% filter(Library == "DHMB2", Strain == "C57BL/6J", Genotype == "MYC")
DHMB3 <- cell_type_enrich %>% filter(Library == "DHMB3", Strain == "C57BL/6J", Genotype == "MYC")
DHMF1 <- cell_type_enrich %>% filter(Library == "DHMF1", Strain == "FVB/NJ", Genotype == "MYC")
DHMF2 <- cell_type_enrich %>% filter(Library == "DHMF2", Strain == "FVB/NJ", Genotype == "MYC")
LB1 <- cell_type_enrich %>% filter(Library == "LB1", Strain == "C57BL/6J", Genotype == "WT")
LB2 <- cell_type_enrich %>% filter(Library == "LB2", Strain == "C57BL/6J", Genotype == "WT")
LF1 <- cell_type_enrich %>% filter(Library == "LF1", Strain == "FVB/NJ", Genotype == "WT")
LF2 <- cell_type_enrich %>% filter(Library == "LF2", Strain == "FVB/NJ", Genotype == "WT")
LHMB1 <- cell_type_enrich %>% filter(Library == "LHMB1", Strain == "C57BL/6J", Genotype == "WT")
LHMB2 <- cell_type_enrich %>% filter(Library == "LHMB2", Strain == "C57BL/6J", Genotype == "MYC")
LHMB3 <- cell_type_enrich %>% filter(Library == "LHMB3", Strain == "C57BL/6J", Genotype == "MYC")
LHMF1 <- cell_type_enrich %>% filter(Library == "LHMF1", Strain == "FVB/NJ", Genotype == "MYC")
LHMF2 <- cell_type_enrich %>% filter(Library == "LHMF2", Strain == "FVB/NJ", Genotype == "MYC")
VB1 <- cell_type_enrich %>% filter(Library == "VB1", Strain == "C57BL/6J", Genotype == "WT")
VB2 <- cell_type_enrich %>% filter(Library == "VB2", Strain == "C57BL/6J", Genotype == "WT")
VF1 <- cell_type_enrich %>% filter(Library == "VF1", Strain == "FVB/NJ", Genotype == "WT")
VF2 <- cell_type_enrich %>% filter(Library == "VF2", Strain == "FVB/NJ", Genotype == "WT")
VHMB1 <- cell_type_enrich %>% filter(Library == "VHMB1", Strain == "C57BL/6J", Genotype == "WT")
VHMB2 <- cell_type_enrich %>% filter(Library == "VHMB2", Strain == "C57BL/6J", Genotype == "MYC")
VHMB3 <- cell_type_enrich %>% filter(Library == "VHMB3", Strain == "C57BL/6J", Genotype == "MYC")
VHMF1 <- cell_type_enrich %>% filter(Library == "VHMF1", Strain == "FVB/NJ", Genotype == "MYC")
VHMF2 <- cell_type_enrich %>% filter(Library == "VHMF2", Strain == "FVB/NJ", Genotype == "MYC")

filtered_library_fraction <- rbind(AB1, AB2, AF1, AF2,
                                   AHMB1, AHMB2, AHMB3, AHMF1, AHMF2,
                                   DB1, DB2, DF1, DF2,
                                   DHMB1, DHMB2, DHMB3, DHMF1, DHMF2,
                                   LB1, LB2, LF1, LF2, 
                                   LHMB1, LHMB2, LHMB3, LHMF1, LHMF2,
                                   VB1, VB2, VF1, VF2,
                                   VHMB1, VHMB2, VHMB3, VHMF1, VHMF2)

# combine DF with cell proportion test results
combined_df <- merge(filtered_library_fraction, prop_test_results_genotype, by.x = "Cluster", by.y = "Cluster")
combined_df <- merge(combined_df, prop_test_results_strain, by.x = "Cluster", by.y = "Cluster")

# Add pvalue to cluster name
library(scales)
combined_df$strain_adjP <- paste(combined_df$Cluster,
                                 "\n FVB vs C57BL6",
                                 "\n logFC = ", 
                                 signif(combined_df$`Strain logFC`, 2),
                                 "\nadj adj p-value = ", scientific(combined_df$`Strain adj.P.Val`, 1))

combined_df$genotype_adjP <- paste(combined_df$Cluster,
                                   "\n Hi-Myc vs WT",
                                   "\n logFC = ", 
                                   signif(combined_df$`Genotype logFC`, 2),
                                   "\nadj p-value = ", scientific(combined_df$`Genotype P.Value`, 1))

# order x axis
combined_df$Genotype <- factor(combined_df$Genotype, levels = c("WT",  "MYC"))

# Plot stroma clusters by strain
Strain_sig <- prop_test_results_strain %>% filter(`Strain adj.P.Val` < 0.05) %>% select(Cluster) 
Strain_sig

stroma_clusters <- c("Endothelial", "Endothelial Selp", "Lymphatic endothelial",
                           "Smooth Muscle", "Pericytes",
                           "Fibroblasts Subglandular", "Fibroblasts Timp1", "Fibroblasts Tgif2","Fibroblasts Interstitial",
                           "Glial")

combined_df %>% 
  filter(Cluster %in% c("Endothelial", "Smooth Muscle")) %>%
  ggplot(aes(x = `Strain`, y = Fraction))+
  geom_jitter(aes(fill = Genotype, color = Genotype), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Stromal Clusters with Significant Difference by Strain") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~strain_adjP, scales = "free", ncol = 2) +
  scale_color_manual(values = c("#a8dadc", "#e63946")) 


# Plot stroma clusters by genotype
Strain_genotype <- prop_test_results_genotype %>% filter(`Genotype adj.P.Val` < 0.05) %>% select(Cluster) 
Strain_genotype

combined_df %>% 
  filter(Cluster %in% c("Fibroblasts Timp1", "Endothelial", "Fibroblasts Interstitial", "Glial")) %>%
  ggplot(aes(x = Genotype, y = Fraction))+
  geom_jitter(aes(fill = Strain, color = Strain), size = 3, 
              width = 0.3) +
  theme_classic() +
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width = 0.5) +
  ggtitle("Stromal Clusters with Significant Differences by Genotype") +
  theme(plot.title = element_text(hjust = 0.5))+
  facet_wrap(~genotype_adjP, scales = "free", ncol = 2) +
  scale_color_manual(values = c("#219ebc", "#ffb703"))  

```

#Differential expression 
Find differentially expressed genes for each cluster
```
# differential gene expression by cluster
Idents(stroma_seurat_filter) <- "stroma"
levels(stroma_seurat_filter) <- stroma_clusters 

seurat_stroma_genes <- FindAllMarkers(stroma_seurat_filter)
saveRDS(seurat_stroma_genes, file = "seurat_stroma_genes.rds")
write.csv(seurat_stroma_genes, file= "seurat_stroma_genes.csv")
```
# save seurat object
```
saveRDS(stroma_seurat_filter, file = "stroma_seurat_filter.rds")
```
