---
title: "Mouse Prostate scRNA-seq Dataset"
output: html_document
date: "2024-01-29"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Generate seurat object
Take 10x genomics outs of mouse prostate 6 month libraries and generate an aggregated seurat object. 
FASTQ and gene x cell count matrices available on NCBI GEO (GSE228945)
```{r}
# Libraries
library(Seurat)
library(dplyr)
library(readr)
library(tidyverse)
library(DataCombine)
library(viridis)
library(ggplot2)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)
```

# Load 10 X genomics data from multiple subjects, create seurat object
```
# Load aggregated count matrices
mouse_wt_myc <- Read10X(data.dir = "C:/Users/mgraha21/OneDrive - Johns Hopkins/Prostate Cancer Seurat/Mouse HiMYC WT/Mouse WT HiMYC counts/mouse_wt_myc/")
```

# Import Counts for Mouse & Human MYC
Function to import MYC aligned counts originally written by co-author Roshan Chikarmane:
Only cells with counts for any type of MYC (mouse or human) are included in imported counts, which is why there are fewer cells in the imported myc_counts than in the full dataset. Cells can be matched with `rownames(myc_counts)` in the imported myc_counts data and `mouse_wt_myc@meta.data$barcode` in the Seurat object.

```
sample_number <- colnames(mouse_wt_myc)
sample_number <- as.numeric(sapply(strsplit(sample_number, "-"), "[[", 2))
table(sample_number)

# Function for MYC imports
myc_import <- function(lib_id){
  myc_counts <- Read10X(data.dir = paste0("C:/Users/mgraha21/OneDrive - Johns Hopkins/Prostate Cancer Seurat/Mouse HiMYC WT/Myc MYC from Anuj/myc_analysis/myc_analysis/", lib_id, "_myc_featurecounts"))
  rownames(myc_counts) <- c("Myc", "MYC")
  colnames(myc_counts) <- paste(colnames(myc_counts), lib_id, sep = "-")
  return(myc_counts)
}

# Loop through all Library IDs
for (id in 1:36){
  assign(paste0(id, "_myc_counts"), myc_import(id))
}

# cbind all of the counts tables into a one table
myc_counts <- do.call(cbind, mget(paste0(1:36, "_myc_counts")))
colnames(myc_counts)

# Remove all of the myc_counts objects for individual library_ids
rm(list=ls(pattern="*_myc_counts"))

print(paste(ncol(myc_counts), " cells in myc_counts, ", nrow(mouse_wt_myc), " cells in the full dataset."))


```

## Merge matrices
Function from co-author Bulouere Wodu:
Merging the human and mice MYC gene alignments. 
matrix1 = matrix with other genes aligned except the human MYC gene
matrix2 = matrix with only the human and mouse genes aligned
```

merge_matrix <- function(matrix1, matrix2) {
  # create a vector with 0s and add it as a new row to matrix1. This new row will hold the data for the human MYC gene
  MYC <- rep(0, times = ncol(matrix1))
  matrix1 <- rbind(matrix1, MYC)
  
  # get the cells present in both matrices
  cells_myc <- intersect(colnames(matrix1), colnames(matrix2))
  
  # replace the 0 in the MYC row with their actual values from matrix2. Only the data for cells present in matrix1 and matrix2 will be replaced.
  matrix1[c("MYC"), cells_myc] <- matrix2[c("MYC"), cells_myc]
  
  # return the merged matrix
  return(matrix1)
}

# merge matrices
counts_matrix_merge <- merge_matrix(mouse_wt_myc, myc_counts)

```
# add Metadata
```
# Initialize the Seurat object
seurat_object <- CreateSeuratObject(counts = counts_matrix_merge, project = "seurat_object", min.cells = 3, min.features = 200)

# retrieve sample/library number
sample_number <- rownames(seurat_object@meta.data)
sample_number <- as.numeric(sapply(strsplit(sample_number, "-"), "[[", 2))
table(sample_number)

# load sample ID information from aggregation of all mouse prostate libraries
library(readr)
aggr_wt_myc <- read_csv("aggr_wt_myc.csv")

# add sample id to seurat object
library_ID <- aggr_wt_myc$library_id
seurat_object$library_id <- library_ID[sample_number]
table(seurat_object$library_id)
table(sample_number)

# add strain ID to seurat object
seurat_object$strain <- "FVB/NJ"
seurat_object$strain[grepl("B", seurat_object$library_id)] <- "C57BL/6J"
table(seurat_object$library_id, seurat_object$strain)

# add lobe ID to seurat object
seurat_object$lobe <- "Anterior"
seurat_object$lobe[grepl("D", seurat_object$library_id)] <- "Dorsal"
seurat_object$lobe[grepl("L", seurat_object$library_id)] <- "Lateral"
seurat_object$lobe[grepl("V", seurat_object$library_id)] <- "Ventral"
table(seurat_object$library_id, seurat_object$lobe)

# add genotype to seurat object
seurat_object$genotype <- "WT"
seurat_object$genotype[grepl("HM", seurat_object$library_id)] <- "MYC"
seurat_object$genotype[grepl("HMB1", seurat_object$library_id)] <- "WT" #mouse HMB1 was genotyped as WT
table(seurat_object$library_id, seurat_object$genotype)

# add genotype and lobe
seurat_object$genotype_lobe <- paste(seurat_object$genotype, seurat_object$lobe)

# add strain and lobe
seurat_object$strain_lobe <- paste(seurat_object$strain, seurat_object$lobe)

# add strain genotype and lobe
seurat_object$strain_genotype_lobe <- paste(seurat_object$strain, seurat_object$genotype_lobe, sep = " ")

```

# QC metrics
```
# Add number of genes per UMI for each cell to metadata
seurat_object$log10GenesPerUMI <- log10(seurat_object$nFeature_RNA) / log10(seurat_object$nCount_RNA)

# Compute percent mito ratio
seurat_object$mitoRatio <- PercentageFeatureSet(object = seurat_object, pattern = "^mt-")
seurat_object$mitoRatio <- seurat_object@meta.data$mitoRatio / 100

summary(seurat_object$mitoRatio)

# mito ratio quantiles
library(stats)
quantile(seurat_object$mitoRatio, probs = c(0, 0.5, 0.8, 0.83, 1))

# Filter out low quality reads using selected thresholds, > 83% of cells retained 
seurat_object <- subset(x = seurat_object,
                           (mitoRatio < 0.25))

summary(seurat_object$mitoRatio)
```

# integrate data by strain
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
# split the dataset into a list by strain (BL6 and FVB)
seurat_object.list <- SplitObject(seurat_object, split.by = "strain")

FVB <- seurat_object.list[["FVB/NJ"]]
BL6 <- seurat_object.list[["C57BL/6J"]]

```

Perform normalization and dimensionality reduction
```
# normalize and run dimensionality reduction on FVB libraries
FVB <- SCTransform(FVB, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE)
FVB <- RunPCA(FVB, npcs = 30, verbose = FALSE)
FVB <- RunUMAP(FVB, reduction = "pca", dims = 1:30, verbose = FALSE)
FVB <- FindNeighbors(FVB, reduction = "pca", dims = 1:30, verbose = FALSE)
FVB <- FindClusters(FVB, resolution = 0.6, verbose = FALSE)

```

Perform integration using pearson residuals
```

BL6 <- SCTransform(BL6, vst.flavor = "v2", verbose = FALSE, return.only.var.genes = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
seurat_object.list <- list(FVB = FVB, BL6 = BL6)
features <- SelectIntegrationFeatures(object.list = seurat_object.list, nfeatures = 3000)
seurat_object.list <- PrepSCTIntegration(object.list = seurat_object.list, anchor.features = features)

# integrate datasets
strain.anchors <- FindIntegrationAnchors(object.list = seurat_object.list, normalization.method = "SCT",
    anchor.features = features)
strain.combined.sct <- IntegrateData(anchorset = strain.anchors, normalization.method = "SCT")
```

Perform an integrated analysis
```
strain.combined.sct <- RunPCA(strain.combined.sct, verbose = FALSE)
strain.combined.sct <- RunUMAP(strain.combined.sct, reduction = "pca", dims = 1:30, verbose = FALSE)
strain.combined.sct <- FindNeighbors(strain.combined.sct, reduction = "pca", dims = 1:30)
strain.combined.sct <- FindClusters(strain.combined.sct, resolution = 0.5)

```
# SCT transform without integration
Perform SCTransform for all genes, keeping integrated clustering
Perform sctransform, dimensionality reduction by PCA and UMAP embedding
https://satijalab.org/seurat/articles/sctransform_v2_vignette.html
```
library(glmGamPoi)

# run sctransform
strain.combined.sct <- SCTransform(strain.combined.sct, method = "glmGamPoi", verbose = FALSE, return.only.var.genes = FALSE, vst.flavor = "v2")
```

# cell type ID

Marker genes for each cell type
```{r}
Basal <- c("Krt5", "Krt15", "Krt14")
Epithelial <- c("Krt8", "Krt18")
Luminal <- "Agr2"
Foxi1 <- c("Foxi1")
sm_pericytes <- c("Tagln", "Acta2")
Smooth_Muscle <- c("Acta1", "Actg2")
Pericytes <- c("Rgs4", "Notch3")
Fibroblast <- c("Pdgfra", "Fbln1")
Endothelial <- c("Eng","Flt1","Cdh5")
Macrophages <- c("C1qb","C1qa","C1qc","Cd68")
Immune <- "Ptprc"
Mast_cells <- c("Mcemp1", "Flt3")
Tcells <- c("Cd2","Cd3d", "Cd3g")

cell_markers <- c(Epithelial, Basal, Luminal,Foxi1,  
                  Endothelial, Smooth_Muscle, sm_pericytes, Pericytes, Fibroblast,
                  Immune,Macrophages,  Mast_cells, Tcells
                  )

```

check expression across clusters
```
library(patchwork)
DoHeatmap(subset(strain.combined.sct, downsample = 100), assay = "SCT", features = cell_markers, group.by = "integrated_snn_res.0.5") +
  scale_fill_viridis()

# cell type ID
Idents(strain.combined.sct) <- "integrated_snn_res.0.5"
new.cluster.ids <- c("Basal",
                     "Luminal", "Basal", "Luminal", "Luminal", "Luminal",
                     "Luminal", "Luminal", "Luminal", "Luminal", "Fibroblasts",
                     "Macrophages", "Luminal", "Endothelial", "Luminal", "Luminal",
                     "Luminal", "Luminal", "Basal", "Smooth Muscle", "Pericytes",
                     "Luminal", "Mast cells", "T cells", "Ionocytes")

names(new.cluster.ids) <- levels(strain.combined.sct)
strain.combined.sct[["cell_type"]] <- new.cluster.ids[as.numeric(as.matrix(strain.combined.sct@meta.data$integrated_snn_res.0.5))+1]
strain.combined.sct <- RenameIdents(strain.combined.sct, new.cluster.ids)
DimPlot(strain.combined.sct, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "cell_type", shuffle = TRUE) + NoLegend()
```

```{r}
# load seurat object
strain.combined.sct <- readRDS(file = "strain.combined.sct_mouseWTMYC.rds")

# Figure 1: Heatmap of cell type specific genes
Idents(strain.combined.sct) <- "cell_type"
levels(strain.combined.sct) <- c("Basal", "Luminal", "Ionocytes", 
                           "Endothelial", "Smooth Muscle","Pericytes", "Fibroblasts", 
                           "Macrophages", "Mast cells", "T cells")

DoHeatmap(subset(strain.combined.sct, downsample = 100), features = cell_markers) +
  scale_fill_viridis()

# UMAPs by cell type, strain, lobe
DimPlot(strain.combined.sct, label = TRUE, raster = TRUE, label.size = 6, repel = TRUE) + ggtitle("Cell Type") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15)) + NoLegend()

DimPlot(strain.combined.sct, label = FALSE, group.by = "strain", shuffle = TRUE, pt.size = 1, raster = TRUE, cols = c("#219ebc", "#ffb703")) + ggtitle("Strain") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

DimPlot(strain.combined.sct, label = FALSE, group.by = "lobe", shuffle = TRUE, pt.size = 1, raster = TRUE, cols = c("#219ebc", "#023047", "#ffb703", "#fb8500")) + ggtitle("Lobe") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

Idents(strain.combined.sct) <- "genotype"
levels(strain.combined.sct) <- c("WT", "MYC")
DimPlot(strain.combined.sct, label = FALSE, shuffle = TRUE, pt.size = 1, raster = TRUE, cols = c("#a8dadc", "#e63946")) + ggtitle("Genotype") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size = 15))

FeaturePlot(strain.combined.sct, features = "MYC", order = TRUE, pt.size = .2, raster = FALSE) + ggtitle("MYC Transgene") +
  theme(text = element_text(size = 15))

```

# Differentially expressed genes between Luminal WT and Hi-MYC
```
Idents(strain.combined.sct) <- "cell_type"
luminal_genotype <- subset(strain.combined.sct, idents = "Luminal") %>% FindMarkers(group.by = "genotype", ident.1 = "MYC", ident.2 = "WT", logfc.threshold = 0.05)
saveRDS(luminal_genotype, file = "luminal_genotype_dge.rds")
luminal_genotype <- readRDS(file = "luminal_genotype_dge.rds")
top_genes_luminal_genotype <- luminal_genotype %>% rownames_to_column("gene")%>% filter(avg_log2FC > 0.1) %>% filter(avg_log2FC > 0.1, pct.2 < 0.05) %>% top_n(50, -p_val_adj)

top_genes_luminal_genotype <- luminal_genotype %>% rownames_to_column("gene")%>% filter(avg_log2FC > 0.2, pct.2 < 0.3) %>% top_n(50, -p_val_adj)

# add lobe specific metadata to luminal cluster
Idents(strain.combined.sct) <- "cell_type"
strain.combined.sct$luminal_lobe <- subset(strain.combined.sct, idents = "Luminal")$lobe
strain.combined.sct$celltype_genotype <- paste(strain.combined.sct$genotype, strain.combined.sct$cell_type, strain.combined.sct$luminal_lobe, sep = " ")
strain.combined.sct$celltype_genotype <- gsub(" NA", "", strain.combined.sct$celltype_genotype)
table(strain.combined.sct$celltype_genotype)

```

Add strain with genotype in metadata
```
# add lobe specific metadata to luminal cluster
Idents(strain.combined.sct) <- "cell_type"
strain.combined.sct$luminal_lobe <- subset(strain.combined.sct, idents = "Luminal")$lobe
strain.combined.sct$strain_celltype_genotype <- paste(strain.combined.sct$genotype, strain.combined.sct$strain, strain.combined.sct$cell_type, strain.combined.sct$luminal_lobe, sep = " ")
strain.combined.sct$strain_celltype_genotype <- gsub(" NA", "", strain.combined.sct$strain_celltype_genotype)
table(strain.combined.sct$strain_celltype_genotype)
```

Heatmap of top luminal genes
```
# make dataframe for heatmap
Idents(strain.combined.sct) <- "cell_type"
AvgExp <- subset(strain.combined.sct, idents = "Luminal") %>% AverageExpression(group.by = "strain_celltype_genotype")
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

AvgExp_top <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% top_genes_luminal_genotype$gene)
AvgExp_top <- AvgExp_top %>% as.matrix()

# column order
sample_order <- c("WT FVB/NJ Luminal Anterior", "WT C57BL/6J Luminal Anterior", 
                  "WT FVB/NJ Luminal Dorsal", "WT C57BL/6J Luminal Dorsal", 
                  "WT FVB/NJ Luminal Lateral", "WT C57BL/6J Luminal Lateral", 
                  "WT FVB/NJ Luminal Ventral", "WT C57BL/6J Luminal Ventral",
                  "MYC FVB/NJ Luminal Anterior", "MYC C57BL/6J Luminal Anterior", 
                  "MYC FVB/NJ Luminal Dorsal",  "MYC C57BL/6J Luminal Dorsal", 
                  "MYC FVB/NJ Luminal Lateral", "MYC C57BL/6J Luminal Lateral", 
                  "MYC FVB/NJ Luminal Ventral", "MYC C57BL/6J Luminal Ventral")

# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# complex heatmap
Heatmap(AvgExp_top, column_order = sample_order, 
        column_dend_reorder = FALSE, 
        show_row_dend = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm")
))
```

#correlate with MYC expression
```

matrix<- strain.combined.sct@assays$SCT@scale.data
matrix_mod<-as.matrix(matrix)
gene<-as.numeric(matrix_mod["MYC",])
correlations<-apply(matrix_mod,1,function(x){cor(gene,x)})
correlations_df <- as.data.frame(correlations) %>% rownames_to_column("gene")

saveRDS(correlations_df, file = "correlations_df.rds")
correlations_df <- readRDS(file = "correlations_df.rds")

MYC_correlations <- correlations_df %>% arrange(desc(correlations)) %>% filter(correlations > 0)
write.csv(MYC_correlations, file = "MYCcorrelations.csv")
#MYC_correlations <- read_csv("MYCcorrelations.csv")

# upload overlap with Gene Ongology, Biological Processes (GO:BP)
library(readr)
GOBP_MYC_overlap <- read_delim("top_100_GO.tsv", 
    delim = "\t", escape_double = FALSE, 
    col_types = cols(`k/K` = col_number(), 
        `p-value` = col_number(), `# Genes in Gene Set (K)` = col_number(), 
        `# Genes in Overlap (k)` = col_number(), 
        `FDR q-value` = col_number()), trim_ws = TRUE, 
    skip = 8)

GOBP_MYC_overlap <- GOBP_MYC_overlap[1:50,]

# convert human back to mouse gene symbol
convertHumanGeneList <- function(x){

require("biomaRt")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")

genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)

humanx <- unique(genesV2)

# Print the first 6 genes found to the screen
print(head(humanx))
return(humanx)
}

Mouse_genes <- convertHumanGeneList(c(MYC_correlations$gene))  
```

# intersection of top genes and myc
```{r}

library(msigdbr)

msigdbr_col <- msigdbr_collections()

pathway_list <- c("KEGG_CELL_CYCLE", "KEGG_RIBOSOME", "KEGG_OXIDATIVE_PHOSPHORYLATION",
                  "KEGG_PURINE_METABOLISM", "KEGG_PYRIMIDINE_METABOLISM",
                  "KEGG_PROTEASOME", "REACTOME_MITOCHONDRIAL_BIOGENESIS")

KEGGgene_sets <- msigdbr(species = "mouse", category = "C2", subcategory = "KEGG")
REACTOMEGgene_sets <- msigdbr(species = "mouse", category = "C2", subcategory = "REACTOME")
gene_sets <- rbind(KEGGgene_sets, REACTOMEGgene_sets)

fgsea_sets <- gene_sets %>% filter(gs_name %in% pathway_list)
fgsea_sets<- fgsea_sets %>% split(x = .$gene_symbol, f = .$gs_name)

#MYC_correlations <- correlations_df %>% arrange(desc(correlations)) %>% top_n(500, correlations)
#write.csv(MYC_correlations, file = "MYC_correlations_positive.csv")
MYC_correlations <- read_csv("MYC_correlations_positive.csv")
```

Heatmap of KEGG gene sets
```{r}
# make dataframe for heatmap
Idents(strain.combined.sct) <- "cell_type"
AvgExp <- subset(strain.combined.sct, idents = "Luminal") %>% AverageExpression(group.by = "strain_celltype_genotype")
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# cell cycle
intersect_cell_cycle_MYC <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_CELL_CYCLE)
MYC_correlations_cellCycle <- MYC_correlations %>% filter(gene %in% intersect_cell_cycle_MYC) %>% top_n(10, correlations)
AvgExp_cell_cycle <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_cellCycle$gene)
AvgExp_cell_cycle <- AvgExp_cell_cycle %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_cellcycle <- rownames(AvgExp_cell_cycle) %>% as.data.frame()
KEGG_cellcycle$Pathway <- "KEGG_CELL_CYCLE"

# ribosome
intersect_ribosome <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_RIBOSOME)
MYC_correlations_ribosome <- MYC_correlations %>% filter(gene %in% intersect_ribosome) %>% top_n(10, correlations)
AvgExp_ribosome <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_ribosome$gene)
AvgExp_ribosome <- AvgExp_ribosome %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_ribosome <- rownames(AvgExp_ribosome)%>% as.data.frame()
KEGG_ribosome$Pathway <- "KEGG_RIBOSOME"

# OxPhos
intersect_OxPhos <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_OXIDATIVE_PHOSPHORYLATION)
MYC_correlations_OxPhos <- MYC_correlations %>% filter(gene %in% intersect_OxPhos) %>% top_n(10, correlations)
AvgExp_OxPhos <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_OxPhos$gene)
AvgExp_OxPhos <- AvgExp_OxPhos %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_OxPhos <- rownames(AvgExp_OxPhos) %>% as.data.frame()
KEGG_OxPhos$Pathway <- "KEGG_OXIDATIVE_PHOSPHORYLATION"

# Purine Metabolism
intersect_Purine <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_PURINE_METABOLISM)
MYC_correlations_Purine <- MYC_correlations %>% filter(gene %in% intersect_Purine) %>% top_n(10, correlations)
AvgExp_Purine <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_Purine$gene)
AvgExp_Purine <- AvgExp_Purine %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_Purine <- rownames(AvgExp_Purine) %>% as.data.frame()
KEGG_Purine$Pathway <- "KEGG_PURINE_METABOLISM"

# Pyrmidine Metabolism
intersect_Pyrmidine <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_PYRIMIDINE_METABOLISM)
MYC_correlations_Pyrmidine <- MYC_correlations %>% filter(gene %in% intersect_Pyrmidine) %>% top_n(10, correlations)
AvgExp_Pyrmidine <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_Pyrmidine$gene)
AvgExp_Pyrmidine <- AvgExp_Pyrmidine %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_Pyrmidine <- rownames(AvgExp_Pyrmidine) %>% as.data.frame()
KEGG_Pyrmidine$Pathway <- "KEGG_PYRIMIDINE_METABOLISM"

intersect(KEGG_Pyrmidine$., KEGG_Pyrmidine$.)
KEGG_Pyrmidine <- rownames(AvgExp_Pyrmidine) %>% as.data.frame()
KEGG_Purine$Pathway <- "KEGG_PYRIMIDINE_METABOLISM \nKEGG_PURINE_METABOLISM"

# Proteasome Metabolism
intersect_Proteasome <- intersect(MYC_correlations$gene, fgsea_sets$KEGG_PROTEASOME)
MYC_correlations_Proteasome <- MYC_correlations %>% filter(gene %in% intersect_Proteasome) %>% top_n(10, correlations)
AvgExp_Proteasome <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_Proteasome$gene)
AvgExp_Proteasome <- AvgExp_Proteasome %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
KEGG_Proteasome <- rownames(AvgExp_Proteasome) %>% as.data.frame()
KEGG_Proteasome$Pathway <- "KEGG_PROTEASOME"

# Mitochondrial Biogenesis
intersect_Mito <- intersect(MYC_correlations$gene, fgsea_sets$REACTOME_MITOCHONDRIAL_BIOGENESIS)
MYC_correlations_Mito <- MYC_correlations %>% filter(gene %in% intersect_Mito)  %>% top_n(10, correlations)
AvgExp_Mito <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% MYC_correlations_Mito$gene)
AvgExp_Mito <- AvgExp_Mito %>% filter(`WT FVB/NJ Luminal Anterior` < 0 )
REACTOME_Mito <- rownames(AvgExp_Mito) %>% as.data.frame()
REACTOME_Mito$Pathway <- "REACTOME_MITOCHONDRIAL_BIOGENESIS"


# gene list
KEGG_genelist <- rbind(KEGG_cellcycle, KEGG_ribosome, KEGG_Purine, KEGG_Proteasome, REACTOME_Mito)
colnames(KEGG_genelist)[1] <- "gene"
KEGG_genelist <- rbind(c("MYC","HUMAN_TRANSGENE"), KEGG_genelist)

```

annotations
```{r}
# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# filtered expression data
# expression matrix
AvgExp_KEGG <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% KEGG_genelist$gene) %>% rownames_to_column("gene")

# add metadata
AvgExp_meta <- merge(AvgExp_KEGG, KEGG_genelist, by.x = "gene", by.y = "gene")
AvgExp_matrix <- AvgExp_meta %>% tibble::column_to_rownames("gene")
AvgExp_matrix <- AvgExp_matrix[-17] %>% as.matrix() %>% t()

AvgExp_meta$Pathway <- factor(AvgExp_meta$Pathway, levels = c("HUMAN_TRANSGENE",
                                                              "KEGG_CELL_CYCLE",
                                                              "KEGG_RIBOSOME",
                                                              "KEGG_PYRIMIDINE_METABOLISM \nKEGG_PURINE_METABOLISM",
                                                              "KEGG_PROTEASOME",
                                                              "REACTOME_MITOCHONDRIAL_BIOGENESIS"))

# Annotations to be added to the top of the heatmap
top.annotations = HeatmapAnnotation(`Biological Pathway` = AvgExp_meta$Pathway,
    col = list(`Biological Pathway` = c("HUMAN_TRANSGENE" = "light grey",
                                        "KEGG_CELL_CYCLE" = "#3B9AB2", 
                   "KEGG_RIBOSOME" = "#78B7C5",
                   "KEGG_PYRIMIDINE_METABOLISM \nKEGG_PURINE_METABOLISM" = "#EBCC2A", 
                   "KEGG_PROTEASOME" = "#E1AF00",
                   "REACTOME_MITOCHONDRIAL_BIOGENESIS" = "#F21A00")))

# sample order
sample_order <- c("WT FVB/NJ Luminal Anterior", "WT C57BL/6J Luminal Anterior", 
                  "WT FVB/NJ Luminal Dorsal", "WT C57BL/6J Luminal Dorsal", 
                  "WT FVB/NJ Luminal Lateral", "WT C57BL/6J Luminal Lateral", 
                  "WT FVB/NJ Luminal Ventral", "WT C57BL/6J Luminal Ventral",
                  "MYC FVB/NJ Luminal Anterior", "MYC C57BL/6J Luminal Anterior", 
                  "MYC FVB/NJ Luminal Dorsal",  "MYC C57BL/6J Luminal Dorsal", 
                  "MYC FVB/NJ Luminal Lateral", "MYC C57BL/6J Luminal Lateral", 
                  "MYC FVB/NJ Luminal Ventral", "MYC C57BL/6J Luminal Ventral")

```

heatmap
```{r}

# complex heatmap
Heatmap(AvgExp_matrix, top_annotation = top.annotations,
        row_order = sample_order,
        column_order = KEGG_genelist$gene,
        column_dend_reorder = FALSE, 
        row_dend_reorder = FALSE, 
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm")
))

# save heatmap objects
# saveRDS(AvgExp_matrix, file = "AvgExp_matrix.rds")
# saveRDS(top.annotations, file = "top.annotations.rds")
# saveRDS(KEGG_genelist, file = "KEGG_genelist.rds")

```
# save seurat object
```
saveRDS(strain.combined.sct, file = "strain.combined.sct_mouseWTMYC.rds")
```