---
title: "Domino HiMyc Epithelial"
author: "Mindy Graham"
date: "2023-01-20"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Perform cell communication analysis on the epithelial subset of the mouse prostate data
```{r}
# Libraries
library(Seurat)
library(dplyr)
library(tidyverse)
library(readr)
library(DataCombine)
library(viridis)
library(ggplot2)
library(glmGamPoi)

# set working directory
wrkdir <- "C:/Users/mgraha21/OneDrive - Johns Hopkins/Mindy JHMI Onedrive/MYC prostate scRNAseq/Manuscript Version of Analysis/Github uploads" #update specific to user
setwd(wrkdir)

# load seurat object
seurat_epi <- readRDS(file = "mouse_seurat_epi.rds")
```

```
DimPlot(seurat_epi)
table(seurat_epi$library_id)

# subset to ~1500 cells
Idents(seurat_epi) <- "epi_clusters_noLobe"
seurat_object_ds <- subset(seurat_epi, downsample = 250)
DimPlot(seurat_object_ds)
seurat_object_ds
table(seurat_object_ds$epi_clusters_noLobe)
table(seurat_object_ds$genotype)

seurat_object_ds
saveRDS(seurat_object_ds, file = "seurat_object_ds.rds")

# get counts data, faster tsv output
data.table::fwrite(t(as.matrix(seurat_object_ds@assays$RNA@counts)), 
                   "counts.tsv", 
                   nThread = 8,
                   sep = '\t', 
                   col.names = TRUE)
```

# initiate docker script
open directory wsl.localhost\Ubuntu-20.04\home\mindykimgraham\docker_scratch  
copy counts generated file to docker scratch directory  
run the bash script to initiate SCENIC: open Linux (Ubuntu-20.04) command line prompt  
ls (shows list of files in directory, confirm that docker scracth folder is there)  
cd (followed by folder name opens directory, typing cd brings back home directory)  
bash ~/docker_scratch2/mgi_scenic.sh  
If working, should see in docker running aertslab/pyscenic:0.10.0  
FYI: this can take a few hours to run. Any dataset with greater than 1500 cells reach computational limits.  
move outputs to current directory to create a Domino object from outputs : auc_mtx.csv, regulons.csv  

edit create_domino function  
devtools::install_github('Chris-Cherry/domino')
```
library(domino)
trace(domino:::convert_genes, edit = TRUE)
https://github.com/Chris-Cherry/domino/issues/14

function (genes, from, to) 
{
  if (from == "ENSMUSG") {
    srcMart = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
    sourceAtts = "ensembl_gene_id"
  }
  if (from == "ENSG") {
    srcMart = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
    sourceAtts = "ensembl_gene_id"
  }
  if (from == "MGI") {
    srcMart = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl", 
      host = "https://dec2021.archive.ensembl.org/")
    sourceAtts = "mgi_symbol"
  }
  if (from == "HGNC") {
    srcMart = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl", 
      host = "https://dec2021.archive.ensembl.org/")
    sourceAtts = "hgnc_symbol"
  }
  if (to == "MGI") {
    tarMart = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl", 
      host = "https://dec2021.archive.ensembl.org/")
    tarAtts = "mgi_symbol"
  }
  if (to == "HGNC") {
    tarMart = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl", 
      host = "https://dec2021.archive.ensembl.org/")
    tarAtts = "hgnc_symbol"
  }
  genesV2 = biomaRt::getLDS(attributes = sourceAtts, filters = sourceAtts, 
    values = genes, mart = srcMart, attributesL = tarAtts, 
    martL = tarMart, uniqueRows = F)
  return(genesV2)
}
```
# create domino object
```
# load downsampled seurat object
seurat_object_ds <- readRDS(file = "seurat_object_ds.rds")
DimPlot(seurat_object_ds)

# load library
library(domino)
library(tidyverse)
library(stringr) 

counts = seurat_object_ds@assays$SCT@counts
z_scores = seurat_object_ds@assays$SCT@scale.data
clusters = seurat_object_ds@active.ident

data <- read.csv(file = "auc_mtx.csv", header = TRUE, 
    stringsAsFactors = FALSE, sep = ',')
rownames(data) <- colnames(seurat_object_ds@assays[["SCT"]]@data)
data <- data[,-1]
auc <- t(data)

df_regulons <- read.csv(file = "regulons.csv")

# create domino object
dom_object = create_domino(signaling_db = "signaling_db/", 
    features = auc, counts = counts, z_scores = z_scores, clusters = clusters, 
    df = df_regulons, gene_conv = c("HGNC", "MGI"))

# saveRDS(dom_object, file = "dom_object.rds")
```

interact with domino object for downstream analysis
```{r}
# load downsampled seurat object
seurat_object_ds <- readRDS(file = "seurat_object_ds.rds")

library(domino)
dom_object <- readRDS(file = "dom_object.rds")

# building networks
dom_object = build_domino(dom_object, max_tf_per_clust = 10, 
    min_tf_pval = .0001, max_rec_per_tf = 10, rec_tf_cor_threshold = 0.25)

# visualize networks
levels(seurat_object_ds)
signaling_network(dom_object, edge_weight = 1, max_thresh = 2.5,
                  cols = c(`Luminal MYC 1` = "#ffb703",
                           Basal = "light grey", `Reactive Basal` = "light grey",
                           Luminal = "light grey", `Reactive Luminal` = "light grey",
                           `Luminal MYC 1` = "light grey", Urothelial = "light grey"))
signaling_df <- dom_object@signaling %>% as.data.frame() %>% arrange(desc(`L_Luminal MYC 1`))
```

# examine what ligand from Luminal MYC1 cells are driving paracrine related changes in TME
```{r}
# Reactive Basal 
signaling_reactive_basal <- dom_object@cl_signaling_matrices[["Reactive Basal"]] %>% as.data.frame() %>% rownames_to_column("Ligand")

# ligands from Luminal MYC 1 to Reactive Basal
signaling_reactive_basal_LuminalMYC1 <- signaling_reactive_basal %>% arrange(desc(`L_Luminal MYC 1`)) %>% head(10)
signaling_reactive_basal_LuminalMYC1

# Ifnk receptors
lig_receptor <- dom_object@linkages[["rec_lig"]]
lig_receptor <- lapply(X = lig_receptor, paste, collapse = ",")
lig_receptor_df <- unlist(lig_receptor) %>% as.data.frame()

Ifnk_receptors <- lig_receptor_df %>% filter(str_detect(., "Ifnk")) %>% rownames()
Ifnk_receptors
FeaturePlot(seurat_epi, features = Ifnk_receptors, order = TRUE)
```

Heatmap of Luminal MYC 1 ligands targeting cluster Reactive Basal
```{r}
# complex heatmap
library(tidyverse)
library(RColorBrewer)
library(ComplexHeatmap)

# expression matrix
Idents(seurat_epi) <- "epi_clusters_noLobe"
AvgExp <- AverageExpression(seurat_epi)
AvgExp_SCT <- AvgExp$SCT

# row standardization
AvgExp_scale <- t(apply(AvgExp_SCT, 1, scale))
colnames(AvgExp_scale) <- colnames(AvgExp_SCT)
AvgExp_scale <- AvgExp_scale %>% na.omit() %>% as.data.frame()

# filter to Luminal MYC 1 ligands
AvgExp_BasalLy6d <- filter(AvgExp_scale, rownames(AvgExp_scale) %in% c(signaling_reactive_basal_LuminalMYC1$Ligand[1:10])) %>% 
  arrange(desc(`Luminal MYC 1`)) %>% as.matrix()
rownames(AvgExp_BasalLy6d)


# colors
RdYlBu <- brewer.pal(6, "RdYlBu")
levels(seurat_object_ds)
column_groups <- c("Luminal MYC 1", "Reactive Luminal", "Reactive Basal","Urothelial", "Basal", "Luminal", "Luminal MYC 2")

# complex heatmap
ht_ligands <- Heatmap(AvgExp_BasalLy6d,
                      row_order = rownames(AvgExp_BasalLy6d),
                      column_order = column_groups,
        show_column_dend = FALSE,
        show_row_dend = FALSE,
        column_title = "Expression of Ligands Targeting Reactive Basal Cluster",
        col = rev(RdYlBu),
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       heatmap_legend_param = list(
    title = "Z-score",
    legend_height = unit(4, "cm")
))

draw(ht_ligands, padding = unit(c(2, 20, 2, 2), "mm"))

```

transcription factor upregulated in Reactive Basal cluster
```{r}
BasalLy6d_TF <- dom_object@linkages[["clust_tf"]][["Reactive Basal"]]

correlations_tf <- dom_object@cor %>% as.data.frame()
cor_Ifnk_receptors <- correlations_tf %>% filter(rownames(correlations_tf) %in% Ifnk_receptors) %>% t() %>% as.data.frame() %>% rownames_to_column("TF")

# filter for TFs upregulated in BasalLy6d
cor_Ifnk_BasalLy6d <- cor_Ifnk_receptors %>% filter(TF %in% BasalLy6d_TF) %>% column_to_rownames("TF") %>% as.matrix()

# which TF is downstream of receptor
dom_object@linkages[["tf_rec"]]$Irf5...
dom_object@linkages[["tf_rec"]]$Olig1...
dom_object@linkages[["tf_rec"]]$Mybl1...
dom_object@linkages[["tf_rec"]]$Foxa2...
dom_object@linkages[["tf_rec"]]$Zfp958...
dom_object@linkages[["tf_rec"]]$Mafb...
dom_object@linkages[["tf_rec"]]$Egr2... # Ifnar1 is upstream of Egr2
dom_object@linkages[["tf_rec"]]$Trp63... # Ifnar1 is upstream of Trp63
dom_object@linkages[["tf_rec"]]$Rarg...
dom_object@linkages[["tf_rec"]]$Sox4... # Ifnar1 is upstream of Sox4

Egr2_targets <- dom_object@linkages[["tf_targets"]][["Egr2"]]
Trp63_targets <- dom_object@linkages[["tf_targets"]][["Trp63"]]
Sox4_targets <- dom_object@linkages[["tf_targets"]][["Sox4"]]
Irf5_targets <- dom_object@linkages[["tf_targets"]][["Irf5"]]
Irf5_targets
```

ligands of receptors that positive correlate with TF activation
```{r}
correlations_tf <- dom_object@cor %>% as.data.frame() 
correlations_tf_t <- t(correlations_tf) %>% as.data.frame()
cor_tf_BasalLy6d <- correlations_tf_t %>% filter(rownames(correlations_tf_t) %in% BasalLy6d_TF) %>% t() %>% as.data.frame() %>% rownames_to_column("receptor")

receptors_1 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[1])) %>% rownames()
receptors_2 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[2])) %>% rownames()
receptors_3 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[3])) %>% rownames()
receptors_4 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[4])) %>% rownames()
receptors_5 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[5])) %>% rownames()
receptors_6 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[6])) %>% rownames()
receptors_7 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[7])) %>% rownames()
receptors_8 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[8])) %>% rownames()
receptors_9 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[9])) %>% rownames()
receptors_10 <- lig_receptor_df %>% filter(str_detect(., signaling_reactive_basal_LuminalMYC1$Ligand[10])) %>% rownames()

receptors <- c(receptors_1, receptors_2, receptors_3, receptors_4, receptors_5,
               receptors_6, receptors_7, receptors_8, receptors_9, receptors_10)

cor_tf_lr <- cor_tf_BasalLy6d %>% filter(receptor %in% receptors) %>% column_to_rownames("receptor") %>% t()
```

make heatmap of correlation plot
```{r}
# colors
BrBG <- brewer.pal(6, "BrBG")

# complex heatmap
ht_tf <- Heatmap(cor_tf_lr,
        show_column_dend = FALSE,
        show_row_dend = FALSE,
        column_title = "Correlation of activated TFs in Reactive Basal Cluster \nwith Receptors of Luminal MYC 1 Ligands",
        col = BrBG,
        row_names_gp = gpar(fontsize = 12),
       column_names_rot = 45,
       row_names_side = "left",
       heatmap_legend_param = list(
    title = "Correlation",
    legend_height = unit(4, "cm")
))
draw(ht_tf, padding = unit(c(2, 20, 2, 2), "mm"))


# highest correlation with Irf5
cor_tf_lr_df <- cor_tf_lr %>% t() %>% as.data.frame() 
cor_tf_lr_df <- cor_tf_lr_df %>% arrange(desc(Irf5...)) %>% head()
cor_tf_lr_df
```

# IRF5, IFNAR 1/2
The interferon response factor 5 has high activity in the reactive basal cluster, with strong correlation with the interferon receptor IFNAR
```{r}
library(wesanderson)
pal <- wes_palette("Zissou1", 6, type = "continuous")

DimPlot(seurat_epi)
Irf5_targets <- c("Gsn", "Tpm2", "Oaf", "Cd47", "Stx11")
VlnPlot(seurat_epi, features = Irf5_targets, stack = TRUE, flip = TRUE, cols = pal) + 
  NoLegend() + ggtitle("IRF5 Targets") 

FeaturePlot(seurat_epi, features = c("Irf5"), order = TRUE, raster = FALSE) & scale_color_gradientn(colours = pal) & NoAxes()
FeaturePlot(seurat_epi, features = c("Ifnar1", "Ifnar2"), order = TRUE, raster = FALSE) & scale_color_gradientn(colours = pal) & NoAxes()

```
